<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Applications | Darwin Core Marine Example Compendium</title>
  <meta name="description" content="Scaling multiple dimensions of diversity in biological data" />
  <meta name="generator" content="bookdown 0.40 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Applications | Darwin Core Marine Example Compendium" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Scaling multiple dimensions of diversity in biological data" />
  <meta name="github-repo" content="ioos/bio_data_guide" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Applications | Darwin Core Marine Example Compendium" />
  
  <meta name="twitter:description" content="Scaling multiple dimensions of diversity in biological data" />
  

<meta name="author" content="By: Standardizing Marine Biological Data Working Group " />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="intro.html"/>
<link rel="next" href="frequently-asked-questions.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Darwin Core Marine Example Compendium</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="applications.html"><a href="applications.html"><i class="fa fa-check"></i><b>2</b> Applications</a>
<ul>
<li class="chapter" data-level="2.1" data-path="applications.html"><a href="applications.html#aligning-data-to-darwin-core---event-core-with-extended-measurement-or-fact"><i class="fa fa-check"></i><b>2.1</b> Aligning Data to Darwin Core - Event Core with Extended Measurement or Fact</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="applications.html"><a href="applications.html#general-information-about-this-notebook"><i class="fa fa-check"></i><b>2.1.1</b> General information about this notebook</a></li>
<li class="chapter" data-level="2.1.2" data-path="applications.html"><a href="applications.html#event-file"><i class="fa fa-check"></i><b>2.1.2</b> Event file</a></li>
<li class="chapter" data-level="2.1.3" data-path="applications.html"><a href="applications.html#occurrence-file"><i class="fa fa-check"></i><b>2.1.3</b> Occurrence file</a></li>
<li class="chapter" data-level="2.1.4" data-path="applications.html"><a href="applications.html#extended-measurement-or-fact-extension-file"><i class="fa fa-check"></i><b>2.1.4</b> Extended Measurement or Fact extension file</a></li>
<li class="chapter" data-level="2.1.5" data-path="applications.html"><a href="applications.html#cleaning-up-event-and-occurrence-files"><i class="fa fa-check"></i><b>2.1.5</b> Cleaning up Event and Occurrence files</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="applications.html"><a href="applications.html#salmon-ocean-ecology-data"><i class="fa fa-check"></i><b>2.2</b> Salmon Ocean Ecology Data</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="applications.html"><a href="applications.html#intro-1"><i class="fa fa-check"></i><b>2.2.1</b> Intro</a></li>
<li class="chapter" data-level="2.2.2" data-path="applications.html"><a href="applications.html#event"><i class="fa fa-check"></i><b>2.2.2</b> event</a></li>
<li class="chapter" data-level="2.2.3" data-path="applications.html"><a href="applications.html#occurrence"><i class="fa fa-check"></i><b>2.2.3</b> occurrence</a></li>
<li class="chapter" data-level="2.2.4" data-path="applications.html"><a href="applications.html#measurementorfact"><i class="fa fa-check"></i><b>2.2.4</b> measurementOrFact</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="applications.html"><a href="applications.html#hakai-seagrass"><i class="fa fa-check"></i><b>2.3</b> Hakai Seagrass</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="applications.html"><a href="applications.html#setup"><i class="fa fa-check"></i><b>2.3.1</b> Setup</a></li>
<li class="chapter" data-level="2.3.2" data-path="applications.html"><a href="applications.html#convert-data-to-darwin-core---extended-measurement-or-fact-format"><i class="fa fa-check"></i><b>2.3.2</b> Convert Data to Darwin Core - Extended Measurement or Fact format</a></li>
<li class="chapter" data-level="2.3.3" data-path="applications.html"><a href="applications.html#session-info"><i class="fa fa-check"></i><b>2.3.3</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="applications.html"><a href="applications.html#trawl-data"><i class="fa fa-check"></i><b>2.4</b> Trawl Data</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="applications.html"><a href="applications.html#workflow-overview"><i class="fa fa-check"></i><b>2.4.1</b> Workflow Overview</a></li>
<li class="chapter" data-level="2.4.2" data-path="applications.html"><a href="applications.html#faq"><i class="fa fa-check"></i><b>2.4.2</b> FAQ</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="applications.html"><a href="applications.html#dataset-edna"><i class="fa fa-check"></i><b>2.5</b> dataset-edna</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="applications.html"><a href="applications.html#introduction"><i class="fa fa-check"></i><b>2.5.1</b> Introduction</a></li>
<li class="chapter" data-level="2.5.2" data-path="applications.html"><a href="applications.html#published-data"><i class="fa fa-check"></i><b>2.5.2</b> Published data</a></li>
<li class="chapter" data-level="2.5.3" data-path="applications.html"><a href="applications.html#repo-structure"><i class="fa fa-check"></i><b>2.5.3</b> Repo structure</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="applications.html"><a href="applications.html#converting-atn-netcdf-file-to-dawrin-core"><i class="fa fa-check"></i><b>2.6</b> Converting ATN netCDF file to Dawrin Core</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="applications.html"><a href="applications.html#downloading-and-preprocessing-the-source-data"><i class="fa fa-check"></i><b>2.6.1</b> Downloading and preprocessing the source data</a></li>
<li class="chapter" data-level="2.6.2" data-path="applications.html"><a href="applications.html#converting-to-darwin-core"><i class="fa fa-check"></i><b>2.6.2</b> Converting to Darwin Core</a></li>
<li class="chapter" data-level="2.6.3" data-path="applications.html"><a href="applications.html#sessioninfo"><i class="fa fa-check"></i><b>2.6.3</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="frequently-asked-questions.html"><a href="frequently-asked-questions.html"><i class="fa fa-check"></i><b>A</b> Frequently Asked Questions</a></li>
<li class="chapter" data-level="B" data-path="tools.html"><a href="tools.html"><i class="fa fa-check"></i><b>B</b> Tools</a>
<ul>
<li class="chapter" data-level="B.1" data-path="tools.html"><a href="tools.html#r"><i class="fa fa-check"></i><b>B.1</b> R</a></li>
<li class="chapter" data-level="B.2" data-path="tools.html"><a href="tools.html#python"><i class="fa fa-check"></i><b>B.2</b> Python</a></li>
<li class="chapter" data-level="B.3" data-path="tools.html"><a href="tools.html#google-sheets"><i class="fa fa-check"></i><b>B.3</b> Google Sheets</a></li>
<li class="chapter" data-level="B.4" data-path="tools.html"><a href="tools.html#validators"><i class="fa fa-check"></i><b>B.4</b> Validators</a></li>
</ul></li>
<li class="chapter" data-level="C" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>C</b> Extras</a>
<ul>
<li class="chapter" data-level="C.1" data-path="extras.html"><a href="extras.html#ecological-metadata-language-eml"><i class="fa fa-check"></i><b>C.1</b> Ecological Metadata Language (EML)</a></li>
<li class="chapter" data-level="C.2" data-path="extras.html"><a href="extras.html#example-using-github-to-resolve-errors"><i class="fa fa-check"></i><b>C.2</b> Example using GitHub to resolve errors</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Darwin Core Marine Example Compendium</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="applications" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">Chapter 2</span> Applications<a href="applications.html#applications" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This chapter contains a series of example applications to convert source data to the Darwin Core standard. You can find
these examples (and more!) in the GitHub repository under the <a href="https://github.com/ioos/bio_data_guide/tree/main/datasets"><code>datasets/</code></a>
directory.</p>
<div id="aligning-data-to-darwin-core---event-core-with-extended-measurement-or-fact" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Aligning Data to Darwin Core - Event Core with Extended Measurement or Fact<a href="applications.html#aligning-data-to-darwin-core---event-core-with-extended-measurement-or-fact" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Abby Benson<br />
January 9, 2022</p>
<div id="general-information-about-this-notebook" class="section level3 hasAnchor" number="2.1.1">
<h3><span class="header-section-number">2.1.1</span> General information about this notebook<a href="applications.html#general-information-about-this-notebook" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Script to process the Texas Parks and Wildlife Department (TPWD) Aransas Bay bag seine data from
the format used by the Houston Advanced Research Center (HARC) for bays in Texas. Taxonomy was processed using a separate script (<a href="https://www.sciencebase.gov/catalog/file/get/53a887f4e4b075096c60cfdd?f=__disk__ab%2F6e%2Ff8%2Fab6ef8426ea328cb6c54d13ee7b6b7ce791d23f8">TPWD_Taxonomy.R</a>) using a taxa list pulled from the pdf “<a href="https://www.sciencebase.gov/catalog/file/get/53a887f4e4b075096c60cfdd?f=__disk__d9%2Ff6%2F46%2Fd9f646b40cf7c6cc3fa77d1e7aebe88f46cf7145">2009 Resource Monitoring Operations Manual</a>”. All original data, processed data and scripts are stored on <a href="https://www.sciencebase.gov/catalog/item/53a887f4e4b075096c60cfdd">an item</a> in USGS ScienceBase.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="applications.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load some of the libraries</span></span>
<span id="cb1-2"><a href="applications.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reshape2)</span>
<span id="cb1-3"><a href="applications.html#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-4"><a href="applications.html#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="applications.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the data</span></span>
<span id="cb2-2"><a href="applications.html#cb2-2" aria-hidden="true" tabindex="-1"></a>BagSeine <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;https://www.sciencebase.gov/catalog/file/get/53a887f4e4b075096c60cfdd?f=__disk__6e%2F6a%2F67%2F6e6a678c41cf928e025fd30339789cc8b893a815&amp;allowOpen=true&quot;</span>, <span class="at">stringsAsFactors=</span><span class="cn">FALSE</span>, <span class="at">strip.white =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Note that if not already done you’ll need to run the <a href="https://www.sciencebase.gov/catalog/file/get/53a887f4e4b075096c60cfdd?f=__disk__ab%2F6e%2Ff8%2Fab6ef8426ea328cb6c54d13ee7b6b7ce791d23f8">TPWD_Taxonomy.R</a> script to get the taxaList file squared away or load the taxonomy file to the World Register of Marine Species Taxon Match Tool <a href="https://www.marinespecies.org/aphia.php?p=match" class="uri">https://www.marinespecies.org/aphia.php?p=match</a></p>
</div>
<div id="event-file" class="section level3 hasAnchor" number="2.1.2">
<h3><span class="header-section-number">2.1.2</span> Event file<a href="applications.html#event-file" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To start we will create the Darwin Core <strong>Event</strong> file. This is the file that will have all the information about the sampling event such as date, location, depth, sampling protocol. Basically anything about the cruise or the way the sampling was done will go in this file. You can see all the Darwin Core terms that are part of the event file here <a href="http://tools.gbif.org/dwca-validator/extension.do?id=dwc:Event" class="uri">http://tools.gbif.org/dwca-validator/extension.do?id=dwc:Event</a>.</p>
<p>The original format for these TPWD HARC files has all of the information associated as the event in the first approximately 50 columns and then all of the information about the occurrence (species) as columns for each species. We will need to start by limiting to the event information only.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="applications.html#cb3-1" aria-hidden="true" tabindex="-1"></a>event <span class="ot">&lt;-</span> BagSeine[,<span class="dv">1</span><span class="sc">:</span><span class="dv">47</span>]</span></code></pre></div>
<p>Next there are several pieces of information that need
1) to be added like the geodeticDatum
2) to be pieced together from multiple columns like datasetID or
3) minor changes like the minimum and maximum depth.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="applications.html#cb4-1" aria-hidden="true" tabindex="-1"></a>event <span class="ot">&lt;-</span> event <span class="sc">%&gt;%</span></span>
<span id="cb4-2"><a href="applications.html#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">type =</span> <span class="st">&quot;Event&quot;</span>,</span>
<span id="cb4-3"><a href="applications.html#cb4-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">modified =</span> lubridate<span class="sc">::</span><span class="fu">today</span>(),</span>
<span id="cb4-4"><a href="applications.html#cb4-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">language =</span> <span class="st">&quot;en&quot;</span>,</span>
<span id="cb4-5"><a href="applications.html#cb4-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">license =</span> <span class="st">&quot;http://creativecommons.org/publicdomain/zero/1.0/legalcode&quot;</span>,</span>
<span id="cb4-6"><a href="applications.html#cb4-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">institutionCode =</span> <span class="st">&quot;TPWD&quot;</span>,</span>
<span id="cb4-7"><a href="applications.html#cb4-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">ownerInstitutionCode =</span> <span class="st">&quot;HARC&quot;</span>,</span>
<span id="cb4-8"><a href="applications.html#cb4-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">coordinateUncertaintyInMeters =</span> <span class="st">&quot;100&quot;</span>,</span>
<span id="cb4-9"><a href="applications.html#cb4-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">geodeticDatum =</span> <span class="st">&quot;WGS84&quot;</span>,</span>
<span id="cb4-10"><a href="applications.html#cb4-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">georeferenceProtocol =</span> <span class="st">&quot;Handheld GPS&quot;</span>,</span>
<span id="cb4-11"><a href="applications.html#cb4-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">country =</span> <span class="st">&quot;United States&quot;</span>,</span>
<span id="cb4-12"><a href="applications.html#cb4-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">countryCode =</span> <span class="st">&quot;US&quot;</span>,</span>
<span id="cb4-13"><a href="applications.html#cb4-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">stateProvince =</span> <span class="st">&quot;Texas&quot;</span>,</span>
<span id="cb4-14"><a href="applications.html#cb4-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">datasetID =</span> <span class="fu">gsub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>, <span class="fu">paste</span>(<span class="st">&quot;TPWD_HARC_Texas&quot;</span>, event<span class="sc">$</span>Bay, event<span class="sc">$</span>Gear_Type)),</span>
<span id="cb4-15"><a href="applications.html#cb4-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">eventID =</span> <span class="fu">paste</span>(<span class="st">&quot;Station&quot;</span>, event<span class="sc">$</span>station_code, <span class="st">&quot;Date&quot;</span>, event<span class="sc">$</span>completion_dttm, <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>),</span>
<span id="cb4-16"><a href="applications.html#cb4-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">sampleSizeUnit =</span> <span class="st">&quot;hectares&quot;</span>,</span>
<span id="cb4-17"><a href="applications.html#cb4-17" aria-hidden="true" tabindex="-1"></a>         <span class="at">CompDate =</span> lubridate<span class="sc">::</span><span class="fu">mdy_hms</span>(event<span class="sc">$</span>CompDate, <span class="at">tz=</span><span class="st">&quot;America/Chicago&quot;</span>), </span>
<span id="cb4-18"><a href="applications.html#cb4-18" aria-hidden="true" tabindex="-1"></a>         <span class="at">StartDate =</span> lubridate<span class="sc">::</span><span class="fu">mdy_hms</span>(event<span class="sc">$</span>StartDate, <span class="at">tz=</span><span class="st">&quot;America/Chicago&quot;</span>),</span>
<span id="cb4-19"><a href="applications.html#cb4-19" aria-hidden="true" tabindex="-1"></a>         <span class="at">minimumDepthInMeters =</span> <span class="fu">ifelse</span>(start_shallow_water_depth_num <span class="sc">&lt;</span> start_deep_water_depth_num, </span>
<span id="cb4-20"><a href="applications.html#cb4-20" aria-hidden="true" tabindex="-1"></a>                                       start_shallow_water_depth_num, start_deep_water_depth_num),</span>
<span id="cb4-21"><a href="applications.html#cb4-21" aria-hidden="true" tabindex="-1"></a>         <span class="at">maximumDepthInMeters =</span> <span class="fu">ifelse</span>(start_deep_water_depth_num <span class="sc">&gt;</span> start_shallow_water_depth_num,</span>
<span id="cb4-22"><a href="applications.html#cb4-22" aria-hidden="true" tabindex="-1"></a>                                       start_deep_water_depth_num, start_shallow_water_depth_num))</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="applications.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(event[,<span class="dv">48</span><span class="sc">:</span><span class="dv">64</span>], <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb5-2"><a href="applications.html#cb5-2" aria-hidden="true" tabindex="-1"></a>    type   modified language                                                    license institutionCode</span>
<span id="cb5-3"><a href="applications.html#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>  Event <span class="dv">2022-01-09</span>       en http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD</span>
<span id="cb5-4"><a href="applications.html#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>  Event <span class="dv">2022-01-09</span>       en http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD</span>
<span id="cb5-5"><a href="applications.html#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>  Event <span class="dv">2022-01-09</span>       en http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD</span>
<span id="cb5-6"><a href="applications.html#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>  Event <span class="dv">2022-01-09</span>       en http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD</span>
<span id="cb5-7"><a href="applications.html#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>  Event <span class="dv">2022-01-09</span>       en http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD</span>
<span id="cb5-8"><a href="applications.html#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>  Event <span class="dv">2022-01-09</span>       en http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD</span>
<span id="cb5-9"><a href="applications.html#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span>  Event <span class="dv">2022-01-09</span>       en http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD</span>
<span id="cb5-10"><a href="applications.html#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="dv">8</span>  Event <span class="dv">2022-01-09</span>       en http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD</span>
<span id="cb5-11"><a href="applications.html#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="dv">9</span>  Event <span class="dv">2022-01-09</span>       en http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD</span>
<span id="cb5-12"><a href="applications.html#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span> Event <span class="dv">2022-01-09</span>       en http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD</span>
<span id="cb5-13"><a href="applications.html#cb5-13" aria-hidden="true" tabindex="-1"></a>   ownerInstitutionCode coordinateUncertaintyInMeters geodeticDatum georeferenceProtocol       country</span>
<span id="cb5-14"><a href="applications.html#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>                  HARC                           <span class="dv">100</span>         WGS84         Handheld GPS United States</span>
<span id="cb5-15"><a href="applications.html#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>                  HARC                           <span class="dv">100</span>         WGS84         Handheld GPS United States</span>
<span id="cb5-16"><a href="applications.html#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>                  HARC                           <span class="dv">100</span>         WGS84         Handheld GPS United States</span>
<span id="cb5-17"><a href="applications.html#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>                  HARC                           <span class="dv">100</span>         WGS84         Handheld GPS United States</span>
<span id="cb5-18"><a href="applications.html#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>                  HARC                           <span class="dv">100</span>         WGS84         Handheld GPS United States</span>
<span id="cb5-19"><a href="applications.html#cb5-19" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>                  HARC                           <span class="dv">100</span>         WGS84         Handheld GPS United States</span>
<span id="cb5-20"><a href="applications.html#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span>                  HARC                           <span class="dv">100</span>         WGS84         Handheld GPS United States</span>
<span id="cb5-21"><a href="applications.html#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="dv">8</span>                  HARC                           <span class="dv">100</span>         WGS84         Handheld GPS United States</span>
<span id="cb5-22"><a href="applications.html#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="dv">9</span>                  HARC                           <span class="dv">100</span>         WGS84         Handheld GPS United States</span>
<span id="cb5-23"><a href="applications.html#cb5-23" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span>                 HARC                           <span class="dv">100</span>         WGS84         Handheld GPS United States</span>
<span id="cb5-24"><a href="applications.html#cb5-24" aria-hidden="true" tabindex="-1"></a>   countryCode stateProvince                             datasetID                                eventID</span>
<span id="cb5-25"><a href="applications.html#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>           US         Texas TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_95_Date_09JAN1997<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">35</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb5-26"><a href="applications.html#cb5-26" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>           US         Texas TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_95_Date_18AUG2000<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">02</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb5-27"><a href="applications.html#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>           US         Texas TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_28JUN2005<span class="sc">:</span><span class="dv">08</span><span class="sc">:</span><span class="dv">41</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb5-28"><a href="applications.html#cb5-28" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>           US         Texas TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_23AUG2006<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">47</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb5-29"><a href="applications.html#cb5-29" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>           US         Texas TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_17OCT2006<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">23</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb5-30"><a href="applications.html#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>           US         Texas TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_19FEB1996<span class="sc">:</span><span class="dv">10</span><span class="sc">:</span><span class="dv">27</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb5-31"><a href="applications.html#cb5-31" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span>           US         Texas TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_11JUN2001<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">12</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb5-32"><a href="applications.html#cb5-32" aria-hidden="true" tabindex="-1"></a><span class="dv">8</span>           US         Texas TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_16MAR1992<span class="sc">:</span><span class="dv">09</span><span class="sc">:</span><span class="dv">46</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb5-33"><a href="applications.html#cb5-33" aria-hidden="true" tabindex="-1"></a><span class="dv">9</span>           US         Texas TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_25SEP1996<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">28</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb5-34"><a href="applications.html#cb5-34" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span>          US         Texas TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_08MAY1997<span class="sc">:</span><span class="dv">13</span><span class="sc">:</span><span class="dv">20</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb5-35"><a href="applications.html#cb5-35" aria-hidden="true" tabindex="-1"></a>   sampleSizeUnit minimumDepthInMeters maximumDepthInMeters</span>
<span id="cb5-36"><a href="applications.html#cb5-36" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>        hectares                  <span class="fl">0.0</span>                  <span class="fl">0.6</span></span>
<span id="cb5-37"><a href="applications.html#cb5-37" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>        hectares                  <span class="fl">0.1</span>                  <span class="fl">0.5</span></span>
<span id="cb5-38"><a href="applications.html#cb5-38" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>        hectares                  <span class="fl">0.4</span>                  <span class="fl">0.6</span></span>
<span id="cb5-39"><a href="applications.html#cb5-39" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>        hectares                  <span class="fl">0.2</span>                  <span class="fl">0.4</span></span>
<span id="cb5-40"><a href="applications.html#cb5-40" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>        hectares                  <span class="fl">0.7</span>                  <span class="fl">0.8</span></span>
<span id="cb5-41"><a href="applications.html#cb5-41" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>        hectares                  <span class="fl">0.1</span>                  <span class="fl">0.3</span></span>
<span id="cb5-42"><a href="applications.html#cb5-42" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span>        hectares                  <span class="fl">0.4</span>                  <span class="fl">0.5</span></span>
<span id="cb5-43"><a href="applications.html#cb5-43" aria-hidden="true" tabindex="-1"></a><span class="dv">8</span>        hectares                  <span class="fl">0.0</span>                  <span class="fl">0.4</span></span>
<span id="cb5-44"><a href="applications.html#cb5-44" aria-hidden="true" tabindex="-1"></a><span class="dv">9</span>        hectares                  <span class="fl">0.3</span>                  <span class="fl">0.7</span></span>
<span id="cb5-45"><a href="applications.html#cb5-45" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span>       hectares                  <span class="fl">0.4</span>                  <span class="fl">0.6</span></span></code></pre></div>
<p>For this dataset there was a start timestamp and end timestamp that we can use to identify the sampling effort which can be really valuable information for downstream users when trying to reuse data from multiple projects.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="applications.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Calculate duration of bag seine event</span></span>
<span id="cb6-2"><a href="applications.html#cb6-2" aria-hidden="true" tabindex="-1"></a>event<span class="sc">$</span>samplingEffort <span class="ot">&lt;-</span> <span class="st">&quot;&quot;</span></span>
<span id="cb6-3"><a href="applications.html#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(event)){</span>
<span id="cb6-4"><a href="applications.html#cb6-4" aria-hidden="true" tabindex="-1"></a>  event[i,]<span class="sc">$</span>samplingEffort <span class="ot">&lt;-</span> <span class="fu">abs</span>(lubridate<span class="sc">::</span><span class="fu">as.duration</span>(event[i,]<span class="sc">$</span>CompDate <span class="sc">-</span> event[i,]<span class="sc">$</span>StartDate))</span>
<span id="cb6-5"><a href="applications.html#cb6-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb6-6"><a href="applications.html#cb6-6" aria-hidden="true" tabindex="-1"></a>event<span class="sc">$</span>samplingEffort <span class="ot">&lt;-</span> <span class="fu">paste</span>(event<span class="sc">$</span>samplingEffort, <span class="st">&quot;seconds&quot;</span>, <span class="at">sep =</span> <span class="st">&quot; &quot;</span>)</span></code></pre></div>
<p>Finally there were a few columns that were a direct match to a Darwin Core term and therefore just need to be renamed to follow the standard.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="applications.html#cb7-1" aria-hidden="true" tabindex="-1"></a>event <span class="ot">&lt;-</span> event <span class="sc">%&gt;%</span></span>
<span id="cb7-2"><a href="applications.html#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">samplingProtocol =</span> Gear_Type,</span>
<span id="cb7-3"><a href="applications.html#cb7-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">locality =</span> Estuary,</span>
<span id="cb7-4"><a href="applications.html#cb7-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">waterBody =</span> SubBay,</span>
<span id="cb7-5"><a href="applications.html#cb7-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">decimalLatitude =</span> Latitude,</span>
<span id="cb7-6"><a href="applications.html#cb7-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">decimalLongitude =</span> Longitude,</span>
<span id="cb7-7"><a href="applications.html#cb7-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">sampleSizeValue =</span> surface_area_num,</span>
<span id="cb7-8"><a href="applications.html#cb7-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">eventDate =</span> CompDate)</span></code></pre></div>
</div>
<div id="occurrence-file" class="section level3 hasAnchor" number="2.1.3">
<h3><span class="header-section-number">2.1.3</span> Occurrence file<a href="applications.html#occurrence-file" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The next file we need to create is the <strong>Occurrence</strong> file. This file includes all the information about the species that were observed. An occurrence in Darwin Core is the intersection of an organism at a time and a place. We have already done the work to identify the time and place in the event file so we don’t need to do that again here. What we do need to is identify all the information about the organisms. Another piece of information that goes in here is basisOfRecord which is a required field and has a controlled vocabulary. For the data we work with you’ll usually put <code>HumanObservation</code> or <code>MachineObservation</code>. If it’s eDNA data you’ll use <code>MaterialSample</code>. If your data are part of a museum collection you’ll use <code>PreservedSpecimen</code>.</p>
<p>Important to note that there is overlap in the Darwin Core terms that “allowed” to be in the event file and in the occurrence file. This is because data can be submitted as “Occurrence Only” where you don’t have a separate event file. In that case, the location and date information will need to be included in the occurrence file. Since we are formatting this dataset as a sampling event we will not include location and date information in the occurrence file. To see all the Darwin Core terms that can go in the occurrence file go here <a href="https://tools.gbif.org/dwca-validator/extension.do?id=dwc:occurrence" class="uri">https://tools.gbif.org/dwca-validator/extension.do?id=dwc:occurrence</a>.</p>
<p>This dataset in its original format is in “wide format”. All that means is that data that we would expect to be encoded as values in the rows are instead column headers. We have to pull all the scientific names out of the column headers and turn them into actual values in the data.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="applications.html#cb8-1" aria-hidden="true" tabindex="-1"></a>occurrence <span class="ot">&lt;-</span> <span class="fu">melt</span>(BagSeine, <span class="at">id=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">47</span>, <span class="at">measure=</span><span class="dv">48</span><span class="sc">:</span><span class="dv">109</span>, <span class="at">variable.name=</span><span class="st">&quot;vernacularName&quot;</span>, <span class="at">value.name=</span><span class="st">&quot;relativeAbundance&quot;</span>)</span></code></pre></div>
<p>You’ll notice when we did that step we went from 5481 obs (or rows) in the data to 334341 obs. We went from wide to long.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="applications.html#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(BagSeine)</span>
<span id="cb9-2"><a href="applications.html#cb9-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">5481</span>  <span class="dv">109</span></span>
<span id="cb9-3"><a href="applications.html#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(occurrence)</span>
<span id="cb9-4"><a href="applications.html#cb9-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">334341</span>     <span class="dv">49</span></span></code></pre></div>
<p>Now as with the event file we have several pieces of information that need to be added or changed to make sure the data are following Darwin Core. We always want to include as much information as possible to make the data as reusable as possible.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="applications.html#cb10-1" aria-hidden="true" tabindex="-1"></a>occurrence <span class="ot">&lt;-</span> occurrence <span class="sc">%&gt;%</span></span>
<span id="cb10-2"><a href="applications.html#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">vernacularName =</span> <span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">.&quot;</span>,<span class="st">&#39; &#39;</span>, vernacularName),</span>
<span id="cb10-3"><a href="applications.html#cb10-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">eventID =</span> <span class="fu">paste</span>(<span class="st">&quot;Station&quot;</span>, station_code, <span class="st">&quot;Date&quot;</span>, completion_dttm, <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>),</span>
<span id="cb10-4"><a href="applications.html#cb10-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceStatus =</span> <span class="fu">ifelse</span>(relativeAbundance <span class="sc">==</span> <span class="dv">0</span>, <span class="st">&quot;Absent&quot;</span>, <span class="st">&quot;Present&quot;</span>),</span>
<span id="cb10-5"><a href="applications.html#cb10-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">basisOfRecord =</span> <span class="st">&quot;HumanObservation&quot;</span>,</span>
<span id="cb10-6"><a href="applications.html#cb10-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">organismQuantityType =</span> <span class="st">&quot;Relative Abundance&quot;</span>,</span>
<span id="cb10-7"><a href="applications.html#cb10-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">collectionCode =</span> <span class="fu">paste</span>(Bay, Gear_Type, <span class="at">sep =</span> <span class="st">&quot; &quot;</span>))</span></code></pre></div>
<p>We will match the taxa list with our occurrence file data to bring in the taxonomic information that we pulled from WoRMS. To save time you’ll just import the processed taxa list which includes the taxonomic hierarchy and the required term scientificNameID which is one of the most important pieces of information to include for OBIS.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="applications.html#cb11-1" aria-hidden="true" tabindex="-1"></a>taxaList <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;https://www.sciencebase.gov/catalog/file/get/53a887f4e4b075096c60cfdd?f=__disk__49%2F0a%2F73%2F490a7337fa94039715809496b22f5d003b8a79a2&amp;allowOpen=true&quot;</span>, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-2"><a href="applications.html#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Merge taxaList with occurrence</span></span>
<span id="cb11-3"><a href="applications.html#cb11-3" aria-hidden="true" tabindex="-1"></a>occurrence <span class="ot">&lt;-</span> <span class="fu">merge</span>(occurrence, taxaList, <span class="at">by =</span> <span class="st">&quot;vernacularName&quot;</span>, <span class="at">all.x =</span> T)</span>
<span id="cb11-4"><a href="applications.html#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Test that all the vernacularNames found a match in taxaList_updated</span></span>
<span id="cb11-5"><a href="applications.html#cb11-5" aria-hidden="true" tabindex="-1"></a>Hmisc<span class="sc">::</span><span class="fu">describe</span>(occurrence<span class="sc">$</span>scientificNameID)</span>
<span id="cb11-6"><a href="applications.html#cb11-6" aria-hidden="true" tabindex="-1"></a>       n  missing distinct </span>
<span id="cb11-7"><a href="applications.html#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="dv">334341</span>        <span class="dv">0</span>       <span class="dv">61</span> </span>
<span id="cb11-8"><a href="applications.html#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="applications.html#cb11-9" aria-hidden="true" tabindex="-1"></a>lowest <span class="sc">:</span> urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">105792</span> urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">107034</span> urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">107379</span> urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">126983</span> urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">127089</span></span>
<span id="cb11-10"><a href="applications.html#cb11-10" aria-hidden="true" tabindex="-1"></a>highest<span class="sc">:</span> urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">367528</span> urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">396707</span> urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">421784</span> urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">422069</span> urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">443955</span></span></code></pre></div>
<p>For that last line of code we are expecting to see no missing values for scientificNameID. Every row in the file should have a value in scientificNameID which should be a WoRMS LSID that look like this <code>urn:lsid:marinespecies.org:taxname:144531</code></p>
<p>We need to create a unique ID for each row in the occurrence file. This is known as the <code>occurrenceID</code> and is a required term. The <code>occurrenceID</code> needs to be globally unique and needs to be permanent and kept in place if any updates to the dataset are made. You should not create brand new occurrenceIDs when you update a dataset. To facilitate this I like to build the <code>occurrenceID</code> from pieces of information available in the dataset to create a unique ID for each row in the occurrence file. For this dataset I used the <code>eventID</code> (Station + Date) plus the scientific name. This only works if there is only one scientific name per station per date so if you have different ages or sexes of species at the same station and date this method of creating the occurrenceID won’t work for you.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="applications.html#cb12-1" aria-hidden="true" tabindex="-1"></a>occurrence<span class="sc">$</span>occurrenceID <span class="ot">&lt;-</span> <span class="fu">paste</span>(occurrence<span class="sc">$</span>eventID, <span class="fu">gsub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>,occurrence<span class="sc">$</span>scientificName), <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>)</span>
<span id="cb12-2"><a href="applications.html#cb12-2" aria-hidden="true" tabindex="-1"></a>occurrence[<span class="dv">1</span>,]<span class="sc">$</span>occurrenceID</span>
<span id="cb12-3"><a href="applications.html#cb12-3" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">&quot;Station_95_Date_09JAN1997:14:35:00.000_Atractosteus_spatula&quot;</span></span></code></pre></div>
<p>For the occurrence file we only have one column to rename. We could have avoided this step if we had named it <code>organismQuantity</code> up above but I kept this to remind me what the data providers had called this.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="applications.html#cb13-1" aria-hidden="true" tabindex="-1"></a>occurrence <span class="ot">&lt;-</span> occurrence <span class="sc">%&gt;%</span></span>
<span id="cb13-2"><a href="applications.html#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">organismQuantity =</span> relativeAbundance)</span></code></pre></div>
</div>
<div id="extended-measurement-or-fact-extension-file" class="section level3 hasAnchor" number="2.1.4">
<h3><span class="header-section-number">2.1.4</span> Extended Measurement or Fact extension file<a href="applications.html#extended-measurement-or-fact-extension-file" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The final file we are going to create is the <strong>Extended Measurement or Fact extension (emof)</strong>. This is a bit like a catch all for any measurements or facts that are not captured in Darwin Core. Darwin Core does not have terms for things like temperature, salinity, gear type, cruise number, length, weight, etc. We are going to create a long format file where each of these is a set of rows in the extended measurement or fact file. You can find all the terms in this extension here <a href="https://tools.gbif.org/dwca-validator/extension.do?id=http://rs.iobis.org/obis/terms/ExtendedMeasurementOrFact" class="uri">https://tools.gbif.org/dwca-validator/extension.do?id=http://rs.iobis.org/obis/terms/ExtendedMeasurementOrFact</a>.</p>
<p>OBIS uses the BODC NERC Vocabulary Server to provide explicit definitions for each of the measurements <a href="https://vocab.nerc.ac.uk/search_nvs/" class="uri">https://vocab.nerc.ac.uk/search_nvs/</a>.</p>
<p>For this dataset I was only able to find code definitions provided by the data providers for some of the measurements. I included the ones that I was able to find code definitions and left out any that I couldn’t find those for. The ones I was able to find code definitions for were <code>Total.Of.Samples_Count</code>, <code>gear_size</code>, <code>start_wind_speed_num</code>, <code>start_barometric_pressure_num</code>, <code>start_temperature_num</code>, <code>start_salinity_num</code>, <code>start_dissolved_oxygen_num</code>. All the others I left out.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="applications.html#cb14-1" aria-hidden="true" tabindex="-1"></a>totalOfSamples <span class="ot">&lt;-</span> event[<span class="fu">c</span>(<span class="st">&quot;Total.Of.Samples_Count&quot;</span>, <span class="st">&quot;eventID&quot;</span>)]</span>
<span id="cb14-2"><a href="applications.html#cb14-2" aria-hidden="true" tabindex="-1"></a>totalOfSamples <span class="ot">&lt;-</span> totalOfSamples[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(totalOfSamples<span class="sc">$</span>Total.Of.Samples_Count)),]</span>
<span id="cb14-3"><a href="applications.html#cb14-3" aria-hidden="true" tabindex="-1"></a>totalOfSamples <span class="ot">&lt;-</span> totalOfSamples <span class="sc">%&gt;%</span> </span>
<span id="cb14-4"><a href="applications.html#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;Total number of samples used to calculate relative abundance&quot;</span>,</span>
<span id="cb14-5"><a href="applications.html#cb14-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnit =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb14-6"><a href="applications.html#cb14-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementTypeID =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb14-7"><a href="applications.html#cb14-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb14-8"><a href="applications.html#cb14-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-9"><a href="applications.html#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">measurementValue =</span> Total.Of.Samples_Count)</span>
<span id="cb14-10"><a href="applications.html#cb14-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-11"><a href="applications.html#cb14-11" aria-hidden="true" tabindex="-1"></a>gear_size <span class="ot">&lt;-</span> event[<span class="fu">c</span>(<span class="st">&quot;gear_size&quot;</span>, <span class="st">&quot;eventID&quot;</span>)]</span>
<span id="cb14-12"><a href="applications.html#cb14-12" aria-hidden="true" tabindex="-1"></a>gear_size <span class="ot">&lt;-</span> gear_size[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(gear_size<span class="sc">$</span>gear_size)),]</span>
<span id="cb14-13"><a href="applications.html#cb14-13" aria-hidden="true" tabindex="-1"></a>gear_size <span class="ot">&lt;-</span> gear_size <span class="sc">%&gt;%</span> </span>
<span id="cb14-14"><a href="applications.html#cb14-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;gear size&quot;</span>,</span>
<span id="cb14-15"><a href="applications.html#cb14-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnit =</span> <span class="st">&quot;meters&quot;</span>,</span>
<span id="cb14-16"><a href="applications.html#cb14-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementTypeID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P01/current/MTHAREA1/&quot;</span>,</span>
<span id="cb14-17"><a href="applications.html#cb14-17" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P06/current/ULAA/&quot;</span>,</span>
<span id="cb14-18"><a href="applications.html#cb14-18" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-19"><a href="applications.html#cb14-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">measurementValue =</span> gear_size)</span>
<span id="cb14-20"><a href="applications.html#cb14-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-21"><a href="applications.html#cb14-21" aria-hidden="true" tabindex="-1"></a>start_wind_speed_num <span class="ot">&lt;-</span> event[<span class="fu">c</span>(<span class="st">&quot;start_wind_speed_num&quot;</span>, <span class="st">&quot;eventID&quot;</span>)]</span>
<span id="cb14-22"><a href="applications.html#cb14-22" aria-hidden="true" tabindex="-1"></a>start_wind_speed_num <span class="ot">&lt;-</span> start_wind_speed_num[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(start_wind_speed_num<span class="sc">$</span>start_wind_speed_num)),]</span>
<span id="cb14-23"><a href="applications.html#cb14-23" aria-hidden="true" tabindex="-1"></a>start_wind_speed_num <span class="ot">&lt;-</span> start_wind_speed_num <span class="sc">%&gt;%</span> </span>
<span id="cb14-24"><a href="applications.html#cb14-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;wind speed&quot;</span>,</span>
<span id="cb14-25"><a href="applications.html#cb14-25" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnit =</span> <span class="st">&quot;not provided&quot;</span>,</span>
<span id="cb14-26"><a href="applications.html#cb14-26" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementTypeID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P01/current/EWSBZZ01/&quot;</span>,</span>
<span id="cb14-27"><a href="applications.html#cb14-27" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb14-28"><a href="applications.html#cb14-28" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-29"><a href="applications.html#cb14-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">measurementValue =</span> start_wind_speed_num)</span>
<span id="cb14-30"><a href="applications.html#cb14-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-31"><a href="applications.html#cb14-31" aria-hidden="true" tabindex="-1"></a>start_barometric_pressure_num <span class="ot">&lt;-</span> event[<span class="fu">c</span>(<span class="st">&quot;start_barometric_pressure_num&quot;</span>, <span class="st">&quot;eventID&quot;</span>)]</span>
<span id="cb14-32"><a href="applications.html#cb14-32" aria-hidden="true" tabindex="-1"></a>start_barometric_pressure_num <span class="ot">&lt;-</span> start_barometric_pressure_num[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(start_barometric_pressure_num<span class="sc">$</span>start_barometric_pressure_num)),]</span>
<span id="cb14-33"><a href="applications.html#cb14-33" aria-hidden="true" tabindex="-1"></a>start_barometric_pressure_num <span class="ot">&lt;-</span> start_barometric_pressure_num <span class="sc">%&gt;%</span> </span>
<span id="cb14-34"><a href="applications.html#cb14-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;barometric pressure&quot;</span>,</span>
<span id="cb14-35"><a href="applications.html#cb14-35" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnit =</span> <span class="st">&quot;not provided&quot;</span>,</span>
<span id="cb14-36"><a href="applications.html#cb14-36" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementTypeID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P07/current/CFSN0015/&quot;</span>,</span>
<span id="cb14-37"><a href="applications.html#cb14-37" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb14-38"><a href="applications.html#cb14-38" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-39"><a href="applications.html#cb14-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">measurementValue =</span> start_barometric_pressure_num)</span>
<span id="cb14-40"><a href="applications.html#cb14-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-41"><a href="applications.html#cb14-41" aria-hidden="true" tabindex="-1"></a>start_temperature_num <span class="ot">&lt;-</span> event[<span class="fu">c</span>(<span class="st">&quot;start_temperature_num&quot;</span>, <span class="st">&quot;eventID&quot;</span>)]</span>
<span id="cb14-42"><a href="applications.html#cb14-42" aria-hidden="true" tabindex="-1"></a>start_temperature_num <span class="ot">&lt;-</span> start_temperature_num[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(start_temperature_num<span class="sc">$</span>start_temperature_num)),]</span>
<span id="cb14-43"><a href="applications.html#cb14-43" aria-hidden="true" tabindex="-1"></a>start_temperature_num <span class="ot">&lt;-</span> start_temperature_num <span class="sc">%&gt;%</span> </span>
<span id="cb14-44"><a href="applications.html#cb14-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;water temperature&quot;</span>,</span>
<span id="cb14-45"><a href="applications.html#cb14-45" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnit =</span> <span class="st">&quot;Celsius&quot;</span>,</span>
<span id="cb14-46"><a href="applications.html#cb14-46" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementTypeID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P01/current/TEMPPR01/&quot;</span>,</span>
<span id="cb14-47"><a href="applications.html#cb14-47" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P06/current/UPAA/&quot;</span>,</span>
<span id="cb14-48"><a href="applications.html#cb14-48" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-49"><a href="applications.html#cb14-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">measurementValue =</span> start_temperature_num)</span>
<span id="cb14-50"><a href="applications.html#cb14-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-51"><a href="applications.html#cb14-51" aria-hidden="true" tabindex="-1"></a>start_salinity_num <span class="ot">&lt;-</span> event[<span class="fu">c</span>(<span class="st">&quot;start_salinity_num&quot;</span>, <span class="st">&quot;eventID&quot;</span>)]</span>
<span id="cb14-52"><a href="applications.html#cb14-52" aria-hidden="true" tabindex="-1"></a>start_salinity_num <span class="ot">&lt;-</span> start_salinity_num[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(start_salinity_num<span class="sc">$</span>start_salinity_num)),]</span>
<span id="cb14-53"><a href="applications.html#cb14-53" aria-hidden="true" tabindex="-1"></a>start_salinity_num <span class="ot">&lt;-</span> start_salinity_num <span class="sc">%&gt;%</span> </span>
<span id="cb14-54"><a href="applications.html#cb14-54" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;salinity&quot;</span>,</span>
<span id="cb14-55"><a href="applications.html#cb14-55" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnit =</span> <span class="st">&quot;ppt&quot;</span>,</span>
<span id="cb14-56"><a href="applications.html#cb14-56" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementTypeID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P01/current/ODSDM021/&quot;</span>,</span>
<span id="cb14-57"><a href="applications.html#cb14-57" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P06/current/UPPT/&quot;</span>,</span>
<span id="cb14-58"><a href="applications.html#cb14-58" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-59"><a href="applications.html#cb14-59" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">measurementValue =</span> start_salinity_num)</span>
<span id="cb14-60"><a href="applications.html#cb14-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-61"><a href="applications.html#cb14-61" aria-hidden="true" tabindex="-1"></a>start_dissolved_oxygen_num <span class="ot">&lt;-</span> event[<span class="fu">c</span>(<span class="st">&quot;start_dissolved_oxygen_num&quot;</span>, <span class="st">&quot;eventID&quot;</span>)]</span>
<span id="cb14-62"><a href="applications.html#cb14-62" aria-hidden="true" tabindex="-1"></a>start_dissolved_oxygen_num <span class="ot">&lt;-</span> start_dissolved_oxygen_num[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(start_dissolved_oxygen_num<span class="sc">$</span>start_dissolved_oxygen_num)),]</span>
<span id="cb14-63"><a href="applications.html#cb14-63" aria-hidden="true" tabindex="-1"></a>start_dissolved_oxygen_num <span class="ot">&lt;-</span> start_dissolved_oxygen_num <span class="sc">%&gt;%</span> </span>
<span id="cb14-64"><a href="applications.html#cb14-64" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;dissolved oxygen&quot;</span>,</span>
<span id="cb14-65"><a href="applications.html#cb14-65" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnit =</span> <span class="st">&quot;ppm&quot;</span>,</span>
<span id="cb14-66"><a href="applications.html#cb14-66" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementTypeID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P09/current/DOX2/&quot;</span>,</span>
<span id="cb14-67"><a href="applications.html#cb14-67" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P06/current/UPPM/&quot;</span>,</span>
<span id="cb14-68"><a href="applications.html#cb14-68" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-69"><a href="applications.html#cb14-69" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">measurementValue =</span> start_dissolved_oxygen_num)</span>
<span id="cb14-70"><a href="applications.html#cb14-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-71"><a href="applications.html#cb14-71" aria-hidden="true" tabindex="-1"></a>alternate_station_code <span class="ot">&lt;-</span> event[<span class="fu">c</span>(<span class="st">&quot;alternate_station_code&quot;</span>, <span class="st">&quot;eventID&quot;</span>)]</span>
<span id="cb14-72"><a href="applications.html#cb14-72" aria-hidden="true" tabindex="-1"></a>alternate_station_code <span class="ot">&lt;-</span> alternate_station_code[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(alternate_station_code<span class="sc">$</span>alternate_station_code)),]</span>
<span id="cb14-73"><a href="applications.html#cb14-73" aria-hidden="true" tabindex="-1"></a>alternate_station_code <span class="ot">&lt;-</span> alternate_station_code <span class="sc">%&gt;%</span> </span>
<span id="cb14-74"><a href="applications.html#cb14-74" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;alternate station code&quot;</span>,</span>
<span id="cb14-75"><a href="applications.html#cb14-75" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnit =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb14-76"><a href="applications.html#cb14-76" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementTypeID =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb14-77"><a href="applications.html#cb14-77" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb14-78"><a href="applications.html#cb14-78" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-79"><a href="applications.html#cb14-79" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">measurementValue =</span> alternate_station_code)</span>
<span id="cb14-80"><a href="applications.html#cb14-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-81"><a href="applications.html#cb14-81" aria-hidden="true" tabindex="-1"></a>organismQuantity <span class="ot">&lt;-</span> occurrence[<span class="fu">c</span>(<span class="st">&quot;organismQuantity&quot;</span>, <span class="st">&quot;eventID&quot;</span>, <span class="st">&quot;occurrenceID&quot;</span>)]</span>
<span id="cb14-82"><a href="applications.html#cb14-82" aria-hidden="true" tabindex="-1"></a>organismQuantity <span class="ot">&lt;-</span> organismQuantity[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(organismQuantity<span class="sc">$</span>organismQuantity)),]</span>
<span id="cb14-83"><a href="applications.html#cb14-83" aria-hidden="true" tabindex="-1"></a>organismQuantity <span class="ot">&lt;-</span> organismQuantity <span class="sc">%&gt;%</span> </span>
<span id="cb14-84"><a href="applications.html#cb14-84" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;relative abundance&quot;</span>,</span>
<span id="cb14-85"><a href="applications.html#cb14-85" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnit =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb14-86"><a href="applications.html#cb14-86" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementTypeID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/S06/current/S0600020/&quot;</span>,</span>
<span id="cb14-87"><a href="applications.html#cb14-87" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-88"><a href="applications.html#cb14-88" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">measurementValue =</span> organismQuantity)</span>
<span id="cb14-89"><a href="applications.html#cb14-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-90"><a href="applications.html#cb14-90" aria-hidden="true" tabindex="-1"></a><span class="co"># Bind the separate measurements together into one file  </span></span>
<span id="cb14-91"><a href="applications.html#cb14-91" aria-hidden="true" tabindex="-1"></a>mof <span class="ot">&lt;-</span> <span class="fu">rbind</span>(totalOfSamples, start_barometric_pressure_num, start_dissolved_oxygen_num, </span>
<span id="cb14-92"><a href="applications.html#cb14-92" aria-hidden="true" tabindex="-1"></a>             start_salinity_num, start_temperature_num, start_wind_speed_num, gear_size,</span>
<span id="cb14-93"><a href="applications.html#cb14-93" aria-hidden="true" tabindex="-1"></a>             alternate_station_code, organismQuantity)</span>
<span id="cb14-94"><a href="applications.html#cb14-94" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mof)</span>
<span id="cb14-95"><a href="applications.html#cb14-95" aria-hidden="true" tabindex="-1"></a> measurementValue                                eventID</span>
<span id="cb14-96"><a href="applications.html#cb14-96" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>               <span class="dv">18</span> Station_95_Date_09JAN1997<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">35</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb14-97"><a href="applications.html#cb14-97" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>              <span class="dv">103</span> Station_95_Date_18AUG2000<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">02</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb14-98"><a href="applications.html#cb14-98" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>              <span class="dv">401</span> Station_96_Date_28JUN2005<span class="sc">:</span><span class="dv">08</span><span class="sc">:</span><span class="dv">41</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb14-99"><a href="applications.html#cb14-99" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>               <span class="dv">35</span> Station_96_Date_23AUG2006<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">47</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb14-100"><a href="applications.html#cb14-100" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>               <span class="dv">57</span> Station_96_Date_17OCT2006<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">23</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb14-101"><a href="applications.html#cb14-101" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>                <span class="dv">5</span> Station_96_Date_19FEB1996<span class="sc">:</span><span class="dv">10</span><span class="sc">:</span><span class="dv">27</span><span class="sc">:</span><span class="fl">00.000</span></span>
<span id="cb14-102"><a href="applications.html#cb14-102" aria-hidden="true" tabindex="-1"></a>                                               measurementType measurementUnit measurementTypeID</span>
<span id="cb14-103"><a href="applications.html#cb14-103" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> Total number of samples used to calculate relative abundance                                  </span>
<span id="cb14-104"><a href="applications.html#cb14-104" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> Total number of samples used to calculate relative abundance                                  </span>
<span id="cb14-105"><a href="applications.html#cb14-105" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> Total number of samples used to calculate relative abundance                                  </span>
<span id="cb14-106"><a href="applications.html#cb14-106" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> Total number of samples used to calculate relative abundance                                  </span>
<span id="cb14-107"><a href="applications.html#cb14-107" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> Total number of samples used to calculate relative abundance                                  </span>
<span id="cb14-108"><a href="applications.html#cb14-108" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span> Total number of samples used to calculate relative abundance                                  </span>
<span id="cb14-109"><a href="applications.html#cb14-109" aria-hidden="true" tabindex="-1"></a>  measurementUnitID occurrenceID</span>
<span id="cb14-110"><a href="applications.html#cb14-110" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>                               </span>
<span id="cb14-111"><a href="applications.html#cb14-111" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>                               </span>
<span id="cb14-112"><a href="applications.html#cb14-112" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>                               </span>
<span id="cb14-113"><a href="applications.html#cb14-113" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>                               </span>
<span id="cb14-114"><a href="applications.html#cb14-114" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>                               </span>
<span id="cb14-115"><a href="applications.html#cb14-115" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>                               </span>
<span id="cb14-116"><a href="applications.html#cb14-116" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(mof)</span>
<span id="cb14-117"><a href="applications.html#cb14-117" aria-hidden="true" tabindex="-1"></a>       measurementValue                                 eventID    measurementType measurementUnit</span>
<span id="cb14-118"><a href="applications.html#cb14-118" aria-hidden="true" tabindex="-1"></a><span class="dv">334336</span>        <span class="fl">0.0000000</span> Station_217_Date_03APR2003<span class="sc">:</span><span class="dv">13</span><span class="sc">:</span><span class="dv">28</span><span class="sc">:</span><span class="fl">00.000</span> relative abundance                </span>
<span id="cb14-119"><a href="applications.html#cb14-119" aria-hidden="true" tabindex="-1"></a><span class="dv">334337</span>        <span class="fl">0.0000000</span> Station_217_Date_24FEB2006<span class="sc">:</span><span class="dv">10</span><span class="sc">:</span><span class="dv">12</span><span class="sc">:</span><span class="fl">00.000</span> relative abundance                </span>
<span id="cb14-120"><a href="applications.html#cb14-120" aria-hidden="true" tabindex="-1"></a><span class="dv">334338</span>        <span class="fl">0.1428571</span> Station_217_Date_23JUN2001<span class="sc">:</span><span class="dv">12</span><span class="sc">:</span><span class="dv">28</span><span class="sc">:</span><span class="fl">00.000</span> relative abundance                </span>
<span id="cb14-121"><a href="applications.html#cb14-121" aria-hidden="true" tabindex="-1"></a><span class="dv">334339</span>        <span class="fl">0.0000000</span> Station_212_Date_23MAY1990<span class="sc">:</span><span class="dv">10</span><span class="sc">:</span><span class="dv">43</span><span class="sc">:</span><span class="fl">00.000</span> relative abundance                </span>
<span id="cb14-122"><a href="applications.html#cb14-122" aria-hidden="true" tabindex="-1"></a><span class="dv">334340</span>        <span class="fl">0.1224490</span> Station_212_Date_24JUL1990<span class="sc">:</span><span class="dv">09</span><span class="sc">:</span><span class="dv">34</span><span class="sc">:</span><span class="fl">00.000</span> relative abundance                </span>
<span id="cb14-123"><a href="applications.html#cb14-123" aria-hidden="true" tabindex="-1"></a><span class="dv">334341</span>        <span class="fl">0.0000000</span> Station_212_Date_21MAR2001<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">52</span><span class="sc">:</span><span class="fl">00.000</span> relative abundance                </span>
<span id="cb14-124"><a href="applications.html#cb14-124" aria-hidden="true" tabindex="-1"></a>                                              measurementTypeID measurementUnitID</span>
<span id="cb14-125"><a href="applications.html#cb14-125" aria-hidden="true" tabindex="-1"></a><span class="dv">334336</span> http<span class="sc">:</span><span class="er">//</span>vocab.nerc.ac.uk<span class="sc">/</span>collection<span class="sc">/</span>S06<span class="sc">/</span>current<span class="sc">/</span>S0600020<span class="sc">/</span>                  </span>
<span id="cb14-126"><a href="applications.html#cb14-126" aria-hidden="true" tabindex="-1"></a><span class="dv">334337</span> http<span class="sc">:</span><span class="er">//</span>vocab.nerc.ac.uk<span class="sc">/</span>collection<span class="sc">/</span>S06<span class="sc">/</span>current<span class="sc">/</span>S0600020<span class="sc">/</span>                  </span>
<span id="cb14-127"><a href="applications.html#cb14-127" aria-hidden="true" tabindex="-1"></a><span class="dv">334338</span> http<span class="sc">:</span><span class="er">//</span>vocab.nerc.ac.uk<span class="sc">/</span>collection<span class="sc">/</span>S06<span class="sc">/</span>current<span class="sc">/</span>S0600020<span class="sc">/</span>                  </span>
<span id="cb14-128"><a href="applications.html#cb14-128" aria-hidden="true" tabindex="-1"></a><span class="dv">334339</span> http<span class="sc">:</span><span class="er">//</span>vocab.nerc.ac.uk<span class="sc">/</span>collection<span class="sc">/</span>S06<span class="sc">/</span>current<span class="sc">/</span>S0600020<span class="sc">/</span>                  </span>
<span id="cb14-129"><a href="applications.html#cb14-129" aria-hidden="true" tabindex="-1"></a><span class="dv">334340</span> http<span class="sc">:</span><span class="er">//</span>vocab.nerc.ac.uk<span class="sc">/</span>collection<span class="sc">/</span>S06<span class="sc">/</span>current<span class="sc">/</span>S0600020<span class="sc">/</span>                  </span>
<span id="cb14-130"><a href="applications.html#cb14-130" aria-hidden="true" tabindex="-1"></a><span class="dv">334341</span> http<span class="sc">:</span><span class="er">//</span>vocab.nerc.ac.uk<span class="sc">/</span>collection<span class="sc">/</span>S06<span class="sc">/</span>current<span class="sc">/</span>S0600020<span class="sc">/</span>                  </span>
<span id="cb14-131"><a href="applications.html#cb14-131" aria-hidden="true" tabindex="-1"></a>                                                        occurrenceID</span>
<span id="cb14-132"><a href="applications.html#cb14-132" aria-hidden="true" tabindex="-1"></a><span class="dv">334336</span> Station_217_Date_03APR2003<span class="sc">:</span><span class="dv">13</span><span class="sc">:</span><span class="dv">28</span><span class="sc">:</span><span class="fl">00.000</span>_Litopenaeus_setiferus</span>
<span id="cb14-133"><a href="applications.html#cb14-133" aria-hidden="true" tabindex="-1"></a><span class="dv">334337</span> Station_217_Date_24FEB2006<span class="sc">:</span><span class="dv">10</span><span class="sc">:</span><span class="dv">12</span><span class="sc">:</span><span class="fl">00.000</span>_Litopenaeus_setiferus</span>
<span id="cb14-134"><a href="applications.html#cb14-134" aria-hidden="true" tabindex="-1"></a><span class="dv">334338</span> Station_217_Date_23JUN2001<span class="sc">:</span><span class="dv">12</span><span class="sc">:</span><span class="dv">28</span><span class="sc">:</span><span class="fl">00.000</span>_Litopenaeus_setiferus</span>
<span id="cb14-135"><a href="applications.html#cb14-135" aria-hidden="true" tabindex="-1"></a><span class="dv">334339</span> Station_212_Date_23MAY1990<span class="sc">:</span><span class="dv">10</span><span class="sc">:</span><span class="dv">43</span><span class="sc">:</span><span class="fl">00.000</span>_Litopenaeus_setiferus</span>
<span id="cb14-136"><a href="applications.html#cb14-136" aria-hidden="true" tabindex="-1"></a><span class="dv">334340</span> Station_212_Date_24JUL1990<span class="sc">:</span><span class="dv">09</span><span class="sc">:</span><span class="dv">34</span><span class="sc">:</span><span class="fl">00.000</span>_Litopenaeus_setiferus</span>
<span id="cb14-137"><a href="applications.html#cb14-137" aria-hidden="true" tabindex="-1"></a><span class="dv">334341</span> Station_212_Date_21MAR2001<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">52</span><span class="sc">:</span><span class="fl">00.000</span>_Litopenaeus_setiferus</span>
<span id="cb14-138"><a href="applications.html#cb14-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-139"><a href="applications.html#cb14-139" aria-hidden="true" tabindex="-1"></a><span class="co"># Write out the file</span></span>
<span id="cb14-140"><a href="applications.html#cb14-140" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(mof, <span class="at">file =</span> (<span class="fu">paste0</span>(event[<span class="dv">1</span>,]<span class="sc">$</span>datasetID, <span class="st">&quot;_mof_&quot;</span>, lubridate<span class="sc">::</span><span class="fu">today</span>(),<span class="st">&quot;.csv&quot;</span>)), <span class="at">fileEncoding =</span> <span class="st">&quot;UTF-8&quot;</span>, <span class="at">row.names =</span> F, <span class="at">na =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
</div>
<div id="cleaning-up-event-and-occurrence-files" class="section level3 hasAnchor" number="2.1.5">
<h3><span class="header-section-number">2.1.5</span> Cleaning up Event and Occurrence files<a href="applications.html#cleaning-up-event-and-occurrence-files" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now that we have all of our files created we can clean up the <strong>Event</strong> and <strong>Occurrence</strong> files to remove the columns that are not following Darwin Core. We had to leave the extra bits in before because we needed them to create the <strong>emof</strong> file above.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="applications.html#cb15-1" aria-hidden="true" tabindex="-1"></a>event <span class="ot">&lt;-</span> event[<span class="fu">c</span>(<span class="st">&quot;samplingProtocol&quot;</span>,<span class="st">&quot;locality&quot;</span>,<span class="st">&quot;waterBody&quot;</span>,<span class="st">&quot;decimalLatitude&quot;</span>,<span class="st">&quot;decimalLongitude&quot;</span>,</span>
<span id="cb15-2"><a href="applications.html#cb15-2" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;eventDate&quot;</span>,<span class="st">&quot;sampleSizeValue&quot;</span>,<span class="st">&quot;minimumDepthInMeters&quot;</span>,</span>
<span id="cb15-3"><a href="applications.html#cb15-3" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;maximumDepthInMeters&quot;</span>,<span class="st">&quot;type&quot;</span>,<span class="st">&quot;modified&quot;</span>,<span class="st">&quot;language&quot;</span>,<span class="st">&quot;license&quot;</span>,<span class="st">&quot;institutionCode&quot;</span>,</span>
<span id="cb15-4"><a href="applications.html#cb15-4" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;ownerInstitutionCode&quot;</span>,<span class="st">&quot;coordinateUncertaintyInMeters&quot;</span>,</span>
<span id="cb15-5"><a href="applications.html#cb15-5" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;geodeticDatum&quot;</span>, <span class="st">&quot;georeferenceProtocol&quot;</span>,<span class="st">&quot;country&quot;</span>,<span class="st">&quot;countryCode&quot;</span>,<span class="st">&quot;stateProvince&quot;</span>,</span>
<span id="cb15-6"><a href="applications.html#cb15-6" aria-hidden="true" tabindex="-1"></a>                 <span class="st">&quot;datasetID&quot;</span>,<span class="st">&quot;eventID&quot;</span>,<span class="st">&quot;sampleSizeUnit&quot;</span>,<span class="st">&quot;samplingEffort&quot;</span>)]</span>
<span id="cb15-7"><a href="applications.html#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(event)</span>
<span id="cb15-8"><a href="applications.html#cb15-8" aria-hidden="true" tabindex="-1"></a>  samplingProtocol                locality   waterBody decimalLatitude decimalLongitude</span>
<span id="cb15-9"><a href="applications.html#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>        Bag Seine Mission<span class="sc">-</span>Aransas Estuary Aransas Bay        <span class="fl">28.13472</span>        <span class="sc">-</span><span class="fl">97.00833</span></span>
<span id="cb15-10"><a href="applications.html#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>        Bag Seine Mission<span class="sc">-</span>Aransas Estuary Aransas Bay        <span class="fl">28.13528</span>        <span class="sc">-</span><span class="fl">97.00722</span></span>
<span id="cb15-11"><a href="applications.html#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>        Bag Seine Mission<span class="sc">-</span>Aransas Estuary Aransas Bay        <span class="fl">28.13444</span>        <span class="sc">-</span><span class="fl">96.99611</span></span>
<span id="cb15-12"><a href="applications.html#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>        Bag Seine Mission<span class="sc">-</span>Aransas Estuary Aransas Bay        <span class="fl">28.13444</span>        <span class="sc">-</span><span class="fl">96.99611</span></span>
<span id="cb15-13"><a href="applications.html#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>        Bag Seine Mission<span class="sc">-</span>Aransas Estuary Aransas Bay        <span class="fl">28.13444</span>        <span class="sc">-</span><span class="fl">96.99611</span></span>
<span id="cb15-14"><a href="applications.html#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>        Bag Seine Mission<span class="sc">-</span>Aransas Estuary Aransas Bay        <span class="fl">28.13472</span>        <span class="sc">-</span><span class="fl">96.99583</span></span>
<span id="cb15-15"><a href="applications.html#cb15-15" aria-hidden="true" tabindex="-1"></a>            eventDate sampleSizeValue minimumDepthInMeters maximumDepthInMeters  type   modified language</span>
<span id="cb15-16"><a href="applications.html#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="dv">1997-01-09</span> <span class="dv">14</span><span class="sc">:</span><span class="dv">35</span><span class="sc">:</span><span class="dv">00</span>            <span class="fl">0.03</span>                  <span class="fl">0.0</span>                  <span class="fl">0.6</span> Event <span class="dv">2022-01-09</span>       en</span>
<span id="cb15-17"><a href="applications.html#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> <span class="dv">2000-08-18</span> <span class="dv">11</span><span class="sc">:</span><span class="dv">02</span><span class="sc">:</span><span class="dv">00</span>            <span class="fl">0.03</span>                  <span class="fl">0.1</span>                  <span class="fl">0.5</span> Event <span class="dv">2022-01-09</span>       en</span>
<span id="cb15-18"><a href="applications.html#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> <span class="dv">2005-06-28</span> <span class="dv">08</span><span class="sc">:</span><span class="dv">41</span><span class="sc">:</span><span class="dv">00</span>            <span class="fl">0.03</span>                  <span class="fl">0.4</span>                  <span class="fl">0.6</span> Event <span class="dv">2022-01-09</span>       en</span>
<span id="cb15-19"><a href="applications.html#cb15-19" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> <span class="dv">2006-08-23</span> <span class="dv">11</span><span class="sc">:</span><span class="dv">47</span><span class="sc">:</span><span class="dv">00</span>            <span class="fl">0.03</span>                  <span class="fl">0.2</span>                  <span class="fl">0.4</span> Event <span class="dv">2022-01-09</span>       en</span>
<span id="cb15-20"><a href="applications.html#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> <span class="dv">2006-10-17</span> <span class="dv">14</span><span class="sc">:</span><span class="dv">23</span><span class="sc">:</span><span class="dv">00</span>            <span class="fl">0.03</span>                  <span class="fl">0.7</span>                  <span class="fl">0.8</span> Event <span class="dv">2022-01-09</span>       en</span>
<span id="cb15-21"><a href="applications.html#cb15-21" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span> <span class="dv">1996-02-19</span> <span class="dv">10</span><span class="sc">:</span><span class="dv">27</span><span class="sc">:</span><span class="dv">00</span>            <span class="fl">0.03</span>                  <span class="fl">0.1</span>                  <span class="fl">0.3</span> Event <span class="dv">2022-01-09</span>       en</span>
<span id="cb15-22"><a href="applications.html#cb15-22" aria-hidden="true" tabindex="-1"></a>                                                     license institutionCode ownerInstitutionCode</span>
<span id="cb15-23"><a href="applications.html#cb15-23" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD                 HARC</span>
<span id="cb15-24"><a href="applications.html#cb15-24" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD                 HARC</span>
<span id="cb15-25"><a href="applications.html#cb15-25" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD                 HARC</span>
<span id="cb15-26"><a href="applications.html#cb15-26" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD                 HARC</span>
<span id="cb15-27"><a href="applications.html#cb15-27" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD                 HARC</span>
<span id="cb15-28"><a href="applications.html#cb15-28" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span> http<span class="sc">:</span><span class="er">//</span>creativecommons.org<span class="sc">/</span>publicdomain<span class="sc">/</span>zero<span class="sc">/</span><span class="fl">1.0</span><span class="sc">/</span>legalcode            TPWD                 HARC</span>
<span id="cb15-29"><a href="applications.html#cb15-29" aria-hidden="true" tabindex="-1"></a>  coordinateUncertaintyInMeters geodeticDatum georeferenceProtocol       country countryCode stateProvince</span>
<span id="cb15-30"><a href="applications.html#cb15-30" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>                           <span class="dv">100</span>         WGS84         Handheld GPS United States          US         Texas</span>
<span id="cb15-31"><a href="applications.html#cb15-31" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>                           <span class="dv">100</span>         WGS84         Handheld GPS United States          US         Texas</span>
<span id="cb15-32"><a href="applications.html#cb15-32" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>                           <span class="dv">100</span>         WGS84         Handheld GPS United States          US         Texas</span>
<span id="cb15-33"><a href="applications.html#cb15-33" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>                           <span class="dv">100</span>         WGS84         Handheld GPS United States          US         Texas</span>
<span id="cb15-34"><a href="applications.html#cb15-34" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>                           <span class="dv">100</span>         WGS84         Handheld GPS United States          US         Texas</span>
<span id="cb15-35"><a href="applications.html#cb15-35" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>                           <span class="dv">100</span>         WGS84         Handheld GPS United States          US         Texas</span>
<span id="cb15-36"><a href="applications.html#cb15-36" aria-hidden="true" tabindex="-1"></a>                              datasetID                                eventID sampleSizeUnit</span>
<span id="cb15-37"><a href="applications.html#cb15-37" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_95_Date_09JAN1997<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">35</span><span class="sc">:</span><span class="fl">00.000</span>       hectares</span>
<span id="cb15-38"><a href="applications.html#cb15-38" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_95_Date_18AUG2000<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">02</span><span class="sc">:</span><span class="fl">00.000</span>       hectares</span>
<span id="cb15-39"><a href="applications.html#cb15-39" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_28JUN2005<span class="sc">:</span><span class="dv">08</span><span class="sc">:</span><span class="dv">41</span><span class="sc">:</span><span class="fl">00.000</span>       hectares</span>
<span id="cb15-40"><a href="applications.html#cb15-40" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_23AUG2006<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">47</span><span class="sc">:</span><span class="fl">00.000</span>       hectares</span>
<span id="cb15-41"><a href="applications.html#cb15-41" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_17OCT2006<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">23</span><span class="sc">:</span><span class="fl">00.000</span>       hectares</span>
<span id="cb15-42"><a href="applications.html#cb15-42" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span> TPWD_HARC_Texas_Aransas_Bay_Bag_Seine Station_96_Date_19FEB1996<span class="sc">:</span><span class="dv">10</span><span class="sc">:</span><span class="dv">27</span><span class="sc">:</span><span class="fl">00.000</span>       hectares</span>
<span id="cb15-43"><a href="applications.html#cb15-43" aria-hidden="true" tabindex="-1"></a>  samplingEffort</span>
<span id="cb15-44"><a href="applications.html#cb15-44" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>    <span class="dv">120</span> seconds</span>
<span id="cb15-45"><a href="applications.html#cb15-45" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>    <span class="dv">120</span> seconds</span>
<span id="cb15-46"><a href="applications.html#cb15-46" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>    <span class="dv">120</span> seconds</span>
<span id="cb15-47"><a href="applications.html#cb15-47" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>    <span class="dv">120</span> seconds</span>
<span id="cb15-48"><a href="applications.html#cb15-48" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>    <span class="dv">120</span> seconds</span>
<span id="cb15-49"><a href="applications.html#cb15-49" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>    <span class="dv">120</span> seconds</span>
<span id="cb15-50"><a href="applications.html#cb15-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-51"><a href="applications.html#cb15-51" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(event, <span class="at">file =</span> <span class="fu">paste0</span>(event[<span class="dv">1</span>,]<span class="sc">$</span>datasetID, <span class="st">&quot;_event_&quot;</span>, lubridate<span class="sc">::</span><span class="fu">today</span>(),<span class="st">&quot;.csv&quot;</span>), <span class="at">fileEncoding =</span> <span class="st">&quot;UTF-8&quot;</span>, <span class="at">row.names =</span> F, <span class="at">na =</span> <span class="st">&quot;&quot;</span>)                    </span>
<span id="cb15-52"><a href="applications.html#cb15-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-53"><a href="applications.html#cb15-53" aria-hidden="true" tabindex="-1"></a>occurrence <span class="ot">&lt;-</span> occurrence[<span class="fu">c</span>(<span class="st">&quot;vernacularName&quot;</span>,<span class="st">&quot;eventID&quot;</span>,<span class="st">&quot;occurrenceStatus&quot;</span>,<span class="st">&quot;basisOfRecord&quot;</span>,</span>
<span id="cb15-54"><a href="applications.html#cb15-54" aria-hidden="true" tabindex="-1"></a>                           <span class="st">&quot;scientificName&quot;</span>,<span class="st">&quot;scientificNameID&quot;</span>,<span class="st">&quot;kingdom&quot;</span>,<span class="st">&quot;phylum&quot;</span>,<span class="st">&quot;class&quot;</span>,</span>
<span id="cb15-55"><a href="applications.html#cb15-55" aria-hidden="true" tabindex="-1"></a>                           <span class="st">&quot;order&quot;</span>,<span class="st">&quot;family&quot;</span>,<span class="st">&quot;genus&quot;</span>,</span>
<span id="cb15-56"><a href="applications.html#cb15-56" aria-hidden="true" tabindex="-1"></a>                           <span class="st">&quot;scientificNameAuthorship&quot;</span>,<span class="st">&quot;taxonRank&quot;</span>, <span class="st">&quot;organismQuantity&quot;</span>,</span>
<span id="cb15-57"><a href="applications.html#cb15-57" aria-hidden="true" tabindex="-1"></a>                           <span class="st">&quot;organismQuantityType&quot;</span>, <span class="st">&quot;occurrenceID&quot;</span>,<span class="st">&quot;collectionCode&quot;</span>)]</span>
<span id="cb15-58"><a href="applications.html#cb15-58" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(occurrence)</span>
<span id="cb15-59"><a href="applications.html#cb15-59" aria-hidden="true" tabindex="-1"></a>  vernacularName                                eventID occurrenceStatus    basisOfRecord</span>
<span id="cb15-60"><a href="applications.html#cb15-60" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>  Alligator gar Station_95_Date_09JAN1997<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">35</span><span class="sc">:</span><span class="fl">00.000</span>           Absent HumanObservation</span>
<span id="cb15-61"><a href="applications.html#cb15-61" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>  Alligator gar Station_95_Date_18AUG2000<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">02</span><span class="sc">:</span><span class="fl">00.000</span>           Absent HumanObservation</span>
<span id="cb15-62"><a href="applications.html#cb15-62" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>  Alligator gar Station_96_Date_28JUN2005<span class="sc">:</span><span class="dv">08</span><span class="sc">:</span><span class="dv">41</span><span class="sc">:</span><span class="fl">00.000</span>           Absent HumanObservation</span>
<span id="cb15-63"><a href="applications.html#cb15-63" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>  Alligator gar Station_96_Date_23AUG2006<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">47</span><span class="sc">:</span><span class="fl">00.000</span>           Absent HumanObservation</span>
<span id="cb15-64"><a href="applications.html#cb15-64" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>  Alligator gar Station_96_Date_17OCT2006<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">23</span><span class="sc">:</span><span class="fl">00.000</span>           Absent HumanObservation</span>
<span id="cb15-65"><a href="applications.html#cb15-65" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>  Alligator gar Station_96_Date_19FEB1996<span class="sc">:</span><span class="dv">10</span><span class="sc">:</span><span class="dv">27</span><span class="sc">:</span><span class="fl">00.000</span>           Absent HumanObservation</span>
<span id="cb15-66"><a href="applications.html#cb15-66" aria-hidden="true" tabindex="-1"></a>        scientificName                          scientificNameID  kingdom   phylum       class</span>
<span id="cb15-67"><a href="applications.html#cb15-67" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> Atractosteus spatula urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">279822</span> Animalia Chordata Actinopteri</span>
<span id="cb15-68"><a href="applications.html#cb15-68" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> Atractosteus spatula urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">279822</span> Animalia Chordata Actinopteri</span>
<span id="cb15-69"><a href="applications.html#cb15-69" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> Atractosteus spatula urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">279822</span> Animalia Chordata Actinopteri</span>
<span id="cb15-70"><a href="applications.html#cb15-70" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> Atractosteus spatula urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">279822</span> Animalia Chordata Actinopteri</span>
<span id="cb15-71"><a href="applications.html#cb15-71" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> Atractosteus spatula urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">279822</span> Animalia Chordata Actinopteri</span>
<span id="cb15-72"><a href="applications.html#cb15-72" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span> Atractosteus spatula urn<span class="sc">:</span>lsid<span class="sc">:</span>marinespecies.org<span class="sc">:</span>taxname<span class="sc">:</span><span class="dv">279822</span> Animalia Chordata Actinopteri</span>
<span id="cb15-73"><a href="applications.html#cb15-73" aria-hidden="true" tabindex="-1"></a>             order        family        genus scientificNameAuthorship taxonRank organismQuantity</span>
<span id="cb15-74"><a href="applications.html#cb15-74" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> Lepisosteiformes Lepisosteidae <span class="fu">Atractosteus</span>         (Lacepède, <span class="dv">1803</span>)   Species                <span class="dv">0</span></span>
<span id="cb15-75"><a href="applications.html#cb15-75" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> Lepisosteiformes Lepisosteidae <span class="fu">Atractosteus</span>         (Lacepède, <span class="dv">1803</span>)   Species                <span class="dv">0</span></span>
<span id="cb15-76"><a href="applications.html#cb15-76" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> Lepisosteiformes Lepisosteidae <span class="fu">Atractosteus</span>         (Lacepède, <span class="dv">1803</span>)   Species                <span class="dv">0</span></span>
<span id="cb15-77"><a href="applications.html#cb15-77" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span> Lepisosteiformes Lepisosteidae <span class="fu">Atractosteus</span>         (Lacepède, <span class="dv">1803</span>)   Species                <span class="dv">0</span></span>
<span id="cb15-78"><a href="applications.html#cb15-78" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> Lepisosteiformes Lepisosteidae <span class="fu">Atractosteus</span>         (Lacepède, <span class="dv">1803</span>)   Species                <span class="dv">0</span></span>
<span id="cb15-79"><a href="applications.html#cb15-79" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span> Lepisosteiformes Lepisosteidae <span class="fu">Atractosteus</span>         (Lacepède, <span class="dv">1803</span>)   Species                <span class="dv">0</span></span>
<span id="cb15-80"><a href="applications.html#cb15-80" aria-hidden="true" tabindex="-1"></a>  organismQuantityType                                                occurrenceID        collectionCode</span>
<span id="cb15-81"><a href="applications.html#cb15-81" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>   Relative Abundance Station_95_Date_09JAN1997<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">35</span><span class="sc">:</span><span class="fl">00.000</span>_Atractosteus_spatula Aransas Bay Bag Seine</span>
<span id="cb15-82"><a href="applications.html#cb15-82" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>   Relative Abundance Station_95_Date_18AUG2000<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">02</span><span class="sc">:</span><span class="fl">00.000</span>_Atractosteus_spatula Aransas Bay Bag Seine</span>
<span id="cb15-83"><a href="applications.html#cb15-83" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>   Relative Abundance Station_96_Date_28JUN2005<span class="sc">:</span><span class="dv">08</span><span class="sc">:</span><span class="dv">41</span><span class="sc">:</span><span class="fl">00.000</span>_Atractosteus_spatula Aransas Bay Bag Seine</span>
<span id="cb15-84"><a href="applications.html#cb15-84" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>   Relative Abundance Station_96_Date_23AUG2006<span class="sc">:</span><span class="dv">11</span><span class="sc">:</span><span class="dv">47</span><span class="sc">:</span><span class="fl">00.000</span>_Atractosteus_spatula Aransas Bay Bag Seine</span>
<span id="cb15-85"><a href="applications.html#cb15-85" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>   Relative Abundance Station_96_Date_17OCT2006<span class="sc">:</span><span class="dv">14</span><span class="sc">:</span><span class="dv">23</span><span class="sc">:</span><span class="fl">00.000</span>_Atractosteus_spatula Aransas Bay Bag Seine</span>
<span id="cb15-86"><a href="applications.html#cb15-86" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>   Relative Abundance Station_96_Date_19FEB1996<span class="sc">:</span><span class="dv">10</span><span class="sc">:</span><span class="dv">27</span><span class="sc">:</span><span class="fl">00.000</span>_Atractosteus_spatula Aransas Bay Bag Seine</span>
<span id="cb15-87"><a href="applications.html#cb15-87" aria-hidden="true" tabindex="-1"></a>                           </span>
<span id="cb15-88"><a href="applications.html#cb15-88" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(occurrence, <span class="at">file =</span> <span class="fu">paste0</span>(event[<span class="dv">1</span>,]<span class="sc">$</span>datasetID, <span class="st">&quot;_occurrence_&quot;</span>,lubridate<span class="sc">::</span><span class="fu">today</span>(),<span class="st">&quot;.csv&quot;</span>), <span class="at">fileEncoding =</span> <span class="st">&quot;UTF-8&quot;</span>, <span class="at">row.names =</span> F, <span class="at">na =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="salmon-ocean-ecology-data" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Salmon Ocean Ecology Data<a href="applications.html#salmon-ocean-ecology-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="intro-1" class="section level3 hasAnchor" number="2.2.1">
<h3><span class="header-section-number">2.2.1</span> Intro<a href="applications.html#intro-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>One of the goals of the Hakai Institute and the Canadian Integrated Ocean Observing System (CIOOS) is to facilitate Open Science and FAIR (findable, accessible, interoperable, reusable) ecological and oceanographic data. In a concerted effort to adopt or establish how best to do that, several Hakai and CIOOS staff attended an International Ocean Observing System (IOOS) Code Sprint in Ann Arbour, Michigan between October 7–11, 2019, to discuss how to implement FAIR data principles for biological data collected in the marine environment.</p>
<p>The <a href="https://dwc.tdwg.org">Darwin Core</a> is a highly structured data format that standardizes data table relations, vocabularies, and defines field names. The Darwin Core defines three table types: <code>event</code>, <code>occurrence</code>, and <code>measurementOrFact</code>. This intuitively captures the way most ecologists conduct their research. Typically, a survey (event) is conducted and measurements, counts, or observations (collectively measurementOrFacts) are made regarding a specific habitat or species (occurrence).</p>
<p>In the following script I demonstrate how I go about converting a subset of the data collected from the Hakai Institute Juvenile Salmon Program and discuss challenges, solutions, pros and cons, and when and what’s worthwhile to convert to Darwin Core.</p>
<p>The conversion of a dataset to Darwin Core is much easier if your data are already tidy (normalized) in which you represent your data in separate tables that reflect the hierarchical and related nature of your observations. If your data are not already in a consistent and structured format, the conversion would likely be very arduos and not intuitive.</p>
</div>
<div id="event" class="section level3 hasAnchor" number="2.2.2">
<h3><span class="header-section-number">2.2.2</span> event<a href="applications.html#event" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The first step is to consider what you will define as an event in your data set. I defined the capture of fish using a purse seine net as the <code>event</code>. Therefore, each row in the <code>event</code> table is one deployment of a seine net and is assigned a unique <code>eventID</code>.</p>
<p>My process for conversion was to make a new table called <code>event</code> and map the standard Darwin Core column names to pre-existing columns that serve the same purpose in my original <code>seine_data</code> table and populate the other required fields.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="applications.html#cb16-1" aria-hidden="true" tabindex="-1"></a>event <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">eventID =</span> survey_seines<span class="sc">$</span>seine_id,</span>
<span id="cb16-2"><a href="applications.html#cb16-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">eventDate =</span> <span class="fu">date</span>(survey_seines<span class="sc">$</span>survey_date),</span>
<span id="cb16-3"><a href="applications.html#cb16-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">decimalLatitude =</span> survey_seines<span class="sc">$</span>lat,</span>
<span id="cb16-4"><a href="applications.html#cb16-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">decimalLongitude =</span> survey_seines<span class="sc">$</span>long,</span>
<span id="cb16-5"><a href="applications.html#cb16-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">geodeticDatum =</span> <span class="st">&quot;EPSG:4326 WGS84&quot;</span>,</span>
<span id="cb16-6"><a href="applications.html#cb16-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">minimumDepthInMeters =</span> <span class="dv">0</span>,</span>
<span id="cb16-7"><a href="applications.html#cb16-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">maximumDepthInMeters =</span> <span class="dv">9</span>, <span class="co"># seine depth is 9 m</span></span>
<span id="cb16-8"><a href="applications.html#cb16-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">samplingProtocol =</span> <span class="st">&quot;http://dx.doi.org/10.21966/1.566666&quot;</span> <span class="co"># This is the DOI for the Hakai Salmon Data Package that contains the smnpling protocol, as well as the complete data package</span></span>
<span id="cb16-9"><a href="applications.html#cb16-9" aria-hidden="true" tabindex="-1"></a>               ) </span>
<span id="cb16-10"><a href="applications.html#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="applications.html#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(event, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;datasets&quot;</span>, <span class="st">&quot;hakai_salmon_data&quot;</span>, <span class="st">&quot;event.csv&quot;</span>))</span></code></pre></div>
</div>
<div id="occurrence" class="section level3 hasAnchor" number="2.2.3">
<h3><span class="header-section-number">2.2.3</span> occurrence<a href="applications.html#occurrence" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Next you’ll want to determine what constitutes an occurrence for your data set. Because each event caputers fish, I consider each fish to be an occurrence. Therefore, the unit of observation (each row) in the occurrence table is a fish. To link each occurence to an event you need to include the <code>eventID</code> column for every occurrence so that you know what seine (event) each fish (occurrence) came from. You must also provide a globally unique identifier for each occurrence. I already have a locally unique identifier for each fish in the original <code>fish_data</code> table called <code>ufn</code>. To make it globally unique I pre-pend the organization and research program metadata to the <code>ufn</code> column.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="applications.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">#TODO: Include bycatch data as well</span></span>
<span id="cb17-2"><a href="applications.html#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="applications.html#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="do">## make table long first</span></span>
<span id="cb17-4"><a href="applications.html#cb17-4" aria-hidden="true" tabindex="-1"></a>seines_total_long <span class="ot">&lt;-</span> survey_seines <span class="sc">%&gt;%</span> </span>
<span id="cb17-5"><a href="applications.html#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(seine_id, so_total, pi_total, cu_total, co_total, he_total, ck_total) <span class="sc">%&gt;%</span> </span>
<span id="cb17-6"><a href="applications.html#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>seine_id, <span class="at">names_to =</span> <span class="st">&quot;scientificName&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;n&quot;</span>)</span>
<span id="cb17-7"><a href="applications.html#cb17-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-8"><a href="applications.html#cb17-8" aria-hidden="true" tabindex="-1"></a>seines_total_long<span class="sc">$</span>scientificName <span class="ot">&lt;-</span> <span class="fu">recode</span>(seines_total_long<span class="sc">$</span>scientificName, <span class="at">so_total =</span> <span class="st">&quot;Oncorhynchus nerka&quot;</span>, <span class="at">pi_total =</span> <span class="st">&quot;Oncorhynchus gorbushca&quot;</span>, <span class="at">cu_total =</span> <span class="st">&quot;Oncorhynchus keta&quot;</span>, <span class="at">co_total =</span> <span class="st">&quot;Oncorhynchus kisutch&quot;</span>, <span class="at">ck_total =</span> <span class="st">&quot;Oncorhynchus tshawytscha&quot;</span>, <span class="at">he_total =</span> <span class="st">&quot;Clupea pallasii&quot;</span>) </span>
<span id="cb17-9"><a href="applications.html#cb17-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-10"><a href="applications.html#cb17-10" aria-hidden="true" tabindex="-1"></a>seines_taken_long <span class="ot">&lt;-</span> survey_seines <span class="sc">%&gt;%</span></span>
<span id="cb17-11"><a href="applications.html#cb17-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(seine_id, so_taken, pi_taken, cu_taken, co_taken, he_taken, ck_taken) <span class="sc">%&gt;%</span> </span>
<span id="cb17-12"><a href="applications.html#cb17-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>seine_id, <span class="at">names_to =</span> <span class="st">&quot;scientificName&quot;</span>, <span class="at">values_to =</span> <span class="st">&quot;n_taken&quot;</span>) </span>
<span id="cb17-13"><a href="applications.html#cb17-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-14"><a href="applications.html#cb17-14" aria-hidden="true" tabindex="-1"></a>seines_taken_long<span class="sc">$</span>scientificName <span class="ot">&lt;-</span> <span class="fu">recode</span>(seines_taken_long<span class="sc">$</span>scientificName, <span class="at">so_taken =</span> <span class="st">&quot;Oncorhynchus nerka&quot;</span>, <span class="at">pi_taken =</span> <span class="st">&quot;Oncorhynchus gorbushca&quot;</span>, <span class="at">cu_taken =</span> <span class="st">&quot;Oncorhynchus keta&quot;</span>, <span class="at">co_taken =</span> <span class="st">&quot;Oncorhynchus kisutch&quot;</span>, <span class="at">ck_taken =</span> <span class="st">&quot;Oncorhynchus tshawytscha&quot;</span>, <span class="at">he_taken =</span> <span class="st">&quot;Clupea pallasii&quot;</span>) </span>
<span id="cb17-15"><a href="applications.html#cb17-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-16"><a href="applications.html#cb17-16" aria-hidden="true" tabindex="-1"></a><span class="do">## remove records that have already been assigned an ID  </span></span>
<span id="cb17-17"><a href="applications.html#cb17-17" aria-hidden="true" tabindex="-1"></a>seines_long <span class="ot">&lt;-</span>  <span class="fu">full_join</span>(seines_total_long, seines_taken_long, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;seine_id&quot;</span>, <span class="st">&quot;scientificName&quot;</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb17-18"><a href="applications.html#cb17-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb17-19"><a href="applications.html#cb17-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">n_not_taken =</span> n <span class="sc">-</span> n_taken) <span class="sc">%&gt;%</span> <span class="co">#so_total includes the number taken so I subtract n_taken to get n_not_taken</span></span>
<span id="cb17-20"><a href="applications.html#cb17-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>n_taken, <span class="sc">-</span>n) <span class="sc">%&gt;%</span> </span>
<span id="cb17-21"><a href="applications.html#cb17-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(n_not_taken <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb17-22"><a href="applications.html#cb17-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-23"><a href="applications.html#cb17-23" aria-hidden="true" tabindex="-1"></a>all_fish_caught <span class="ot">&lt;-</span></span>
<span id="cb17-24"><a href="applications.html#cb17-24" aria-hidden="true" tabindex="-1"></a>  seines_long[<span class="fu">rep</span>(<span class="fu">seq.int</span>(<span class="dv">1</span>, <span class="fu">nrow</span>(seines_long)), seines_long<span class="sc">$</span>n_not_taken), <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>] <span class="sc">%&gt;%</span> </span>
<span id="cb17-25"><a href="applications.html#cb17-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>n_not_taken) <span class="sc">%&gt;%</span> </span>
<span id="cb17-26"><a href="applications.html#cb17-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">prefix =</span> <span class="st">&quot;hakai-jsp-&quot;</span>,</span>
<span id="cb17-27"><a href="applications.html#cb17-27" aria-hidden="true" tabindex="-1"></a>         <span class="at">suffix =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(.),</span>
<span id="cb17-28"><a href="applications.html#cb17-28" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="fu">paste0</span>(prefix, suffix)</span>
<span id="cb17-29"><a href="applications.html#cb17-29" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb17-30"><a href="applications.html#cb17-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>prefix, <span class="sc">-</span>suffix)</span>
<span id="cb17-31"><a href="applications.html#cb17-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-32"><a href="applications.html#cb17-32" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb17-33"><a href="applications.html#cb17-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-34"><a href="applications.html#cb17-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Change species names to full Scientific names </span></span>
<span id="cb17-35"><a href="applications.html#cb17-35" aria-hidden="true" tabindex="-1"></a>latin <span class="ot">&lt;-</span> <span class="fu">fct_recode</span>(fish_data<span class="sc">$</span>species, <span class="st">&quot;Oncorhynchus nerka&quot;</span> <span class="ot">=</span> <span class="st">&quot;SO&quot;</span>, <span class="st">&quot;Oncorhynchus gorbuscha&quot;</span> <span class="ot">=</span> <span class="st">&quot;PI&quot;</span>, <span class="st">&quot;Oncorhynchus keta&quot;</span> <span class="ot">=</span> <span class="st">&quot;CU&quot;</span>, <span class="st">&quot;Oncorhynchus kisutch&quot;</span> <span class="ot">=</span> <span class="st">&quot;CO&quot;</span>, <span class="st">&quot;Clupea pallasii&quot;</span> <span class="ot">=</span> <span class="st">&quot;HE&quot;</span>, <span class="st">&quot;Oncorhynchus tshawytscha&quot;</span> <span class="ot">=</span> <span class="st">&quot;CK&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb17-36"><a href="applications.html#cb17-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.character</span>()</span>
<span id="cb17-37"><a href="applications.html#cb17-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-38"><a href="applications.html#cb17-38" aria-hidden="true" tabindex="-1"></a>fish_retained_data <span class="ot">&lt;-</span> fish_data <span class="sc">%&gt;%</span> </span>
<span id="cb17-39"><a href="applications.html#cb17-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">scientificName =</span> latin) <span class="sc">%&gt;%</span> </span>
<span id="cb17-40"><a href="applications.html#cb17-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>species) <span class="sc">%&gt;%</span> </span>
<span id="cb17-41"><a href="applications.html#cb17-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">prefix =</span> <span class="st">&quot;hakai-jsp-&quot;</span>,</span>
<span id="cb17-42"><a href="applications.html#cb17-42" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="fu">paste0</span>(prefix, ufn)) <span class="sc">%&gt;%</span> </span>
<span id="cb17-43"><a href="applications.html#cb17-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>semsp_id, <span class="sc">-</span>prefix, <span class="sc">-</span>ufn, <span class="sc">-</span>fork_length_field, <span class="sc">-</span>fork_length, <span class="sc">-</span>weight, <span class="sc">-</span>weight_field)</span>
<span id="cb17-44"><a href="applications.html#cb17-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-45"><a href="applications.html#cb17-45" aria-hidden="true" tabindex="-1"></a>occurrence <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(all_fish_caught, fish_retained_data) <span class="sc">%&gt;%</span> </span>
<span id="cb17-46"><a href="applications.html#cb17-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">basisOfRecord =</span> <span class="st">&quot;HumanObservation&quot;</span>,</span>
<span id="cb17-47"><a href="applications.html#cb17-47" aria-hidden="true" tabindex="-1"></a>        <span class="at">occurenceStatus =</span> <span class="st">&quot;present&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb17-48"><a href="applications.html#cb17-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">eventID =</span> seine_id)</span></code></pre></div>
<p>For each occuerence of the six different fish species that I caught I need to match the species name that I provide with the official <code>scientificName</code> that is part of the World Register of Marine Species database <a href="http://www.marinespecies.org/" class="uri">http://www.marinespecies.org/</a></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="applications.html#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># I went directly to the WoRMS webite (http://www.marinespecies.org/) to download the full taxonomic levels for the salmon species I have and put the WoRMS output (species_matched.xls) table in this project directory which is read in below and joined with the occurrence table</span></span>
<span id="cb18-2"><a href="applications.html#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="applications.html#cb18-3" aria-hidden="true" tabindex="-1"></a>species_matched <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;datasets&quot;</span>, <span class="st">&quot;hakai_salmon_data&quot;</span>, <span class="st">&quot;raw_data&quot;</span>, <span class="st">&quot;species_matched.xls&quot;</span>))</span>
<span id="cb18-4"><a href="applications.html#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="applications.html#cb18-5" aria-hidden="true" tabindex="-1"></a>occurrence <span class="ot">&lt;-</span> <span class="fu">left_join</span>(occurrence, species_matched, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;scientificName&quot;</span> <span class="ot">=</span> <span class="st">&quot;ScientificName&quot;</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb18-6"><a href="applications.html#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(occurrenceID, basisOfRecord, scientificName, eventID, <span class="at">occurrenceStatus =</span> occurenceStatus, Kingdom, Phylum, Class, Order, Family, Genus, Species)</span>
<span id="cb18-7"><a href="applications.html#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="applications.html#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(occurrence, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;datasets&quot;</span>, <span class="st">&quot;hakai_salmon_data&quot;</span>, <span class="st">&quot;occurrence.csv&quot;</span>))</span></code></pre></div>
</div>
<div id="measurementorfact" class="section level3 hasAnchor" number="2.2.4">
<h3><span class="header-section-number">2.2.4</span> measurementOrFact<a href="applications.html#measurementorfact" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To convert all your measurements or facts from your normal format to Darwin Core you essentially need to put all your measurements into one column called measurementType and a corresponding column called MeasurementValue. This standardizes the column names are in the <code>measurementOrFact</code> table. There are a number of predefined <code>measurementType</code>s listed on the <a href="https://www.bodc.ac.uk/resources/vocabularies/">NERC</a> database that should be used where possible. I found it difficult to navigate this page to find the correct <code>measurementType</code>.</p>
<p>Here I convert length, and weight measurements that relate to an event and an occurrence and call those <code>measurementTypes</code> as <code>length</code> and <code>weight</code>.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="applications.html#cb19-1" aria-hidden="true" tabindex="-1"></a>fish_data<span class="sc">$</span>weight <span class="ot">&lt;-</span> <span class="fu">coalesce</span>(fish_data<span class="sc">$</span>weight, fish_data<span class="sc">$</span>weight_field)</span>
<span id="cb19-2"><a href="applications.html#cb19-2" aria-hidden="true" tabindex="-1"></a>fish_data<span class="sc">$</span>fork_length <span class="ot">&lt;-</span> <span class="fu">coalesce</span>(fish_data<span class="sc">$</span>fork_length, fish_data<span class="sc">$</span>fork_length_field)</span>
<span id="cb19-3"><a href="applications.html#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="applications.html#cb19-4" aria-hidden="true" tabindex="-1"></a>fish_length <span class="ot">&lt;-</span> fish_data <span class="sc">%&gt;%</span></span>
<span id="cb19-5"><a href="applications.html#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">occurrenceID =</span> <span class="fu">paste0</span>(<span class="st">&quot;hakai-jsp-&quot;</span>, ufn)) <span class="sc">%&gt;%</span> </span>
<span id="cb19-6"><a href="applications.html#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(occurrenceID, <span class="at">eventID =</span> seine_id, fork_length, weight) <span class="sc">%&gt;%</span> </span>
<span id="cb19-7"><a href="applications.html#cb19-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;fork length&quot;</span>, <span class="at">measurementValue =</span> fork_length) <span class="sc">%&gt;%</span> </span>
<span id="cb19-8"><a href="applications.html#cb19-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(eventID, occurrenceID, measurementType, measurementValue) <span class="sc">%&gt;%</span> </span>
<span id="cb19-9"><a href="applications.html#cb19-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementUnit =</span> <span class="st">&quot;millimeters&quot;</span>,</span>
<span id="cb19-10"><a href="applications.html#cb19-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P06/current/UXMM/&quot;</span>)</span>
<span id="cb19-11"><a href="applications.html#cb19-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-12"><a href="applications.html#cb19-12" aria-hidden="true" tabindex="-1"></a>fish_weight <span class="ot">&lt;-</span> fish_data <span class="sc">%&gt;%</span> </span>
<span id="cb19-13"><a href="applications.html#cb19-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">occurrenceID =</span> <span class="fu">paste0</span>(<span class="st">&quot;hakai-jsp-&quot;</span>, ufn)) <span class="sc">%&gt;%</span> </span>
<span id="cb19-14"><a href="applications.html#cb19-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(occurrenceID, <span class="at">eventID =</span> seine_id, fork_length, weight) <span class="sc">%&gt;%</span> </span>
<span id="cb19-15"><a href="applications.html#cb19-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementType =</span> <span class="st">&quot;mass&quot;</span>, <span class="at">measurementValue =</span> weight) <span class="sc">%&gt;%</span> </span>
<span id="cb19-16"><a href="applications.html#cb19-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(eventID, occurrenceID, measurementType, measurementValue) <span class="sc">%&gt;%</span> </span>
<span id="cb19-17"><a href="applications.html#cb19-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">measurementUnit =</span> <span class="st">&quot;grams&quot;</span>,</span>
<span id="cb19-18"><a href="applications.html#cb19-18" aria-hidden="true" tabindex="-1"></a>         <span class="at">measurementUnitID =</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P06/current/UGRM/&quot;</span>)</span>
<span id="cb19-19"><a href="applications.html#cb19-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-20"><a href="applications.html#cb19-20" aria-hidden="true" tabindex="-1"></a>measurementOrFact <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(fish_length, fish_weight) <span class="sc">%&gt;%</span> </span>
<span id="cb19-21"><a href="applications.html#cb19-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>(measurementValue)</span>
<span id="cb19-22"><a href="applications.html#cb19-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-23"><a href="applications.html#cb19-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(fish_length, fish_weight)</span>
<span id="cb19-24"><a href="applications.html#cb19-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-25"><a href="applications.html#cb19-25" aria-hidden="true" tabindex="-1"></a><span class="fu">write_csv</span>(measurementOrFact, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;datasets&quot;</span>, <span class="st">&quot;hakai_salmon_data&quot;</span>, <span class="st">&quot;measurementOrFact.csv&quot;</span>))</span></code></pre></div>
</div>
</div>
<div id="hakai-seagrass" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Hakai Seagrass<a href="applications.html#hakai-seagrass" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="setup" class="section level3 hasAnchor" number="2.3.1">
<h3><span class="header-section-number">2.3.1</span> Setup<a href="applications.html#setup" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This section clears the workspace, checks the working directory, and
installs packages (if required) and loads packages, and loads necessary
datasets</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="applications.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;knitr&quot;</span>)</span>
<span id="cb20-2"><a href="applications.html#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Knitr global chunk options</span></span>
<span id="cb20-3"><a href="applications.html#cb20-3" aria-hidden="true" tabindex="-1"></a>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">message =</span> <span class="cn">FALSE</span>,</span>
<span id="cb20-4"><a href="applications.html#cb20-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">warning =</span> <span class="cn">FALSE</span>,</span>
<span id="cb20-5"><a href="applications.html#cb20-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">error   =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<div id="load-data" class="section level4 hasAnchor" number="2.3.1.1">
<h4><span class="header-section-number">2.3.1.1</span> Load Data<a href="applications.html#load-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>First load the seagrass density survey data, set variable classes, and have a quick look</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="applications.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load density data</span></span>
<span id="cb21-2"><a href="applications.html#cb21-2" aria-hidden="true" tabindex="-1"></a>seagrassDensity <span class="ot">&lt;-</span> </span>
<span id="cb21-3"><a href="applications.html#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read.csv</span>(<span class="st">&quot;raw_data/seagrass_density_survey.csv&quot;</span>,</span>
<span id="cb21-4"><a href="applications.html#cb21-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">colClass =</span> <span class="st">&quot;character&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb21-5"><a href="applications.html#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">date             =</span> <span class="fu">ymd</span>(date),</span>
<span id="cb21-6"><a href="applications.html#cb21-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">depth            =</span> <span class="fu">as.numeric</span>(depth),</span>
<span id="cb21-7"><a href="applications.html#cb21-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">transect_dist    =</span> <span class="fu">factor</span>(transect_dist),</span>
<span id="cb21-8"><a href="applications.html#cb21-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">collected_start  =</span> <span class="fu">ymd_hms</span>(collected_start),</span>
<span id="cb21-9"><a href="applications.html#cb21-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">collected_end    =</span> <span class="fu">ymd_hms</span>(collected_end),</span>
<span id="cb21-10"><a href="applications.html#cb21-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">density          =</span> <span class="fu">as.numeric</span>(density),</span>
<span id="cb21-11"><a href="applications.html#cb21-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">density_msq      =</span> <span class="fu">as.numeric</span>(density_msq),</span>
<span id="cb21-12"><a href="applications.html#cb21-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">canopy_height_cm =</span> <span class="fu">as.numeric</span>(canopy_height_cm),</span>
<span id="cb21-13"><a href="applications.html#cb21-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">flowering_shoots =</span> <span class="fu">as.numeric</span>(flowering_shoots)) <span class="sc">%T&gt;%</span></span>
<span id="cb21-14"><a href="applications.html#cb21-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glimpse</span>()</span></code></pre></div>
<pre><code>## Rows: 3,031
## Columns: 22
## $ X                &lt;chr&gt; &quot;1&quot;, &quot;2&quot;, &quot;3&quot;, &quot;4&quot;, &quot;5&quot;, &quot;6&quot;, &quot;7&quot;, &quot;8&quot;, &quot;9&quot;, &quot;10&quot;, &quot;1…
## $ organization     &lt;chr&gt; &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;,…
## $ work_area        &lt;chr&gt; &quot;CALVERT&quot;, &quot;CALVERT&quot;, &quot;CALVERT&quot;, &quot;CALVERT&quot;, &quot;CALVERT&quot;…
## $ project          &lt;chr&gt; &quot;MARINEGEO&quot;, &quot;MARINEGEO&quot;, &quot;MARINEGEO&quot;, &quot;MARINEGEO&quot;, &quot;…
## $ survey           &lt;chr&gt; &quot;PRUTH_BAY&quot;, &quot;PRUTH_BAY&quot;, &quot;PRUTH_BAY&quot;, &quot;PRUTH_BAY&quot;, &quot;…
## $ site_id          &lt;chr&gt; &quot;PRUTH_BAY_INTERIOR4&quot;, &quot;PRUTH_BAY_INTERIOR4&quot;, &quot;PRUTH_…
## $ date             &lt;date&gt; 2016-05-13, 2016-05-13, 2016-05-13, 2016-05-13, 2016…
## $ sampling_bout    &lt;chr&gt; &quot;4&quot;, &quot;4&quot;, &quot;4&quot;, &quot;4&quot;, &quot;4&quot;, &quot;4&quot;, &quot;4&quot;, &quot;6&quot;, &quot;6&quot;, &quot;6&quot;, &quot;6&quot;…
## $ dive_supervisor  &lt;chr&gt; &quot;Zach&quot;, &quot;Zach&quot;, &quot;Zach&quot;, &quot;Zach&quot;, &quot;Zach&quot;, &quot;Zach&quot;, &quot;Zach…
## $ collector        &lt;chr&gt; &quot;Derek&quot;, &quot;Derek&quot;, &quot;Derek&quot;, &quot;Derek&quot;, &quot;Derek&quot;, &quot;Derek&quot;,…
## $ hakai_id         &lt;chr&gt; &quot;2016-05-13_PRUTH_BAY_INTERIOR4_0&quot;, &quot;2016-05-13_PRUTH…
## $ sample_type      &lt;chr&gt; &quot;seagrass_density&quot;, &quot;seagrass_density&quot;, &quot;seagrass_den…
## $ depth            &lt;dbl&gt; 6.0, 6.0, 6.0, 6.0, 5.0, 6.0, 6.0, 9.1, 9.0, 8.9, 9.0…
## $ transect_dist    &lt;fct&gt; 0, 5, 10, 15, 20, 25, 30, 10, 15, 20, 25, 30, 0, 5, 1…
## $ collected_start  &lt;dttm&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ collected_end    &lt;dttm&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ density          &lt;dbl&gt; 13, 10, 18, 22, 16, 31, 9, 5, 6, 6, 6, 3, 13, 30, 23,…
## $ density_msq      &lt;dbl&gt; 208, 160, 288, 352, 256, 496, 144, 80, 96, 96, 96, 48…
## $ canopy_height_cm &lt;dbl&gt; 60, 63, 80, 54, 55, 50, 63, 85, 80, 90, 95, 75, 60, 6…
## $ flowering_shoots &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, 0, 0, 0, 0, 0, NA, NA, NA…
## $ comments         &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
## $ quality_log      &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…</code></pre>
<p>Next, load the habitat survey data, and same as above, set variable classes as necessary,
and have a quick look.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="applications.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load habitat data, set variable classes, have a quick look</span></span>
<span id="cb23-2"><a href="applications.html#cb23-2" aria-hidden="true" tabindex="-1"></a>seagrassHabitat <span class="ot">&lt;-</span></span>
<span id="cb23-3"><a href="applications.html#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read.csv</span>(<span class="st">&quot;raw_data/seagrass_habitat_survey.csv&quot;</span>,</span>
<span id="cb23-4"><a href="applications.html#cb23-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">colClasses =</span> <span class="st">&quot;character&quot;</span>) <span class="sc">%&gt;%</span>  </span>
<span id="cb23-5"><a href="applications.html#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">date            =</span> <span class="fu">ymd</span>(date),</span>
<span id="cb23-6"><a href="applications.html#cb23-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">depth           =</span> <span class="fu">as.numeric</span>(depth),</span>
<span id="cb23-7"><a href="applications.html#cb23-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">hakai_id        =</span> <span class="fu">str_pad</span>(hakai_id, <span class="dv">5</span>, <span class="at">pad =</span> <span class="st">&quot;0&quot;</span>),</span>
<span id="cb23-8"><a href="applications.html#cb23-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">transect_dist   =</span> <span class="fu">factor</span>(transect_dist),</span>
<span id="cb23-9"><a href="applications.html#cb23-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">collected_start =</span> <span class="fu">ymd_hms</span>(collected_start),</span>
<span id="cb23-10"><a href="applications.html#cb23-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">collected_end   =</span> <span class="fu">ymd_hms</span>(collected_end)) <span class="sc">%T&gt;%</span></span>
<span id="cb23-11"><a href="applications.html#cb23-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glimpse</span>()</span></code></pre></div>
<pre><code>## Rows: 2,052
## Columns: 28
## $ X                &lt;chr&gt; &quot;1&quot;, &quot;2&quot;, &quot;3&quot;, &quot;4&quot;, &quot;5&quot;, &quot;6&quot;, &quot;7&quot;, &quot;8&quot;, &quot;9&quot;, &quot;10&quot;, &quot;1…
## $ organization     &lt;chr&gt; &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;,…
## $ work_area        &lt;chr&gt; &quot;CALVERT&quot;, &quot;CALVERT&quot;, &quot;CALVERT&quot;, &quot;CALVERT&quot;, &quot;CALVERT&quot;…
## $ project          &lt;chr&gt; &quot;MARINEGEO&quot;, &quot;MARINEGEO&quot;, &quot;MARINEGEO&quot;, &quot;MARINEGEO&quot;, &quot;…
## $ survey           &lt;chr&gt; &quot;CHOKED_PASS&quot;, &quot;CHOKED_PASS&quot;, &quot;CHOKED_PASS&quot;, &quot;CHOKED_…
## $ site_id          &lt;chr&gt; &quot;CHOKED_PASS_INTERIOR6&quot;, &quot;CHOKED_PASS_INTERIOR6&quot;, &quot;CH…
## $ date             &lt;date&gt; 2017-11-22, 2017-11-22, 2017-11-22, 2017-11-22, 2017…
## $ sampling_bout    &lt;chr&gt; &quot;6&quot;, &quot;6&quot;, &quot;6&quot;, &quot;6&quot;, &quot;6&quot;, &quot;6&quot;, &quot;1&quot;, &quot;1&quot;, &quot;1&quot;, &quot;1&quot;, &quot;1&quot;…
## $ dive_supervisor  &lt;chr&gt; &quot;gillian&quot;, &quot;gillian&quot;, &quot;gillian&quot;, &quot;gillian&quot;, &quot;gillian&quot;…
## $ collector        &lt;chr&gt; &quot;zach&quot;, &quot;zach&quot;, &quot;zach&quot;, &quot;zach&quot;, &quot;zach&quot;, &quot;zach&quot;, &quot;kyle…
## $ hakai_id         &lt;chr&gt; &quot;10883&quot;, &quot;2017-11-22_CHOKED_PASS_INTERIOR6_5 - 10&quot;, &quot;…
## $ sample_type      &lt;chr&gt; &quot;seagrass_habitat&quot;, &quot;seagrass_habitat&quot;, &quot;seagrass_hab…
## $ depth            &lt;dbl&gt; 9.2, 9.4, 9.3, 9.0, 9.2, 9.2, 3.4, 3.4, 3.4, 3.4, 3.4…
## $ transect_dist    &lt;fct&gt; 0 - 5, 10-May, 15-Oct, 15 - 20, 20 - 25, 25 - 30, 0 -…
## $ collected_start  &lt;dttm&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ collected_end    &lt;dttm&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ bag_uid          &lt;chr&gt; &quot;10883&quot;, NA, NA, &quot;11094&quot;, NA, &quot;11182&quot;, &quot;7119&quot;, NA, &quot;7…
## $ bag_number       &lt;chr&gt; &quot;3557&quot;, NA, NA, &quot;3520&quot;, NA, &quot;903&quot;, &quot;800&quot;, NA, &quot;318&quot;, …
## $ density_range    &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
## $ substrate        &lt;chr&gt; &quot;sand,shell hash&quot;, &quot;sand,shell hash&quot;, &quot;sand,shell has…
## $ patchiness       &lt;chr&gt; &quot;&lt; 1&quot;, &quot;&lt; 1&quot;, &quot;02-Jan&quot;, &quot;&lt; 1&quot;, &quot;&lt; 1&quot;, &quot;&lt; 1&quot;, &quot;&lt; 1&quot;, &quot;…
## $ adj_habitat_1    &lt;chr&gt; &quot;seagrass&quot;, &quot;seagrass&quot;, &quot;seagrass&quot;, &quot;seagrass&quot;, &quot;seag…
## $ adj_habitat_2    &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
## $ sample_collected &lt;chr&gt; &quot;TRUE&quot;, &quot;FALSE&quot;, &quot;FALSE&quot;, &quot;TRUE&quot;, &quot;FALSE&quot;, &quot;TRUE&quot;, &quot;T…
## $ vegetation_1     &lt;chr&gt; NA, NA, NA, NA, NA, NA, &quot;des&quot;, NA, &quot;des&quot;, NA, NA, NA,…
## $ vegetation_2     &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
## $ comments         &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
## $ quality_log      &lt;chr&gt; &quot;1: Flowering shoots 0 for entire transects&quot;, NA, NA,…</code></pre>
<p>Finally, load coordinate data for surveys, and subset necessary variables</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="applications.html#cb25-1" aria-hidden="true" tabindex="-1"></a>coordinates <span class="ot">&lt;-</span> </span>
<span id="cb25-2"><a href="applications.html#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">read.csv</span>(<span class="st">&quot;raw_data/seagrassCoordinates.csv&quot;</span>,</span>
<span id="cb25-3"><a href="applications.html#cb25-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">colClass =</span> <span class="fu">c</span>(<span class="st">&quot;Point.Name&quot;</span> <span class="ot">=</span> <span class="st">&quot;character&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb25-4"><a href="applications.html#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Point.Name, Decimal.Lat, Decimal.Long) <span class="sc">%T&gt;%</span></span>
<span id="cb25-5"><a href="applications.html#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glimpse</span>()</span></code></pre></div>
<pre><code>## Rows: 70
## Columns: 3
## $ Point.Name   &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
## $ Decimal.Lat  &lt;dbl&gt; 52.06200, 52.05200, 51.92270, 51.92500, 51.80900, 51.8090…
## $ Decimal.Long &lt;dbl&gt; -128.4120, -128.4030, -128.4648, -128.4540, -128.2360, -1…</code></pre>
</div>
<div id="merge-datasets" class="section level4 hasAnchor" number="2.3.1.2">
<h4><span class="header-section-number">2.3.1.2</span> Merge Datasets<a href="applications.html#merge-datasets" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Now all the datasets have been loaded, and briefly formatted, we’ll join
together the habitat and density surveys, and the coordinates for these.</p>
<p>The seagrass density surveys collect data at discrete points (ie. 5 metres)
along the transects, while the habitat surveys collect data over sections
(ie. 0 - 5 metres) along the transects. In order to fit these two surveys
together, we’ll narrow the habitat surveys from a range to a point so the
locations will match. Based on how the habitat data is collected, the point
the habitat survey is applied to will be the distance at the end of the
swath (ie. 10-15m will become 15m). To account for no preceeding distance,
the 0m distance will use the 0-5m section of the survey.</p>
<p>First, well make the necessary transformations to the habitat dataset.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="applications.html#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reformat seagrassHabitat to merge with seagrassDensity</span></span>
<span id="cb27-2"><a href="applications.html#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="do">## replicate 0 - 5m transect dist to match with 0m in density survey;</span></span>
<span id="cb27-3"><a href="applications.html#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="do">## rest of habitat bins can map one to one with density (ie. 5 - 10m -&gt; 10m)</span></span>
<span id="cb27-4"><a href="applications.html#cb27-4" aria-hidden="true" tabindex="-1"></a>seagrass0tmp <span class="ot">&lt;-</span> </span>
<span id="cb27-5"><a href="applications.html#cb27-5" aria-hidden="true" tabindex="-1"></a>  seagrassHabitat <span class="sc">%&gt;%</span></span>
<span id="cb27-6"><a href="applications.html#cb27-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transect_dist <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;0 - 5&quot;</span>, <span class="st">&quot;0 - 2.5&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb27-7"><a href="applications.html#cb27-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">transect_dist =</span> <span class="fu">factor</span>(<span class="dv">0</span>))</span>
<span id="cb27-8"><a href="applications.html#cb27-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-9"><a href="applications.html#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="do">## collapse various levels to match with seagrassDensity transect_dist</span></span>
<span id="cb27-10"><a href="applications.html#cb27-10" aria-hidden="true" tabindex="-1"></a>seagrassHabitat<span class="sc">$</span>transect_dist <span class="ot">&lt;-</span> </span>
<span id="cb27-11"><a href="applications.html#cb27-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fct_collapse</span>(seagrassHabitat<span class="sc">$</span>transect_dist,</span>
<span id="cb27-12"><a href="applications.html#cb27-12" aria-hidden="true" tabindex="-1"></a>               <span class="st">&quot;5&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;0 - 5&quot;</span>, <span class="st">&quot;2.5 - 7.5&quot;</span>),</span>
<span id="cb27-13"><a href="applications.html#cb27-13" aria-hidden="true" tabindex="-1"></a>               <span class="st">&quot;10&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;5 - 10&quot;</span>, <span class="st">&quot;7.5 - 12.5&quot;</span>),</span>
<span id="cb27-14"><a href="applications.html#cb27-14" aria-hidden="true" tabindex="-1"></a>               <span class="st">&quot;15&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;10 - 15&quot;</span>, <span class="st">&quot;12.5 - 17.5&quot;</span>),</span>
<span id="cb27-15"><a href="applications.html#cb27-15" aria-hidden="true" tabindex="-1"></a>               <span class="st">&quot;20&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;15 - 20&quot;</span>, <span class="st">&quot;17.5 - 22.5&quot;</span>),</span>
<span id="cb27-16"><a href="applications.html#cb27-16" aria-hidden="true" tabindex="-1"></a>               <span class="st">&quot;25&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;20 - 25&quot;</span>, <span class="st">&quot;22.5 - 27.5&quot;</span>),</span>
<span id="cb27-17"><a href="applications.html#cb27-17" aria-hidden="true" tabindex="-1"></a>               <span class="st">&quot;30&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;25 - 30&quot;</span>, <span class="st">&quot;27.5 - 30&quot;</span>))</span>
<span id="cb27-18"><a href="applications.html#cb27-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-19"><a href="applications.html#cb27-19" aria-hidden="true" tabindex="-1"></a><span class="do">## merge seagrass0tmp into seagrassHabitat to account for 0m samples,</span></span>
<span id="cb27-20"><a href="applications.html#cb27-20" aria-hidden="true" tabindex="-1"></a><span class="do">## set class for date, datetime variables</span></span>
<span id="cb27-21"><a href="applications.html#cb27-21" aria-hidden="true" tabindex="-1"></a>seagrassHabitatFull <span class="ot">&lt;-</span> </span>
<span id="cb27-22"><a href="applications.html#cb27-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbind</span>(seagrass0tmp, seagrassHabitat) <span class="sc">%&gt;%</span></span>
<span id="cb27-23"><a href="applications.html#cb27-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(transect_dist <span class="sc">!=</span> <span class="st">&quot;0 - 2.5&quot;</span>)  <span class="sc">%&gt;%</span> <span class="co"># already captured in seagrass0tmp </span></span>
<span id="cb27-24"><a href="applications.html#cb27-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">droplevels</span>(.)  <span class="co"># remove now unused factor levels</span></span></code></pre></div>
<p>With the distances of habitat and density surveys now corresponding, we can
now merge these two datasets plus there coordinates together, combine
redundant fields, and remove unnecessary fields.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="applications.html#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge seagrassHabitatFull with seagrassDensity, then coordinates</span></span>
<span id="cb28-2"><a href="applications.html#cb28-2" aria-hidden="true" tabindex="-1"></a>seagrass <span class="ot">&lt;-</span> </span>
<span id="cb28-3"><a href="applications.html#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">full_join</span>(seagrassHabitatFull, seagrassDensity, </span>
<span id="cb28-4"><a href="applications.html#cb28-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;organization&quot;</span>,</span>
<span id="cb28-5"><a href="applications.html#cb28-5" aria-hidden="true" tabindex="-1"></a>                   <span class="st">&quot;work_area&quot;</span>,</span>
<span id="cb28-6"><a href="applications.html#cb28-6" aria-hidden="true" tabindex="-1"></a>                   <span class="st">&quot;project&quot;</span>,</span>
<span id="cb28-7"><a href="applications.html#cb28-7" aria-hidden="true" tabindex="-1"></a>                   <span class="st">&quot;survey&quot;</span>,</span>
<span id="cb28-8"><a href="applications.html#cb28-8" aria-hidden="true" tabindex="-1"></a>                   <span class="st">&quot;site_id&quot;</span>, </span>
<span id="cb28-9"><a href="applications.html#cb28-9" aria-hidden="true" tabindex="-1"></a>                   <span class="st">&quot;date&quot;</span>,</span>
<span id="cb28-10"><a href="applications.html#cb28-10" aria-hidden="true" tabindex="-1"></a>                   <span class="st">&quot;transect_dist&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb28-11"><a href="applications.html#cb28-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># merge hakai_id.x and hakai_id.y into single variable field;</span></span>
<span id="cb28-12"><a href="applications.html#cb28-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># use combination of date, site_id, transect_dist, and field uid (hakai_id </span></span>
<span id="cb28-13"><a href="applications.html#cb28-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># when present)</span></span>
<span id="cb28-14"><a href="applications.html#cb28-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">field_uid =</span> <span class="fu">ifelse</span>(sample_collected <span class="sc">==</span> <span class="cn">TRUE</span>, hakai_id.x, <span class="st">&quot;NA&quot;</span>),</span>
<span id="cb28-15"><a href="applications.html#cb28-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">hakai_id =</span> <span class="fu">paste</span>(date, <span class="st">&quot;HAKAI:CALVERT&quot;</span>, site_id, transect_dist, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>),</span>
<span id="cb28-16"><a href="applications.html#cb28-16" aria-hidden="true" tabindex="-1"></a>         <span class="co"># below, aggregate metadata that didn&#39;t merge naturally (ie. due to minor </span></span>
<span id="cb28-17"><a href="applications.html#cb28-17" aria-hidden="true" tabindex="-1"></a>         <span class="co"># differences in watch time or depth gauges)</span></span>
<span id="cb28-18"><a href="applications.html#cb28-18" aria-hidden="true" tabindex="-1"></a>         <span class="at">dive_supervisor =</span> dive_supervisor.x,</span>
<span id="cb28-19"><a href="applications.html#cb28-19" aria-hidden="true" tabindex="-1"></a>         <span class="at">collected_start =</span> <span class="fu">ymd_hms</span>(<span class="fu">ifelse</span>(<span class="fu">is.na</span>(collected_start.x),</span>
<span id="cb28-20"><a href="applications.html#cb28-20" aria-hidden="true" tabindex="-1"></a>                                          collected_start.y, </span>
<span id="cb28-21"><a href="applications.html#cb28-21" aria-hidden="true" tabindex="-1"></a>                                          collected_start.x)),</span>
<span id="cb28-22"><a href="applications.html#cb28-22" aria-hidden="true" tabindex="-1"></a>         <span class="at">collected_end   =</span> <span class="fu">ymd_hms</span>(<span class="fu">ifelse</span>(<span class="fu">is.na</span>(collected_start.x),</span>
<span id="cb28-23"><a href="applications.html#cb28-23" aria-hidden="true" tabindex="-1"></a>                                          collected_start.y,</span>
<span id="cb28-24"><a href="applications.html#cb28-24" aria-hidden="true" tabindex="-1"></a>                                          collected_start.x)),</span>
<span id="cb28-25"><a href="applications.html#cb28-25" aria-hidden="true" tabindex="-1"></a>         <span class="at">depth_m         =</span> <span class="fu">ifelse</span>(<span class="fu">is.na</span>(depth.x), depth.y, depth.x),</span>
<span id="cb28-26"><a href="applications.html#cb28-26" aria-hidden="true" tabindex="-1"></a>         <span class="at">sampling_bout   =</span> sampling_bout.x) <span class="sc">%&gt;%</span></span>
<span id="cb28-27"><a href="applications.html#cb28-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(., coordinates,  <span class="co"># add coordinates</span></span>
<span id="cb28-28"><a href="applications.html#cb28-28" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;site_id&quot;</span> <span class="ot">=</span> <span class="st">&quot;Point.Name&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb28-29"><a href="applications.html#cb28-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>( <span class="sc">-</span> <span class="fu">c</span>(X.x, X.y, hakai_id.x, hakai_id.y,  <span class="co"># remove unnecessary variables</span></span>
<span id="cb28-30"><a href="applications.html#cb28-30" aria-hidden="true" tabindex="-1"></a>              dive_supervisor.x, dive_supervisor.y,</span>
<span id="cb28-31"><a href="applications.html#cb28-31" aria-hidden="true" tabindex="-1"></a>              collected_start.x, collected_start.y,</span>
<span id="cb28-32"><a href="applications.html#cb28-32" aria-hidden="true" tabindex="-1"></a>              collected_end.x, collected_end.y,</span>
<span id="cb28-33"><a href="applications.html#cb28-33" aria-hidden="true" tabindex="-1"></a>              depth.x, depth.y,</span>
<span id="cb28-34"><a href="applications.html#cb28-34" aria-hidden="true" tabindex="-1"></a>              sampling_bout.x, sampling_bout.y)) <span class="sc">%&gt;%</span></span>
<span id="cb28-35"><a href="applications.html#cb28-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">density_msq =</span> <span class="fu">as.character</span>(density_msq),</span>
<span id="cb28-36"><a href="applications.html#cb28-36" aria-hidden="true" tabindex="-1"></a>         <span class="at">canopy_height_cm =</span> <span class="fu">as.character</span>(canopy_height_cm),</span>
<span id="cb28-37"><a href="applications.html#cb28-37" aria-hidden="true" tabindex="-1"></a>         <span class="at">flowering_shoots =</span> <span class="fu">as.character</span>(flowering_shoots),</span>
<span id="cb28-38"><a href="applications.html#cb28-38" aria-hidden="true" tabindex="-1"></a>         <span class="at">depth_m =</span> <span class="fu">as.character</span>(depth_m)) <span class="sc">%T&gt;%</span></span>
<span id="cb28-39"><a href="applications.html#cb28-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glimpse</span>()</span></code></pre></div>
<pre><code>## Rows: 3,743
## Columns: 38
## $ organization     &lt;chr&gt; &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;, &quot;HAKAI&quot;,…
## $ work_area        &lt;chr&gt; &quot;CALVERT&quot;, &quot;CALVERT&quot;, &quot;CALVERT&quot;, &quot;CALVERT&quot;, &quot;CALVERT&quot;…
## $ project          &lt;chr&gt; &quot;MARINEGEO&quot;, &quot;MARINEGEO&quot;, &quot;MARINEGEO&quot;, &quot;MARINEGEO&quot;, &quot;…
## $ survey           &lt;chr&gt; &quot;CHOKED_PASS&quot;, &quot;CHOKED_PASS&quot;, &quot;CHOKED_PASS&quot;, &quot;PRUTH_B…
## $ site_id          &lt;chr&gt; &quot;CHOKED_PASS_INTERIOR6&quot;, &quot;CHOKED_PASS_EDGE1&quot;, &quot;CHOKED…
## $ date             &lt;date&gt; 2017-11-22, 2017-05-19, 2017-05-19, 2017-07-03, 2017…
## $ collector.x      &lt;chr&gt; &quot;zach&quot;, &quot;kyle&quot;, NA, &quot;tanya&quot;, &quot;zach&quot;, &quot;zach&quot;, &quot;zach&quot;, …
## $ sample_type.x    &lt;chr&gt; &quot;seagrass_habitat&quot;, &quot;seagrass_habitat&quot;, &quot;seagrass_hab…
## $ transect_dist    &lt;fct&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
## $ bag_uid          &lt;chr&gt; &quot;10883&quot;, &quot;7119&quot;, &quot;7031&quot;, &quot;2352&quot;, &quot;10255&quot;, &quot;10023&quot;, &quot;1…
## $ bag_number       &lt;chr&gt; &quot;3557&quot;, &quot;800&quot;, &quot;301&quot;, &quot;324&quot;, &quot;3506&quot;, &quot;3555&quot;, &quot;3534&quot;, …
## $ density_range    &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
## $ substrate        &lt;chr&gt; &quot;sand,shell hash&quot;, &quot;sand,shell hash&quot;, &quot;sand,shell has…
## $ patchiness       &lt;chr&gt; &quot;&lt; 1&quot;, &quot;&lt; 1&quot;, &quot;&lt; 1&quot;, &quot;&lt; 1&quot;, &quot;&lt; 1&quot;, &quot;05-Apr&quot;, &quot;04-Mar&quot;…
## $ adj_habitat_1    &lt;chr&gt; &quot;seagrass&quot;, &quot;sand&quot;, &quot;standing kelp&quot;, &quot;seagrass&quot;, &quot;sea…
## $ adj_habitat_2    &lt;chr&gt; NA, NA, NA, NA, NA, NA, &quot;standing kelp&quot;, NA, NA, NA, …
## $ sample_collected &lt;chr&gt; &quot;TRUE&quot;, &quot;TRUE&quot;, &quot;TRUE&quot;, &quot;TRUE&quot;, &quot;TRUE&quot;, &quot;TRUE&quot;, &quot;TRUE…
## $ vegetation_1     &lt;chr&gt; NA, &quot;des&quot;, &quot;des&quot;, &quot;zm&quot;, &quot;des&quot;, NA, NA, NA, NA, NA, NA…
## $ vegetation_2     &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, &quot;…
## $ comments.x       &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
## $ quality_log.x    &lt;chr&gt; &quot;1: Flowering shoots 0 for entire transects&quot;, NA, NA,…
## $ collector.y      &lt;chr&gt; &quot;derek&quot;, &quot;ondine&quot;, &quot;ondine&quot;, &quot;derek&quot;, &quot;derek&quot;, &quot;derek…
## $ sample_type.y    &lt;chr&gt; &quot;seagrass_density&quot;, &quot;seagrass_density&quot;, &quot;seagrass_den…
## $ density          &lt;dbl&gt; 4, 10, 6, 13, 6, 1, 2, 6, 21, 3, 7, 4, 3, 14, 17, 11,…
## $ density_msq      &lt;chr&gt; &quot;64&quot;, &quot;160&quot;, &quot;96&quot;, &quot;208&quot;, &quot;96&quot;, &quot;16&quot;, &quot;32&quot;, &quot;96&quot;, &quot;33…
## $ canopy_height_cm &lt;chr&gt; &quot;80&quot;, &quot;80&quot;, &quot;110&quot;, &quot;60&quot;, &quot;125&quot;, &quot;100&quot;, &quot;100&quot;, &quot;125&quot;, …
## $ flowering_shoots &lt;chr&gt; &quot;0&quot;, NA, NA, NA, NA, NA, NA, &quot;0&quot;, NA, NA, NA, &quot;0&quot;, NA…
## $ comments.y       &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
## $ quality_log.y    &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, &quot;…
## $ field_uid        &lt;chr&gt; &quot;10883&quot;, &quot;07119&quot;, &quot;07031&quot;, &quot;02352&quot;, &quot;10255&quot;, &quot;10023&quot;,…
## $ hakai_id         &lt;chr&gt; &quot;2017-11-22:HAKAI:CALVERT:CHOKED_PASS_INTERIOR6:0&quot;, &quot;…
## $ dive_supervisor  &lt;chr&gt; &quot;gillian&quot;, &quot;gillian,gillian.sadlierbrown&quot;, &quot;gillian,g…
## $ collected_start  &lt;dttm&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ collected_end    &lt;dttm&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ depth_m          &lt;chr&gt; &quot;9.2&quot;, &quot;3.4&quot;, &quot;4.8&quot;, &quot;2.4&quot;, &quot;5.3&quot;, &quot;5.6&quot;, &quot;4.4&quot;, &quot;2.5…
## $ sampling_bout    &lt;chr&gt; &quot;6&quot;, &quot;1&quot;, &quot;3&quot;, &quot;5&quot;, &quot;5&quot;, &quot;3&quot;, &quot;5&quot;, &quot;2&quot;, &quot;1&quot;, &quot;2&quot;, &quot;6&quot;…
## $ Decimal.Lat      &lt;dbl&gt; 51.67482, 51.67882, 51.67493, 51.64532, 51.67349, 51.…
## $ Decimal.Long     &lt;dbl&gt; -128.1195, -128.1148, -128.1237, -128.1193, -128.1180…</code></pre>
</div>
</div>
<div id="convert-data-to-darwin-core---extended-measurement-or-fact-format" class="section level3 hasAnchor" number="2.3.2">
<h3><span class="header-section-number">2.3.2</span> Convert Data to Darwin Core - Extended Measurement or Fact format<a href="applications.html#convert-data-to-darwin-core---extended-measurement-or-fact-format" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The Darwin Core ExtendedMeasurementOrFact (eMoF) extension bases records
around a core event (rather than occurrence as in standard Darwin Core),
allowing for additional measurement variables to be associated with
occurrence data.</p>
<div id="add-event-id-and-occurrence-id-variables-to-dataset" class="section level4 hasAnchor" number="2.3.2.1">
<h4><span class="header-section-number">2.3.2.1</span> Add Event ID and Occurrence ID variables to dataset<a href="applications.html#add-event-id-and-occurrence-id-variables-to-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>As this dataset will be annually updated, rather than using
natural keys (ie. using package::uuid to autogenerate) for event and
occurence IDs, here we will use surrogate keys made up of a concatenation
of date survey, transect location, observation distance, and sample ID
(for occurrenceID, when a sample is present).</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="applications.html#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create and populate eventID variable</span></span>
<span id="cb30-2"><a href="applications.html#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="do">## currently only event is used, but additional surveys and abiotic data</span></span>
<span id="cb30-3"><a href="applications.html#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="do">## are associated with parent events that may be included at a later date</span></span>
<span id="cb30-4"><a href="applications.html#cb30-4" aria-hidden="true" tabindex="-1"></a>seagrass<span class="sc">$</span>eventID <span class="ot">&lt;-</span> seagrass<span class="sc">$</span>hakai_id</span>
<span id="cb30-5"><a href="applications.html#cb30-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-6"><a href="applications.html#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="co"># create and populate occurrenceID; combine eventID with transect_dist </span></span>
<span id="cb30-7"><a href="applications.html#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="co"># and field_uid</span></span>
<span id="cb30-8"><a href="applications.html#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="do">## in the event of &lt;NA&gt; field_uid, no sample was collected, but</span></span>
<span id="cb30-9"><a href="applications.html#cb30-9" aria-hidden="true" tabindex="-1"></a><span class="do">## measurements and occurrence are still taken; no further subsamples</span></span>
<span id="cb30-10"><a href="applications.html#cb30-10" aria-hidden="true" tabindex="-1"></a><span class="do">## are associated with &lt;NA&gt; field_uids</span></span>
<span id="cb30-11"><a href="applications.html#cb30-11" aria-hidden="true" tabindex="-1"></a>seagrass<span class="sc">$</span>occurrenceID <span class="ot">&lt;-</span> </span>
<span id="cb30-12"><a href="applications.html#cb30-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">with</span>(seagrass, </span>
<span id="cb30-13"><a href="applications.html#cb30-13" aria-hidden="true" tabindex="-1"></a>       <span class="fu">paste</span>(eventID, transect_dist, field_uid, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>))</span></code></pre></div>
</div>
<div id="create-event-occurrence-and-emof-tables" class="section level4 hasAnchor" number="2.3.2.2">
<h4><span class="header-section-number">2.3.2.2</span> Create Event, Occurrence, and eMoF tables<a href="applications.html#create-event-occurrence-and-emof-tables" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Now that we’ve created eventIDs and occurrenceIDs to connect all the
variables together, we can begin to create the Event, Occurrence,
and extended Measurement or Fact table necessary for DarwinCore
compliant datasets</p>
<div id="event-table" class="section level5 hasAnchor" number="2.3.2.2.1">
<h5><span class="header-section-number">2.3.2.2.1</span> Event Table<a href="applications.html#event-table" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="applications.html#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset seagrass to create event table</span></span>
<span id="cb31-2"><a href="applications.html#cb31-2" aria-hidden="true" tabindex="-1"></a>seagrassEvent <span class="ot">&lt;-</span></span>
<span id="cb31-3"><a href="applications.html#cb31-3" aria-hidden="true" tabindex="-1"></a>  seagrass <span class="sc">%&gt;%</span></span>
<span id="cb31-4"><a href="applications.html#cb31-4" aria-hidden="true" tabindex="-1"></a>  distinct <span class="sc">%&gt;%</span>  <span class="co"># some duplicates in data stemming from database conflicts</span></span>
<span id="cb31-5"><a href="applications.html#cb31-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(date,</span>
<span id="cb31-6"><a href="applications.html#cb31-6" aria-hidden="true" tabindex="-1"></a>         Decimal.Lat, Decimal.Long, transect_dist,</span>
<span id="cb31-7"><a href="applications.html#cb31-7" aria-hidden="true" tabindex="-1"></a>         depth_m, eventID) <span class="sc">%&gt;%</span></span>
<span id="cb31-8"><a href="applications.html#cb31-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">eventDate                     =</span> date,</span>
<span id="cb31-9"><a href="applications.html#cb31-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">decimalLatitude               =</span> Decimal.Lat,</span>
<span id="cb31-10"><a href="applications.html#cb31-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">decimalLongitude              =</span> Decimal.Long,</span>
<span id="cb31-11"><a href="applications.html#cb31-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">coordinateUncertaintyInMeters =</span> transect_dist,</span>
<span id="cb31-12"><a href="applications.html#cb31-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">minimumDepthInMeters          =</span> depth_m,</span>
<span id="cb31-13"><a href="applications.html#cb31-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">maximumDepthInMeters          =</span> depth_m) <span class="sc">%&gt;%</span></span>
<span id="cb31-14"><a href="applications.html#cb31-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">geodeticDatum  =</span> <span class="st">&quot;WGS84&quot;</span>,</span>
<span id="cb31-15"><a href="applications.html#cb31-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">samplingEffort =</span> <span class="st">&quot;30 metre transect&quot;</span>) <span class="sc">%T&gt;%</span> glimpse</span></code></pre></div>
<pre><code>## Rows: 3,659
## Columns: 8
## $ eventDate                     &lt;date&gt; 2017-11-22, 2017-05-19, 2017-05-19, 201…
## $ decimalLatitude               &lt;dbl&gt; 51.67482, 51.67882, 51.67493, 51.64532, …
## $ decimalLongitude              &lt;dbl&gt; -128.1195, -128.1148, -128.1237, -128.11…
## $ coordinateUncertaintyInMeters &lt;fct&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0…
## $ maximumDepthInMeters          &lt;chr&gt; &quot;9.2&quot;, &quot;3.4&quot;, &quot;4.8&quot;, &quot;2.4&quot;, &quot;5.3&quot;, &quot;5.6&quot;…
## $ eventID                       &lt;chr&gt; &quot;2017-11-22:HAKAI:CALVERT:CHOKED_PASS_IN…
## $ geodeticDatum                 &lt;chr&gt; &quot;WGS84&quot;, &quot;WGS84&quot;, &quot;WGS84&quot;, &quot;WGS84&quot;, &quot;WGS…
## $ samplingEffort                &lt;chr&gt; &quot;30 metre transect&quot;, &quot;30 metre transect&quot;…</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="applications.html#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save event table to csv</span></span>
<span id="cb33-2"><a href="applications.html#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(seagrassEvent, <span class="st">&quot;processed_data/hakaiSeagrassDwcEvent.csv&quot;</span>)</span></code></pre></div>
</div>
<div id="occurrence-table" class="section level5 hasAnchor" number="2.3.2.2.2">
<h5><span class="header-section-number">2.3.2.2.2</span> Occurrence Table<a href="applications.html#occurrence-table" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="applications.html#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset seagrass to create occurrence table</span></span>
<span id="cb34-2"><a href="applications.html#cb34-2" aria-hidden="true" tabindex="-1"></a>seagrassOccurrence <span class="ot">&lt;-</span></span>
<span id="cb34-3"><a href="applications.html#cb34-3" aria-hidden="true" tabindex="-1"></a>  seagrass <span class="sc">%&gt;%</span></span>
<span id="cb34-4"><a href="applications.html#cb34-4" aria-hidden="true" tabindex="-1"></a>  distinct <span class="sc">%&gt;%</span>  <span class="co"># some duplicates in data stemming from database conflicts</span></span>
<span id="cb34-5"><a href="applications.html#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(eventID, occurrenceID) <span class="sc">%&gt;%</span></span>
<span id="cb34-6"><a href="applications.html#cb34-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">basisOfRecord =</span> <span class="st">&quot;HumanObservation&quot;</span>,</span>
<span id="cb34-7"><a href="applications.html#cb34-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">scientificName   =</span> <span class="st">&quot;Zostera subg. Zostera marina&quot;</span>,</span>
<span id="cb34-8"><a href="applications.html#cb34-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceStatus =</span> <span class="st">&quot;present&quot;</span>)</span>
<span id="cb34-9"><a href="applications.html#cb34-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-10"><a href="applications.html#cb34-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Taxonomic name matching</span></span>
<span id="cb34-11"><a href="applications.html#cb34-11" aria-hidden="true" tabindex="-1"></a><span class="co"># in addition to the above metadata, DarwinCore format requires further</span></span>
<span id="cb34-12"><a href="applications.html#cb34-12" aria-hidden="true" tabindex="-1"></a><span class="co"># taxonomic data that can be acquired through the WoRMS register.</span></span>
<span id="cb34-13"><a href="applications.html#cb34-13" aria-hidden="true" tabindex="-1"></a><span class="do">## Load taxonomic info, downloaded via WoRMS tool</span></span>
<span id="cb34-14"><a href="applications.html#cb34-14" aria-hidden="true" tabindex="-1"></a><span class="co"># zmWorms &lt;- </span></span>
<span id="cb34-15"><a href="applications.html#cb34-15" aria-hidden="true" tabindex="-1"></a><span class="co">#   read.delim(&quot;raw_data/zmworms_matched.txt&quot;,</span></span>
<span id="cb34-16"><a href="applications.html#cb34-16" aria-hidden="true" tabindex="-1"></a><span class="co">#              header = TRUE,</span></span>
<span id="cb34-17"><a href="applications.html#cb34-17" aria-hidden="true" tabindex="-1"></a><span class="co">#              nrows  = 1)</span></span>
<span id="cb34-18"><a href="applications.html#cb34-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-19"><a href="applications.html#cb34-19" aria-hidden="true" tabindex="-1"></a>zmWorms <span class="ot">&lt;-</span> <span class="fu">wm_record</span>(<span class="at">id =</span> <span class="dv">145795</span>)</span>
<span id="cb34-20"><a href="applications.html#cb34-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-21"><a href="applications.html#cb34-21" aria-hidden="true" tabindex="-1"></a><span class="co"># join WoRMS name with seagrassOccurrence create above</span></span>
<span id="cb34-22"><a href="applications.html#cb34-22" aria-hidden="true" tabindex="-1"></a>seagrassOccurrence <span class="ot">&lt;-</span> </span>
<span id="cb34-23"><a href="applications.html#cb34-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">full_join</span>(seagrassOccurrence, zmWorms, </span>
<span id="cb34-24"><a href="applications.html#cb34-24" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;scientificName&quot;</span> <span class="ot">=</span> <span class="st">&quot;scientificname&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb34-25"><a href="applications.html#cb34-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(eventID, occurrenceID, basisOfRecord, scientificName, occurrenceStatus, AphiaID,</span>
<span id="cb34-26"><a href="applications.html#cb34-26" aria-hidden="true" tabindex="-1"></a>         url, authority, status, unacceptreason, taxonRankID, rank,</span>
<span id="cb34-27"><a href="applications.html#cb34-27" aria-hidden="true" tabindex="-1"></a>         valid_AphiaID, valid_name, valid_authority, parentNameUsageID,</span>
<span id="cb34-28"><a href="applications.html#cb34-28" aria-hidden="true" tabindex="-1"></a>         kingdom, phylum, class, order, family, genus, citation, lsid,</span>
<span id="cb34-29"><a href="applications.html#cb34-29" aria-hidden="true" tabindex="-1"></a>         isMarine, match_type, modified) <span class="sc">%T&gt;%</span></span>
<span id="cb34-30"><a href="applications.html#cb34-30" aria-hidden="true" tabindex="-1"></a>  glimpse</span></code></pre></div>
<pre><code>## Rows: 3,659
## Columns: 27
## $ eventID           &lt;chr&gt; &quot;2017-11-22:HAKAI:CALVERT:CHOKED_PASS_INTERIOR6:0&quot;, …
## $ occurrenceID      &lt;chr&gt; &quot;2017-11-22:HAKAI:CALVERT:CHOKED_PASS_INTERIOR6:0:0:…
## $ basisOfRecord     &lt;chr&gt; &quot;HumanObservation&quot;, &quot;HumanObservation&quot;, &quot;HumanObserv…
## $ scientificName    &lt;chr&gt; &quot;Zostera subg. Zostera marina&quot;, &quot;Zostera subg. Zoste…
## $ occurrenceStatus  &lt;chr&gt; &quot;present&quot;, &quot;present&quot;, &quot;present&quot;, &quot;present&quot;, &quot;present…
## $ AphiaID           &lt;int&gt; 145795, 145795, 145795, 145795, 145795, 145795, 1457…
## $ url               &lt;chr&gt; &quot;https://www.marinespecies.org/aphia.php?p=taxdetail…
## $ authority         &lt;chr&gt; &quot;Linnaeus, 1753&quot;, &quot;Linnaeus, 1753&quot;, &quot;Linnaeus, 1753&quot;…
## $ status            &lt;chr&gt; &quot;accepted&quot;, &quot;accepted&quot;, &quot;accepted&quot;, &quot;accepted&quot;, &quot;acc…
## $ unacceptreason    &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
## $ taxonRankID       &lt;int&gt; 220, 220, 220, 220, 220, 220, 220, 220, 220, 220, 22…
## $ rank              &lt;chr&gt; &quot;Species&quot;, &quot;Species&quot;, &quot;Species&quot;, &quot;Species&quot;, &quot;Species…
## $ valid_AphiaID     &lt;int&gt; 145795, 145795, 145795, 145795, 145795, 145795, 1457…
## $ valid_name        &lt;chr&gt; &quot;Zostera subg. Zostera marina&quot;, &quot;Zostera subg. Zoste…
## $ valid_authority   &lt;chr&gt; &quot;Linnaeus, 1753&quot;, &quot;Linnaeus, 1753&quot;, &quot;Linnaeus, 1753&quot;…
## $ parentNameUsageID &lt;int&gt; 370435, 370435, 370435, 370435, 370435, 370435, 3704…
## $ kingdom           &lt;chr&gt; &quot;Plantae&quot;, &quot;Plantae&quot;, &quot;Plantae&quot;, &quot;Plantae&quot;, &quot;Plantae…
## $ phylum            &lt;chr&gt; &quot;Tracheophyta&quot;, &quot;Tracheophyta&quot;, &quot;Tracheophyta&quot;, &quot;Tra…
## $ class             &lt;chr&gt; &quot;Magnoliopsida&quot;, &quot;Magnoliopsida&quot;, &quot;Magnoliopsida&quot;, &quot;…
## $ order             &lt;chr&gt; &quot;Alismatales&quot;, &quot;Alismatales&quot;, &quot;Alismatales&quot;, &quot;Alisma…
## $ family            &lt;chr&gt; &quot;Zosteraceae&quot;, &quot;Zosteraceae&quot;, &quot;Zosteraceae&quot;, &quot;Zoster…
## $ genus             &lt;chr&gt; &quot;Zostera&quot;, &quot;Zostera&quot;, &quot;Zostera&quot;, &quot;Zostera&quot;, &quot;Zostera…
## $ citation          &lt;chr&gt; &quot;WoRMS (2024). Zostera subg. Zostera marina Linnaeus…
## $ lsid              &lt;chr&gt; &quot;urn:lsid:marinespecies.org:taxname:145795&quot;, &quot;urn:ls…
## $ isMarine          &lt;int&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
## $ match_type        &lt;chr&gt; &quot;exact&quot;, &quot;exact&quot;, &quot;exact&quot;, &quot;exact&quot;, &quot;exact&quot;, &quot;exact&quot;…
## $ modified          &lt;chr&gt; &quot;2008-12-09T10:03:16.140Z&quot;, &quot;2008-12-09T10:03:16.140…</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="applications.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save occurrence table to csv</span></span>
<span id="cb36-2"><a href="applications.html#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(seagrassOccurrence, <span class="st">&quot;processed_data/hakaiSeagrassDwcOccurrence.csv&quot;</span>)</span></code></pre></div>
</div>
<div id="extended-measurementorfact-table" class="section level5 hasAnchor" number="2.3.2.2.3">
<h5><span class="header-section-number">2.3.2.2.3</span> Extended MeasurementOrFact table<a href="applications.html#extended-measurementorfact-table" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="applications.html#cb37-1" aria-hidden="true" tabindex="-1"></a>seagrassMof <span class="ot">&lt;-</span></span>
<span id="cb37-2"><a href="applications.html#cb37-2" aria-hidden="true" tabindex="-1"></a>  seagrass <span class="sc">%&gt;%</span></span>
<span id="cb37-3"><a href="applications.html#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># select variables for eMoF table</span></span>
<span id="cb37-4"><a href="applications.html#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(date,</span>
<span id="cb37-5"><a href="applications.html#cb37-5" aria-hidden="true" tabindex="-1"></a>         eventID, survey, site_id, transect_dist,</span>
<span id="cb37-6"><a href="applications.html#cb37-6" aria-hidden="true" tabindex="-1"></a>         substrate, patchiness, adj_habitat_1, adj_habitat_2,</span>
<span id="cb37-7"><a href="applications.html#cb37-7" aria-hidden="true" tabindex="-1"></a>         vegetation_1, vegetation_2,</span>
<span id="cb37-8"><a href="applications.html#cb37-8" aria-hidden="true" tabindex="-1"></a>         density_msq, canopy_height_cm, flowering_shoots) <span class="sc">%&gt;%</span></span>
<span id="cb37-9"><a href="applications.html#cb37-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># split substrate into two variables (currently holds two substrate type in same variable)</span></span>
<span id="cb37-10"><a href="applications.html#cb37-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(substrate, <span class="at">sep =</span> <span class="st">&quot;,&quot;</span>, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">&quot;substrate_1&quot;</span>, <span class="st">&quot;substrate_2&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb37-11"><a href="applications.html#cb37-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># change variables names to match NERC database (or to be more descriptive where none exist)</span></span>
<span id="cb37-12"><a href="applications.html#cb37-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">measurementDeterminedDate   =</span> date,</span>
<span id="cb37-13"><a href="applications.html#cb37-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">SubstrateTypeA              =</span> substrate_1,</span>
<span id="cb37-14"><a href="applications.html#cb37-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">SubstrateTypeB              =</span> substrate_2,</span>
<span id="cb37-15"><a href="applications.html#cb37-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">BarePatchLengthWithinSeagrass =</span> patchiness,</span>
<span id="cb37-16"><a href="applications.html#cb37-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">PrimaryAdjacentHabitat      =</span> adj_habitat_1,</span>
<span id="cb37-17"><a href="applications.html#cb37-17" aria-hidden="true" tabindex="-1"></a>         <span class="at">SecondaryAdjacentHabitat    =</span> adj_habitat_2,</span>
<span id="cb37-18"><a href="applications.html#cb37-18" aria-hidden="true" tabindex="-1"></a>         <span class="at">PrimaryAlgaeSp              =</span> vegetation_1,</span>
<span id="cb37-19"><a href="applications.html#cb37-19" aria-hidden="true" tabindex="-1"></a>         <span class="at">SecondaryAlgaeSp            =</span> vegetation_2,</span>
<span id="cb37-20"><a href="applications.html#cb37-20" aria-hidden="true" tabindex="-1"></a>         <span class="at">BedAbund                    =</span> density_msq,</span>
<span id="cb37-21"><a href="applications.html#cb37-21" aria-hidden="true" tabindex="-1"></a>         <span class="at">CanopyHeight                =</span> canopy_height_cm,</span>
<span id="cb37-22"><a href="applications.html#cb37-22" aria-hidden="true" tabindex="-1"></a>         <span class="at">FloweringBedAbund           =</span> flowering_shoots) <span class="sc">%&gt;%</span>  </span>
<span id="cb37-23"><a href="applications.html#cb37-23" aria-hidden="true" tabindex="-1"></a>  <span class="co"># reformat variables into DwC MeasurementOrFact format</span></span>
<span id="cb37-24"><a href="applications.html#cb37-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># (single values variable, with measurement type, unit, etc. variables)</span></span>
<span id="cb37-25"><a href="applications.html#cb37-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>( <span class="sc">-</span> <span class="fu">c</span>(measurementDeterminedDate, eventID, survey, site_id, transect_dist),</span>
<span id="cb37-26"><a href="applications.html#cb37-26" aria-hidden="true" tabindex="-1"></a>                <span class="at">names_to =</span> <span class="st">&quot;measurementType&quot;</span>,</span>
<span id="cb37-27"><a href="applications.html#cb37-27" aria-hidden="true" tabindex="-1"></a>                <span class="at">values_to =</span> <span class="st">&quot;measurementValue&quot;</span>,</span>
<span id="cb37-28"><a href="applications.html#cb37-28" aria-hidden="true" tabindex="-1"></a>                <span class="at">values_ptypes =</span> <span class="fu">list</span>(<span class="at">measurementValue =</span> <span class="st">&quot;character&quot;</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb37-29"><a href="applications.html#cb37-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># use measurement type to fill in remainder of variables relating to </span></span>
<span id="cb37-30"><a href="applications.html#cb37-30" aria-hidden="true" tabindex="-1"></a>  <span class="co"># NERC vocabulary and metadata fields</span></span>
<span id="cb37-31"><a href="applications.html#cb37-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb37-32"><a href="applications.html#cb37-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">measurementTypeID =</span> <span class="fu">case_when</span>(</span>
<span id="cb37-33"><a href="applications.html#cb37-33" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;BedAbund&quot;</span> <span class="sc">~</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL02/&quot;</span>,</span>
<span id="cb37-34"><a href="applications.html#cb37-34" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;CanopyHeight&quot;</span> <span class="sc">~</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P01/current/OBSMAXLX/&quot;</span>,</span>
<span id="cb37-35"><a href="applications.html#cb37-35" aria-hidden="true" tabindex="-1"></a>      <span class="co"># measurementType == &quot;BarePatchWithinSeagrass&quot; ~ &quot;&quot;,</span></span>
<span id="cb37-36"><a href="applications.html#cb37-36" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;FloweringBedAbund&quot;</span> <span class="sc">~</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P01/current/SDBIOL02/&quot;</span>),</span>
<span id="cb37-37"><a href="applications.html#cb37-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">measurementUnit =</span> <span class="fu">case_when</span>(</span>
<span id="cb37-38"><a href="applications.html#cb37-38" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;BedAbund&quot;</span> <span class="sc">~</span> <span class="st">&quot;Number per square metre&quot;</span>,</span>
<span id="cb37-39"><a href="applications.html#cb37-39" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;CanopyHeight&quot;</span> <span class="sc">~</span> <span class="st">&quot;Centimetres&quot;</span>,</span>
<span id="cb37-40"><a href="applications.html#cb37-40" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;BarePatchhLengthWithinSeagrass&quot;</span> <span class="sc">~</span> <span class="st">&quot;Metres&quot;</span>,</span>
<span id="cb37-41"><a href="applications.html#cb37-41" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;FloweringBedAbund&quot;</span> <span class="sc">~</span> <span class="st">&quot;Number per square metre&quot;</span>),</span>
<span id="cb37-42"><a href="applications.html#cb37-42" aria-hidden="true" tabindex="-1"></a>    <span class="at">measurementUnitID =</span> <span class="fu">case_when</span>(</span>
<span id="cb37-43"><a href="applications.html#cb37-43" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;BedAbund&quot;</span> <span class="sc">~</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P06/current/UPMS/&quot;</span>,</span>
<span id="cb37-44"><a href="applications.html#cb37-44" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;CanopyHeight&quot;</span> <span class="sc">~</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P06/current/ULCM/&quot;</span>,</span>
<span id="cb37-45"><a href="applications.html#cb37-45" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;BarePatchhLengthWithinSeagrass&quot;</span> <span class="sc">~</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P06/current/ULAA/2/&quot;</span>,</span>
<span id="cb37-46"><a href="applications.html#cb37-46" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;FloweringBedAbund&quot;</span> <span class="sc">~</span> <span class="st">&quot;http://vocab.nerc.ac.uk/collection/P06/current/UPMS/&quot;</span>),</span>
<span id="cb37-47"><a href="applications.html#cb37-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">measurementAccuracy =</span> <span class="fu">case_when</span>(</span>
<span id="cb37-48"><a href="applications.html#cb37-48" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;CanopyHeight&quot;</span> <span class="sc">~</span> <span class="dv">5</span>),</span>
<span id="cb37-49"><a href="applications.html#cb37-49" aria-hidden="true" tabindex="-1"></a>    <span class="at">measurementMethod =</span> <span class="fu">case_when</span>(</span>
<span id="cb37-50"><a href="applications.html#cb37-50" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;BedAbund&quot;</span> <span class="sc">~</span> <span class="st">&quot;25cmx25cm quadrat count&quot;</span>,</span>
<span id="cb37-51"><a href="applications.html#cb37-51" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;CanopyHeight&quot;</span> <span class="sc">~</span> <span class="st">&quot;in situ with ruler&quot;</span>,</span>
<span id="cb37-52"><a href="applications.html#cb37-52" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;BarePatchhLengthWithinSeagrass&quot;</span> <span class="sc">~</span> <span class="st">&quot;estimated along transect line&quot;</span>,</span>
<span id="cb37-53"><a href="applications.html#cb37-53" aria-hidden="true" tabindex="-1"></a>      measurementType <span class="sc">==</span> <span class="st">&quot;FloweringBedAbund&quot;</span> <span class="sc">~</span> <span class="st">&quot;25cmx25cm quadrat count&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb37-54"><a href="applications.html#cb37-54" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(eventID, measurementDeterminedDate, measurementType, measurementValue,</span>
<span id="cb37-55"><a href="applications.html#cb37-55" aria-hidden="true" tabindex="-1"></a>         measurementTypeID, measurementUnit, measurementUnitID, measurementAccuracy,</span>
<span id="cb37-56"><a href="applications.html#cb37-56" aria-hidden="true" tabindex="-1"></a>         measurementMethod) <span class="sc">%T&gt;%</span></span>
<span id="cb37-57"><a href="applications.html#cb37-57" aria-hidden="true" tabindex="-1"></a><span class="co">#  select(!c(survey, site_id, transect_dist)) %T&gt;%</span></span>
<span id="cb37-58"><a href="applications.html#cb37-58" aria-hidden="true" tabindex="-1"></a>  <span class="fu">glimpse</span>()</span></code></pre></div>
<pre><code>## Rows: 37,430
## Columns: 9
## $ eventID                   &lt;chr&gt; &quot;2017-11-22:HAKAI:CALVERT:CHOKED_PASS_INTERI…
## $ measurementDeterminedDate &lt;date&gt; 2017-11-22, 2017-11-22, 2017-11-22, 2017-11…
## $ measurementType           &lt;chr&gt; &quot;SubstrateTypeA&quot;, &quot;SubstrateTypeB&quot;, &quot;BarePat…
## $ measurementValue          &lt;chr&gt; &quot;sand&quot;, &quot;shell hash&quot;, &quot;&lt; 1&quot;, &quot;seagrass&quot;, NA,…
## $ measurementTypeID         &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, &quot;http://vocab.ne…
## $ measurementUnit           &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, &quot;Number per squa…
## $ measurementUnitID         &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, &quot;http://vocab.ne…
## $ measurementAccuracy       &lt;dbl&gt; NA, NA, NA, NA, NA, NA, NA, NA, 5, NA, NA, N…
## $ measurementMethod         &lt;chr&gt; NA, NA, NA, NA, NA, NA, NA, &quot;25cmx25cm quadr…</code></pre>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="applications.html#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save eMoF table to csv</span></span>
<span id="cb39-2"><a href="applications.html#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(seagrassMof, <span class="st">&quot;processed_data/hakaiSeagrassDwcEmof.csv&quot;</span>)</span></code></pre></div>
</div>
</div>
</div>
<div id="session-info" class="section level3 hasAnchor" number="2.3.3">
<h3><span class="header-section-number">2.3.3</span> Session Info<a href="applications.html#session-info" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Print session information below in case necessary for future reference</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="applications.html#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print Session Info for future reference</span></span>
<span id="cb40-2"><a href="applications.html#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.1.1 (2021-08-10)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] worrms_0.4.3    magrittr_2.0.3  knitr_1.42      here_1.0.1     
##  [5] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.0   dplyr_1.1.2    
##  [9] purrr_1.0.1     readr_2.1.5     tidyr_1.3.0     tibble_3.2.1   
## [13] ggplot2_3.4.2   tidyverse_2.0.0
## 
## loaded via a namespace (and not attached):
##  [1] tidyselect_1.2.1  xfun_0.39         bslib_0.4.2       colorspace_2.1-0 
##  [5] vctrs_0.6.5       generics_0.1.3    htmltools_0.5.5   yaml_2.3.7       
##  [9] utf8_1.2.3        rlang_1.1.1       jquerylib_0.1.4   pillar_1.9.0     
## [13] httpcode_0.3.0    glue_1.6.2        withr_2.5.0       bit64_4.0.5      
## [17] readxl_1.4.3      lifecycle_1.0.3   munsell_0.5.0     gtable_0.3.3     
## [21] cellranger_1.1.0  evaluate_0.21     tzdb_0.3.0        fastmap_1.1.1    
## [25] curl_5.0.0        parallel_4.1.1    fansi_1.0.4       triebeard_0.4.1  
## [29] urltools_1.7.3    Rcpp_1.0.10       scales_1.2.1      cachem_1.0.8     
## [33] vroom_1.6.3       jsonlite_1.8.4    bit_4.0.5         hms_1.1.3        
## [37] digest_0.6.31     stringi_1.7.12    bookdown_0.40     grid_4.1.1       
## [41] rprojroot_2.0.4   cli_3.6.1         tools_4.1.1       sass_0.4.6       
## [45] crul_1.5.0        crayon_1.5.2      pkgconfig_2.0.3   timechange_0.3.0 
## [49] rmarkdown_2.21    rstudioapi_0.16.0 R6_2.5.1          compiler_4.1.1</code></pre>
</div>
</div>
<div id="trawl-data" class="section level2 hasAnchor" number="2.4">
<h2><span class="header-section-number">2.4</span> Trawl Data<a href="applications.html#trawl-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One of the more common datasets that can be standardized to Darwin Core and integrated within OBIS is catch data from e.g. a trawl sampling event, or a zooplankton net tow. Of special concern here are datasets that include both a total (species-specific) catch weight, in addition to individual measurements (for a subset of the overall data). In this case, through our standardization to Darwin Core, we want to ensure that data users understand that the individual measurements are a part of, or subset of, the overall (species-specific) record, whilst at the same time ensure that data providers are not duplicating occurrence records to OBIS.</p>
<p>The GitHub issue related to application is can be found <a href="https://github.com/iobis/env-data/issues/10">here</a></p>
<div id="workflow-overview" class="section level3 hasAnchor" number="2.4.1">
<h3><span class="header-section-number">2.4.1</span> Workflow Overview<a href="applications.html#workflow-overview" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In our current setup, this relationship between the overall catch data and subsetted information is provided in the <a href="https://tools.gbif.org/dwca-validator/extension.do?id=dwc:ResourceRelationship">resourceRelationship</a> extension. This extension <em>cannot</em> currently be harvested by GBIF. The required terms for this extension are <code>resourceID</code>, <code>relatedResourceID</code>, <code>resourceRelationshipID</code> and <code>relationshipOfResource</code>. The <code>relatedResourceID</code> here refers to the <em>object</em> of the relationship, whereas the <code>resourceID</code> refers to the <em>subject</em> of the relationship:</p>
<ul>
<li>resourceRelationshipID: a unique identifier for the relationship between one resource (the subject) and another (relatedResource, object).</li>
<li>resourceID: a unique identifier for the resource that is the subject of the relationship.</li>
<li>relatedResourceID: a unique identifier for the resource that is the object of the relationship.</li>
<li>relationshipOfResource: The relationship of the subject (identified by the resourceID) to the object (relatedResourceID). The relationshipOfResource is a free text field.</li>
</ul>
<p>A few resources have been published to OBIS that contain the resourceRelationship extension (examples). Here, I’ll lay out the process and coding used for the <a href="http://ipt.iobis.org/obiscanada/resource?r=trawl-catch-and-species-abundance-from-the-2019-gulf-of-alaska-international-year-of-the-salmon-expedition">Trawl Catch and Species Abundance from the 2019 Gulf of Alaska International Year of the Salmon Expedition</a>. In the following code chunks some details are omitted to improve the readability - the overall code to standardize the catch data can be found <a href="https://github.com/HakaiInstitute/iys-oos/blob/develop/datasets/GoA_2019/Trawl/2019/data_wrangle_trawl2019.Rmd">here</a>. This dataset includes species-specific total catch data at multiple stations (sampling events). From each catch, individual measurements were also taken. Depending on the number of individual caught in the trawl, this was either the total number of species individuals caught, or only a subset (in case of large numbers of individuals caught).</p>
<p>In this specific data record, we created a single Event Core with three extensions: an <em>occurrence</em> extension, <em>measurement or fact</em> extension, and the <em>resourceRelationship</em> extension. However, in this walk-through I’ll only touch on the Event Core, occurrence extension and resourceRelationship extension.</p>
<p>The trawl data is part of a larger project collecting various data types related to salmon ocean ecology. Therefore, in our Event Core we nested information related to the sampling event in the specific layer. (include a visual representation of the schema). Prior to creating the Event Core, we ensured that e.g. dates and times followed the correct ISO-8601 standards, and converted to the correct time zone.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="applications.html#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Time is recorded numerically (1037 instead of 10:37), so need to change these columns:</span></span>
<span id="cb42-2"><a href="applications.html#cb42-2" aria-hidden="true" tabindex="-1"></a>trawl2019<span class="sc">$</span>END_DEPLOYMENT_TIME <span class="ot">&lt;-</span> <span class="fu">substr</span>(<span class="fu">as.POSIXct</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%04.0f&quot;</span>, trawl2019<span class="sc">$</span>END_DEPLOYMENT_TIME), <span class="at">format =</span> <span class="st">&quot;%H%M&quot;</span>), <span class="dv">12</span>, <span class="dv">16</span>)</span>
<span id="cb42-3"><a href="applications.html#cb42-3" aria-hidden="true" tabindex="-1"></a>trawl2019<span class="sc">$</span>BEGIN_RETRIEVAL_TIME <span class="ot">&lt;-</span> <span class="fu">substr</span>(<span class="fu">as.POSIXct</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%04.0f&quot;</span>, trawl2019<span class="sc">$</span>BEGIN_RETRIEVAL_TIME), <span class="at">format =</span> <span class="st">&quot;%H%M&quot;</span>), <span class="dv">12</span>, <span class="dv">16</span>)</span>
<span id="cb42-4"><a href="applications.html#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Additionally, the vessel time is recorded in &#39;Vladivostok&#39; according to the metadata tab. This has to be converted to UTC.  </span></span>
<span id="cb42-5"><a href="applications.html#cb42-5" aria-hidden="true" tabindex="-1"></a>trawl2019 <span class="ot">&lt;-</span> trawl2019 <span class="sc">%&gt;%</span></span>
<span id="cb42-6"><a href="applications.html#cb42-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">eventDate_start =</span> <span class="fu">format_iso_8601</span>(<span class="fu">as.POSIXct</span>(<span class="fu">paste</span>(EVENT_DATE_START, END_DEPLOYMENT_TIME),</span>
<span id="cb42-7"><a href="applications.html#cb42-7" aria-hidden="true" tabindex="-1"></a>                                                      <span class="at">tz =</span> <span class="st">&quot;Asia/Vladivostok&quot;</span>)),</span>
<span id="cb42-8"><a href="applications.html#cb42-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">eventDate_start =</span> <span class="fu">str_replace</span>(eventDate_start, <span class="st">&quot;</span><span class="sc">\\</span><span class="st">+00:00&quot;</span>, <span class="st">&quot;Z&quot;</span>),</span>
<span id="cb42-9"><a href="applications.html#cb42-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">eventDate_finish =</span> <span class="fu">format_iso_8601</span>(<span class="fu">as.POSIXct</span>(<span class="fu">paste</span>(EVENT_DATE_FINISH, BEGIN_RETRIEVAL_TIME),</span>
<span id="cb42-10"><a href="applications.html#cb42-10" aria-hidden="true" tabindex="-1"></a>                                                       <span class="at">tz =</span> <span class="st">&quot;Asia/Vladivostok&quot;</span>)),</span>
<span id="cb42-11"><a href="applications.html#cb42-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">eventDate_finish =</span> <span class="fu">str_replace</span>(eventDate_finish, <span class="st">&quot;</span><span class="sc">\\</span><span class="st">+00:00&quot;</span>, <span class="st">&quot;Z&quot;</span>),</span>
<span id="cb42-12"><a href="applications.html#cb42-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">eventDate =</span> <span class="fu">paste</span>(eventDate_start, eventDate_finish, <span class="at">sep =</span> <span class="st">&quot;/&quot;</span>),</span>
<span id="cb42-13"><a href="applications.html#cb42-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">project =</span> <span class="st">&quot;IYS&quot;</span>,</span>
<span id="cb42-14"><a href="applications.html#cb42-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">cruise =</span> <span class="fu">paste</span>(project, <span class="st">&quot;GoA2019&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>), </span>
<span id="cb42-15"><a href="applications.html#cb42-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">station =</span> <span class="fu">paste</span>(cruise, TOW_NUMBER, <span class="at">sep=</span><span class="st">&quot;:Stn&quot;</span>),</span>
<span id="cb42-16"><a href="applications.html#cb42-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">trawl =</span> <span class="fu">paste</span>(station, <span class="st">&quot;trawl&quot;</span>, <span class="at">sep=</span><span class="st">&quot;:&quot;</span>))</span></code></pre></div>
<p>Then we created the various layers of our Event Core. We created these layers/data frames from two separate datasets that data are pulled from - one dataset that contains the <em>overall</em> catch data, and one dataset that contains the <em>specimen</em> data:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="applications.html#cb43-1" aria-hidden="true" tabindex="-1"></a>trawl2019_allCatch <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="fu">here</span>(<span class="st">&quot;Trawl&quot;</span>, <span class="st">&quot;2019&quot;</span>, <span class="st">&quot;raw_data&quot;</span>, </span>
<span id="cb43-2"><a href="applications.html#cb43-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="st">&quot;2019_GoA_Fish_Trawl_catchdata.xlsx&quot;</span>), <span class="at">sheet =</span> <span class="st">&quot;CATCH_FINAL&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="applications.html#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">project =</span> <span class="st">&quot;IYS&quot;</span>,</span>
<span id="cb43-4"><a href="applications.html#cb43-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">cruise =</span> <span class="fu">paste</span>(project, <span class="st">&quot;GoA2019&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>),</span>
<span id="cb43-5"><a href="applications.html#cb43-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">station =</span> <span class="fu">paste</span>(cruise, <span class="st">`</span><span class="at">TOW_NUMBER (number)</span><span class="st">`</span>, <span class="at">sep =</span> <span class="st">&quot;:Stn&quot;</span>),</span>
<span id="cb43-6"><a href="applications.html#cb43-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">trawl =</span> <span class="fu">paste</span>(station, <span class="st">&quot;trawl&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>))</span>
<span id="cb43-7"><a href="applications.html#cb43-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-8"><a href="applications.html#cb43-8" aria-hidden="true" tabindex="-1"></a>trawl2019_specimen <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="fu">here</span>(<span class="st">&quot;Trawl&quot;</span>, <span class="st">&quot;2019&quot;</span>, <span class="st">&quot;raw_data&quot;</span>, <span class="st">&quot;2019_GoA_Fish_Specimen_data.xlsx&quot;</span>), </span>
<span id="cb43-9"><a href="applications.html#cb43-9" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">sheet =</span> <span class="st">&quot;SPECIMEN_FINAL&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-10"><a href="applications.html#cb43-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">project =</span> <span class="st">&quot;IYS&quot;</span>,</span>
<span id="cb43-11"><a href="applications.html#cb43-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">cruise =</span> <span class="fu">paste</span>(project, <span class="st">&quot;GoA2019&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>),</span>
<span id="cb43-12"><a href="applications.html#cb43-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">station =</span> <span class="fu">paste</span>(cruise, TOW_NUMBER, <span class="at">sep =</span> <span class="st">&quot;:Stn&quot;</span>),</span>
<span id="cb43-13"><a href="applications.html#cb43-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">trawl =</span> <span class="fu">paste</span>(station, <span class="st">&quot;trawl&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>),</span>
<span id="cb43-14"><a href="applications.html#cb43-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">sample =</span> <span class="fu">paste</span>(trawl, <span class="st">&quot;sample&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>),</span>
<span id="cb43-15"><a href="applications.html#cb43-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">sample =</span> <span class="fu">paste</span>(sample, <span class="fu">row_number</span>(), <span class="at">sep =</span> <span class="st">&quot;&quot;</span>))</span></code></pre></div>
<p>Next we created the Event Core, ensuring that we connect the data to the right layer (i.e. date and time should be connected to the layer associated with the sampling event). Please note that because we are creating multiple layers and nesting information, and then at a later stage combining different tables, this results in cells being populated with <code>NA</code>. These have to be removed prior to publishing the Event Core through the IPT.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="applications.html#cb44-1" aria-hidden="true" tabindex="-1"></a>trawl2019_project <span class="ot">&lt;-</span> trawl2019 <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="applications.html#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">eventID =</span> project) <span class="sc">%&gt;%</span></span>
<span id="cb44-3"><a href="applications.html#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(eventID) <span class="sc">%&gt;%</span></span>
<span id="cb44-4"><a href="applications.html#cb44-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">type =</span> <span class="st">&quot;project&quot;</span>)</span>
<span id="cb44-5"><a href="applications.html#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="applications.html#cb44-6" aria-hidden="true" tabindex="-1"></a>trawl2019_cruise <span class="ot">&lt;-</span> trawl2019 <span class="sc">%&gt;%</span> </span>
<span id="cb44-7"><a href="applications.html#cb44-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">eventID =</span> cruise,</span>
<span id="cb44-8"><a href="applications.html#cb44-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">parentEventID =</span> project) <span class="sc">%&gt;%</span> </span>
<span id="cb44-9"><a href="applications.html#cb44-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(eventID, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-10"><a href="applications.html#cb44-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">type =</span> <span class="st">&quot;cruise&quot;</span>)</span>
<span id="cb44-11"><a href="applications.html#cb44-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-12"><a href="applications.html#cb44-12" aria-hidden="true" tabindex="-1"></a>trawl2019_station <span class="ot">&lt;-</span> trawl2019 <span class="sc">%&gt;%</span> </span>
<span id="cb44-13"><a href="applications.html#cb44-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">eventID =</span> station,</span>
<span id="cb44-14"><a href="applications.html#cb44-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">parentEventID =</span> cruise) <span class="sc">%&gt;%</span> </span>
<span id="cb44-15"><a href="applications.html#cb44-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(eventID, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-16"><a href="applications.html#cb44-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">type =</span> <span class="st">&quot;station&quot;</span>)</span>
<span id="cb44-17"><a href="applications.html#cb44-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-18"><a href="applications.html#cb44-18" aria-hidden="true" tabindex="-1"></a><span class="co"># The coordinates associated to the trawl need to be presented in a LINESTRING.</span></span>
<span id="cb44-19"><a href="applications.html#cb44-19" aria-hidden="true" tabindex="-1"></a><span class="co"># END_LONGITUDE_DD needs to be inverted (has to be between -180 and 180, inclusive). </span></span>
<span id="cb44-20"><a href="applications.html#cb44-20" aria-hidden="true" tabindex="-1"></a>trawl2019_coordinates <span class="ot">&lt;-</span> trawl2019 <span class="sc">%&gt;%</span></span>
<span id="cb44-21"><a href="applications.html#cb44-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">eventID =</span> trawl,</span>
<span id="cb44-22"><a href="applications.html#cb44-22" aria-hidden="true" tabindex="-1"></a>         START_LATITUDE_DD,</span>
<span id="cb44-23"><a href="applications.html#cb44-23" aria-hidden="true" tabindex="-1"></a>         longitude,</span>
<span id="cb44-24"><a href="applications.html#cb44-24" aria-hidden="true" tabindex="-1"></a>         END_LATITUDE_DD,</span>
<span id="cb44-25"><a href="applications.html#cb44-25" aria-hidden="true" tabindex="-1"></a>         END_LONGITUDE_DD) <span class="sc">%&gt;%</span></span>
<span id="cb44-26"><a href="applications.html#cb44-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">END_LONGITUDE_DD =</span> END_LONGITUDE_DD <span class="sc">*</span> <span class="sc">-</span><span class="dv">1</span>,</span>
<span id="cb44-27"><a href="applications.html#cb44-27" aria-hidden="true" tabindex="-1"></a>         <span class="at">footprintWKT =</span> <span class="fu">paste</span>(<span class="st">&quot;LINESTRING (&quot;</span>, longitude, START_LATITUDE_DD, <span class="st">&quot;,&quot;</span>, </span>
<span id="cb44-28"><a href="applications.html#cb44-28" aria-hidden="true" tabindex="-1"></a>                              END_LONGITUDE_DD, END_LATITUDE_DD, <span class="st">&quot;)&quot;</span>)) </span>
<span id="cb44-29"><a href="applications.html#cb44-29" aria-hidden="true" tabindex="-1"></a>trawl2019_linestring <span class="ot">&lt;-</span> obistools<span class="sc">::</span><span class="fu">calculate_centroid</span>(trawl2019_coordinates<span class="sc">$</span>footprintWKT)</span>
<span id="cb44-30"><a href="applications.html#cb44-30" aria-hidden="true" tabindex="-1"></a>trawl2019_linestring <span class="ot">&lt;-</span> <span class="fu">cbind</span>(trawl2019_coordinates, trawl2019_linestring) <span class="sc">%&gt;%</span></span>
<span id="cb44-31"><a href="applications.html#cb44-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(eventID, footprintWKT, decimalLatitude, decimalLongitude, coordinateUncertaintyInMeters)</span>
<span id="cb44-32"><a href="applications.html#cb44-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-33"><a href="applications.html#cb44-33" aria-hidden="true" tabindex="-1"></a>trawl2019_trawl <span class="ot">&lt;-</span> trawl2019 <span class="sc">%&gt;%</span> </span>
<span id="cb44-34"><a href="applications.html#cb44-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">eventID =</span> trawl,</span>
<span id="cb44-35"><a href="applications.html#cb44-35" aria-hidden="true" tabindex="-1"></a>         <span class="at">parentEventID =</span> station,</span>
<span id="cb44-36"><a href="applications.html#cb44-36" aria-hidden="true" tabindex="-1"></a>         eventDate,</span>
<span id="cb44-37"><a href="applications.html#cb44-37" aria-hidden="true" tabindex="-1"></a>         year,</span>
<span id="cb44-38"><a href="applications.html#cb44-38" aria-hidden="true" tabindex="-1"></a>         month,</span>
<span id="cb44-39"><a href="applications.html#cb44-39" aria-hidden="true" tabindex="-1"></a>         day) <span class="sc">%&gt;%</span></span>
<span id="cb44-40"><a href="applications.html#cb44-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">minimumDepthInMeters =</span> <span class="dv">0</span>, <span class="co"># headrope was at the surface</span></span>
<span id="cb44-41"><a href="applications.html#cb44-41" aria-hidden="true" tabindex="-1"></a>         <span class="at">maximumDepthInMeters =</span> trawl2019<span class="sc">$</span>MOUTH_OPENING_HEIGHT,</span>
<span id="cb44-42"><a href="applications.html#cb44-42" aria-hidden="true" tabindex="-1"></a>         <span class="at">samplingProtocol =</span> <span class="st">&quot;midwater trawl&quot;</span>, <span class="co"># when available add DOI to paper here</span></span>
<span id="cb44-43"><a href="applications.html#cb44-43" aria-hidden="true" tabindex="-1"></a>         <span class="at">locality =</span> <span class="fu">case_when</span>(</span>
<span id="cb44-44"><a href="applications.html#cb44-44" aria-hidden="true" tabindex="-1"></a>           trawl2019<span class="sc">$</span>EVENT_SUB_TYPE <span class="sc">==</span> <span class="st">&quot;Can EEZ&quot;</span> <span class="sc">~</span> <span class="st">&quot;Canadian EEZ&quot;</span>),</span>
<span id="cb44-45"><a href="applications.html#cb44-45" aria-hidden="true" tabindex="-1"></a>         <span class="at">locationID =</span> <span class="fu">case_when</span>(</span>
<span id="cb44-46"><a href="applications.html#cb44-46" aria-hidden="true" tabindex="-1"></a>           trawl2019<span class="sc">$</span>EVENT_SUB_TYPE <span class="sc">==</span> <span class="st">&quot;Can EEZ&quot;</span> <span class="sc">~</span> <span class="st">&quot;http://marineregions.org/mrgid/8493&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb44-47"><a href="applications.html#cb44-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(trawl2019_linestring, <span class="at">by =</span> <span class="st">&quot;eventID&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb44-48"><a href="applications.html#cb44-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(eventID, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-49"><a href="applications.html#cb44-49" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">type =</span> <span class="st">&quot;midwater trawl&quot;</span>)</span>
<span id="cb44-50"><a href="applications.html#cb44-50" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb44-51"><a href="applications.html#cb44-51" aria-hidden="true" tabindex="-1"></a>trawl2019_sample <span class="ot">&lt;-</span> trawl2019_specimen <span class="sc">%&gt;%</span></span>
<span id="cb44-52"><a href="applications.html#cb44-52" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">eventID =</span> sample,</span>
<span id="cb44-53"><a href="applications.html#cb44-53" aria-hidden="true" tabindex="-1"></a>         <span class="at">parentEventID =</span> trawl) <span class="sc">%&gt;%</span></span>
<span id="cb44-54"><a href="applications.html#cb44-54" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(eventID, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-55"><a href="applications.html#cb44-55" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">type =</span> <span class="st">&quot;individual sample&quot;</span>)</span>
<span id="cb44-56"><a href="applications.html#cb44-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-57"><a href="applications.html#cb44-57" aria-hidden="true" tabindex="-1"></a>trawl2019_event <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(trawl2019_project, </span>
<span id="cb44-58"><a href="applications.html#cb44-58" aria-hidden="true" tabindex="-1"></a>                             trawl2019_cruise,</span>
<span id="cb44-59"><a href="applications.html#cb44-59" aria-hidden="true" tabindex="-1"></a>                             trawl2019_station,</span>
<span id="cb44-60"><a href="applications.html#cb44-60" aria-hidden="true" tabindex="-1"></a>                             trawl2019_trawl,</span>
<span id="cb44-61"><a href="applications.html#cb44-61" aria-hidden="true" tabindex="-1"></a>                             trawl2019_sample) </span>
<span id="cb44-62"><a href="applications.html#cb44-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-63"><a href="applications.html#cb44-63" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove NAs from the Event Core:</span></span>
<span id="cb44-64"><a href="applications.html#cb44-64" aria-hidden="true" tabindex="-1"></a>trawl2019_event <span class="ot">&lt;-</span> <span class="fu">sapply</span>(trawl2019_event, as.character)</span>
<span id="cb44-65"><a href="applications.html#cb44-65" aria-hidden="true" tabindex="-1"></a>trawl2019_event[<span class="fu">is.na</span>(trawl2019_event)] <span class="ot">&lt;-</span> <span class="st">&quot;&quot;</span></span>
<span id="cb44-66"><a href="applications.html#cb44-66" aria-hidden="true" tabindex="-1"></a>trawl2019_event <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(trawl2019_event)</span></code></pre></div>
<p><strong>TO DO: Add visual of e.g. the top 10 rows of the Event Core.</strong></p>
<p>Now that we created the Event Core, we create the occurrence extension. To do this, we create two separate occurrence data tables: one that includes the occurrence data for the <em>total</em> catch, and one data table for the <em>specimen</em> data. Finally, the Occurrence extension is created by combining these two data frames. Personally, I prefer to re-order it so it makes visual sense to me (nest the specimen occurrence records under their respective overall catch data).</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="applications.html#cb45-1" aria-hidden="true" tabindex="-1"></a>trawl2019_allCatch_worms <span class="ot">&lt;-</span> worrms<span class="sc">::</span><span class="fu">wm_records_names</span>(<span class="fu">unique</span>(trawl2019_allCatch<span class="sc">$</span>scientificname))</span>
<span id="cb45-2"><a href="applications.html#cb45-2" aria-hidden="true" tabindex="-1"></a>trawl2019_occ <span class="ot">&lt;-</span> <span class="fu">left_join</span>(trawl2019_allCatch, trawl2019_allCatch_worms, <span class="at">by =</span> <span class="st">&quot;scientificname&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb45-3"><a href="applications.html#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">eventID =</span> trawl,</span>
<span id="cb45-4"><a href="applications.html#cb45-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">specificEpithet =</span> species,</span>
<span id="cb45-5"><a href="applications.html#cb45-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">scientificNameAuthorship =</span> authority,</span>
<span id="cb45-6"><a href="applications.html#cb45-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">taxonomicStatus =</span> status,</span>
<span id="cb45-7"><a href="applications.html#cb45-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">taxonRank =</span> rank,</span>
<span id="cb45-8"><a href="applications.html#cb45-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">scientificName =</span> scientificname,</span>
<span id="cb45-9"><a href="applications.html#cb45-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">scientificNameID =</span> lsid,</span>
<span id="cb45-10"><a href="applications.html#cb45-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">individualCount =</span> <span class="st">`</span><span class="at">CATCH_COUNT (pieces)(**includes Russian expansion for some species)</span><span class="st">`</span>,</span>
<span id="cb45-11"><a href="applications.html#cb45-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceRemarks =</span> COMMENTS) <span class="sc">%&gt;%</span></span>
<span id="cb45-12"><a href="applications.html#cb45-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">occurrenceID =</span> <span class="fu">paste</span>(eventID, <span class="st">&quot;occ&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>),</span>
<span id="cb45-13"><a href="applications.html#cb45-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceID =</span> <span class="fu">paste</span>(occurrenceID, <span class="fu">row_number</span>(), <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>),</span>
<span id="cb45-14"><a href="applications.html#cb45-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceStatus =</span> <span class="st">&quot;present&quot;</span>,</span>
<span id="cb45-15"><a href="applications.html#cb45-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">sex =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb45-16"><a href="applications.html#cb45-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-17"><a href="applications.html#cb45-17" aria-hidden="true" tabindex="-1"></a>trawl2019_catch_ind_worms <span class="ot">&lt;-</span> worrms<span class="sc">::</span><span class="fu">wm_records_names</span>(<span class="fu">unique</span>(trawl2019_catch_ind<span class="sc">$</span>scientificname)) <span class="sc">%&gt;%</span> <span class="fu">bind_rows</span>()</span>
<span id="cb45-18"><a href="applications.html#cb45-18" aria-hidden="true" tabindex="-1"></a>trawl2019_catch_ind_occ <span class="ot">&lt;-</span> <span class="fu">left_join</span>(trawl2019_catch_ind, trawl2019_catch_ind_worms, <span class="at">by =</span> <span class="st">&quot;scientificname&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb45-19"><a href="applications.html#cb45-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">scientificNameAuthorship =</span> authority,</span>
<span id="cb45-20"><a href="applications.html#cb45-20" aria-hidden="true" tabindex="-1"></a>         <span class="at">taxonomicStatus =</span> status,</span>
<span id="cb45-21"><a href="applications.html#cb45-21" aria-hidden="true" tabindex="-1"></a>         <span class="at">taxonRank =</span> rank,</span>
<span id="cb45-22"><a href="applications.html#cb45-22" aria-hidden="true" tabindex="-1"></a>         <span class="at">scientificName =</span> scientificname,</span>
<span id="cb45-23"><a href="applications.html#cb45-23" aria-hidden="true" tabindex="-1"></a>         <span class="at">scientificNameID =</span> lsid) <span class="sc">%&gt;%</span></span>
<span id="cb45-24"><a href="applications.html#cb45-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">occurrenceID =</span> <span class="fu">paste</span>(eventID, <span class="st">&quot;occ&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>),</span>
<span id="cb45-25"><a href="applications.html#cb45-25" aria-hidden="true" tabindex="-1"></a>         <span class="at">occurrenceStatus =</span> <span class="st">&quot;present&quot;</span>,</span>
<span id="cb45-26"><a href="applications.html#cb45-26" aria-hidden="true" tabindex="-1"></a>         <span class="at">individualCount =</span> <span class="dv">1</span>)</span>
<span id="cb45-27"><a href="applications.html#cb45-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-28"><a href="applications.html#cb45-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine the two occurrence data frames:</span></span>
<span id="cb45-29"><a href="applications.html#cb45-29" aria-hidden="true" tabindex="-1"></a>trawl2019_occ_ext <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(trawl2019_occ_fnl, trawl2019_catch_ind_fnl)</span>
<span id="cb45-30"><a href="applications.html#cb45-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-31"><a href="applications.html#cb45-31" aria-hidden="true" tabindex="-1"></a><span class="co"># To re-order the occurrenceID, use following code:</span></span>
<span id="cb45-32"><a href="applications.html#cb45-32" aria-hidden="true" tabindex="-1"></a>order <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_sort</span>(trawl2019_occ_ext<span class="sc">$</span>occurrenceID, <span class="at">numeric=</span><span class="cn">TRUE</span>)</span>
<span id="cb45-33"><a href="applications.html#cb45-33" aria-hidden="true" tabindex="-1"></a>trawl2019_occ_ext <span class="ot">&lt;-</span> trawl2019_occ_ext[<span class="fu">match</span>(order, trawl2019_occ_ext<span class="sc">$</span>occurrenceID),] <span class="sc">%&gt;%</span></span>
<span id="cb45-34"><a href="applications.html#cb45-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">basisOfRecord =</span> <span class="st">&quot;HumanObservation&quot;</span>)</span></code></pre></div>
<p><strong>TO DO: Add visual of e.g. the top 10 rows of the Occurrence extension.</strong></p>
<p><strong>Please note</strong> that in the <em>overall</em> species-specific occurrence data frame, <em>individualCount</em> was not included. This term should not be used for abundance studies, but to avoid confusion and the appearance that the specimen records are an additional observation on top of the overall catch record, the <em>individualCount</em> term was left blank for the overall catch data.</p>
<p>A resource relationship extension is created to further highlight that the individual samples in the occurrence extension are part of a larger overall catch that was also listed in the occurrence extension. In this extension, we wanted to make sure to highlight that the <em>specimen</em> occurrence records are <em>a subset of</em> the <em>overall</em> catch data through the field <code>relationshipOfResource1</code>. Each of these relationships gets a unique <code>resourceRelationshipID</code>.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="applications.html#cb46-1" aria-hidden="true" tabindex="-1"></a>trawl_resourceRelationship <span class="ot">&lt;-</span> trawl2019_occ_ext <span class="sc">%&gt;%</span></span>
<span id="cb46-2"><a href="applications.html#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(eventID, occurrenceID, scientificName) <span class="sc">%&gt;%</span></span>
<span id="cb46-3"><a href="applications.html#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">resourceID =</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">&quot;sample&quot;</span>, trawl2019_occ_ext<span class="sc">$</span>occurrenceID), trawl2019_occ_ext<span class="sc">$</span>occurrenceID, <span class="cn">NA</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb46-4"><a href="applications.html#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">eventID =</span> <span class="fu">gsub</span>(<span class="st">&quot;:sample.*&quot;</span>, <span class="st">&quot;&quot;</span>, trawl2019_occ_ext<span class="sc">$</span>eventID)) <span class="sc">%&gt;%</span></span>
<span id="cb46-5"><a href="applications.html#cb46-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(eventID, scientificName) <span class="sc">%&gt;%</span></span>
<span id="cb46-6"><a href="applications.html#cb46-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">n</span>() <span class="sc">!=</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb46-7"><a href="applications.html#cb46-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>()</span>
<span id="cb46-8"><a href="applications.html#cb46-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-9"><a href="applications.html#cb46-9" aria-hidden="true" tabindex="-1"></a>trawl_resourceRelationship <span class="ot">&lt;-</span> trawl_resourceRelationship <span class="sc">%&gt;%</span></span>
<span id="cb46-10"><a href="applications.html#cb46-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">relatedResourceID =</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">&quot;sample&quot;</span>, trawl_resourceRelationship<span class="sc">$</span>occurrenceID), <span class="cn">NA</span>, trawl_resourceRelationship<span class="sc">$</span>occurrenceID)) <span class="sc">%&gt;%</span></span>
<span id="cb46-11"><a href="applications.html#cb46-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">relationshipOfResource =</span> <span class="fu">ifelse</span>(<span class="sc">!</span><span class="fu">is.na</span>(resourceID), <span class="st">&quot;is a subset of&quot;</span>, <span class="cn">NA</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb46-12"><a href="applications.html#cb46-12" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(eventID, scientificName) <span class="sc">%&gt;%</span></span>
<span id="cb46-13"><a href="applications.html#cb46-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fill</span>(relatedResourceID) <span class="sc">%&gt;%</span></span>
<span id="cb46-14"><a href="applications.html#cb46-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(resourceID))</span>
<span id="cb46-15"><a href="applications.html#cb46-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-16"><a href="applications.html#cb46-16" aria-hidden="true" tabindex="-1"></a>order <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_sort</span>(trawl_resourceRelationship<span class="sc">$</span>resourceID, <span class="at">numeric =</span> <span class="cn">TRUE</span>)</span>
<span id="cb46-17"><a href="applications.html#cb46-17" aria-hidden="true" tabindex="-1"></a>trawl_resourceRelationship <span class="ot">&lt;-</span> trawl_resourceRelationship[<span class="fu">match</span>(order, trawl_resourceRelationship<span class="sc">$</span>resourceID),]</span>
<span id="cb46-18"><a href="applications.html#cb46-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-19"><a href="applications.html#cb46-19" aria-hidden="true" tabindex="-1"></a>trawl_resourceRelationship <span class="ot">&lt;-</span> trawl_resourceRelationship <span class="sc">%&gt;%</span></span>
<span id="cb46-20"><a href="applications.html#cb46-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">resourceRelationshipID =</span> <span class="fu">paste</span>(relatedResourceID, <span class="st">&quot;rr&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>),</span>
<span id="cb46-21"><a href="applications.html#cb46-21" aria-hidden="true" tabindex="-1"></a>         <span class="at">ID =</span> <span class="fu">sprintf</span>(<span class="st">&quot;%03d&quot;</span>, <span class="fu">row_number</span>()),</span>
<span id="cb46-22"><a href="applications.html#cb46-22" aria-hidden="true" tabindex="-1"></a>         <span class="at">resourceRelationshipID =</span> <span class="fu">paste</span>(resourceRelationshipID, ID, <span class="at">sep =</span> <span class="st">&quot;:&quot;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb46-23"><a href="applications.html#cb46-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(eventID, resourceRelationshipID, resourceID, relationshipOfResource, relatedResourceID)</span></code></pre></div>
<p><strong>TO DO: Add visual of e.g. the top 10 rows of the ResourceRelationship extension.</strong></p>
</div>
<div id="faq" class="section level3 hasAnchor" number="2.4.2">
<h3><span class="header-section-number">2.4.2</span> FAQ<a href="applications.html#faq" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Q1</strong>. Why not use the terms <em>associatedOccurrence</em> or <em>associatedTaxa</em>?
<strong>A</strong>. There seems to be a movement away from the term <em>associatedOccurrence</em> as the <code>resourceRelationship</code> extension has a much broader use case. Some issues that were raised on GitHub exemplify this, see e.g. <a href="https://github.com/tdwg/dwc/issues/324">here</a>. <em>associatedTaxa</em> is used to provide identifiers or names of taxa and the associations of an Occurrence with them. This term is not apt for establishing relationships between taxa, only between specific Occurrences of an organism with other taxa. As noted on the <a href="https://dwc.tdwg.org/terms/#dwc:associatedTaxa">TDWG website</a>, <em>[…] Note that the ResourceRelationship class is an alternative means of representing associations, and with more detail.</em> See also e.g. <a href="https://github.com/tdwg/dwc/issues/331">this issue</a>.</p>
</div>
</div>
<div id="dataset-edna" class="section level2 hasAnchor" number="2.5">
<h2><span class="header-section-number">2.5</span> dataset-edna<a href="applications.html#dataset-edna" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>By <a href="https://github.com/dianalg">Diana LaScala-Gruenewald</a></p>
<p><a href="https://mybinder.org/v2/gh/iobis/dataset-edna/HEAD?labpath=src%2Fconversion_code.ipynb"><img src="https://mybinder.org/badge_logo.svg" alt="Binder" /></a></p>
<div id="introduction" class="section level3 hasAnchor" number="2.5.1">
<h3><span class="header-section-number">2.5.1</span> Introduction<a href="applications.html#introduction" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Rationale:</strong></p>
<p>DNA derived data are increasingly being used to document taxon
occurrences. To ensure these data are useful to the broadest possible
community, <a href="https://www.gbif.org/">GBIF</a> published a guide entitled “<a href="https://docs.gbif-uat.org/publishing-dna-derived-data/1.0/en/">Publishing DNA-derived
data through biodiversity data platforms</a>.”
This guide is supported by the <a href="https://tools.gbif.org/dwca-validator/extension.do?id=http://rs.gbif.org/terms/1.0/DNADerivedData">DNA derived data extension</a>
for <a href="https://dwc.tdwg.org/">Darwin Core</a>, which incorporates <a href="https://gensc.org/mixs/">MIxS</a>
terms into the Darwin Core standard.</p>
<p>This use case draws on both the guide and the extension to illustrate
how to incorporate a DNA derived data extension file into a Darwin Core
archive.</p>
<p>For further information on this use case and the DNA Derived data extension
in general, see the recording of the <a href="https://obis.org/2021/10/13/gendatawebinar/">OBIS Webinar on Genetic Data</a>.</p>
<p><strong>Project abstract:</strong></p>
<p>The example data employed in this use case are from marine
filtered seawater samples collected at a nearshore station in
Monterey Bay, California, USA. They were collected by CTD
rosette and filtered by a peristaltic pump system. Subsequently,
they underwent metabarcoding for the 18S V9 region. The resulting
ASVs, their assigned taxonomy, and the metadata associated with their
collection are the input data for the conversion scripts
presented here.</p>
<p>A selection of samples from this collection were included in the
publication “<a href="https://www.nature.com/articles/s41467-019-14105-1">Environmental DNA reveals seasonal shifts and potential
interactions in a marine community</a>”
which was published with open access in <em>Nature Communications</em> in 2020.</p>
<p><strong>Contacts:</strong>
- Francisco Chavez - Principle Investigator (<a href="chfr@mbari.org">chfr@mbari.org</a>)
- Kathleen Pitz - Research Associate (<a href="kpitz@mbari.org">kpitz@mbari.org</a>)
- Diana LaScala-Gruenewald - Point of Contact (<a href="dianalg@mbari.org">dianalg@mbari.org</a>)</p>
</div>
<div id="published-data" class="section level3 hasAnchor" number="2.5.2">
<h3><span class="header-section-number">2.5.2</span> Published data<a href="applications.html#published-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li><a href="https://www.gbif.org/dataset/e0b59ee7-19ae-4eb0-9217-33317fb50d47">GBIF</a></li>
<li><a href="https://obis.org/dataset/62b97724-da17-4ca7-9b26-b2a22aeaab51">OBIS</a></li>
</ul>
</div>
<div id="repo-structure" class="section level3 hasAnchor" number="2.5.3">
<h3><span class="header-section-number">2.5.3</span> Repo structure<a href="applications.html#repo-structure" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<pre><code>.
+-- README.md                   :Description of this repository
+-- LICENSE                     :Repository license
+-- .gitignore                  :Files and directories to be ignored by git
+-- environment.yml             :Conda environment configuration file for Binder
|
+-- raw
|   +-- asv_table.csv           :Source data containing ASV sequences and number of reads
|   +-- taxa_table.csv          :Source data containing taxon matches for each ASV
|   +-- metadata_table.csv      :Source data containing metadata about samples (e.g. collection information)
|
+-- src
|   +-- conversion_code.py      :Darwin Core mapping script
|   +-- conversion_code.ipynb   :Darwin Core mapping Jupyter Notebook
|   +-- WoRMS.py                :Functions for querying the World Register of Marine Species
|
+-- processed
|   +-- occurrence.csv          :Occurrence file, generated by conversion_code
|   +-- dna_extension.csv       :DNA Derived Data Extension file, generated by conversion_code
</code></pre>
</div>
</div>
<div id="converting-atn-netcdf-file-to-dawrin-core" class="section level2 hasAnchor" number="2.6">
<h2><span class="header-section-number">2.6</span> Converting ATN netCDF file to Dawrin Core<a href="applications.html#converting-atn-netcdf-file-to-dawrin-core" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Created: 2022-03-23
Updated: 2023-11-16</p>
<p>Credit: Stephen Formel, Mathew Biddle</p>
<p>This notebook walks through downloading an example netCDF file from the an Archive package at NCEI and translating it to a Darwin Core Archive compliant package for easy loading and publishing via the Integrated Publishing Toolkit (IPT). The example file follows a specific specification for ATN satellite trajectory observations as documented <a href="https://github.com/ioos/ioos-atn-data/blob/main/templates/atn_trajectory_template.cdl">here</a>. More information about the ATN netCDF specification can be found in the repository <a href="https://github.com/ioos/ioos-atn-data" class="uri">https://github.com/ioos/ioos-atn-data</a>.</p>
<p>This example uses the <a href="https://ropensci.org/blog/2019/11/05/tidync/">tidync</a> package to work with netCDF data.</p>
<p>Data used in this notebook are available from NCEI at the following link <a href="https://www.ncei.noaa.gov/archive/accession/0282699" class="uri">https://www.ncei.noaa.gov/archive/accession/0282699</a>.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="applications.html#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Load libraries</span></span>
<span id="cb48-2"><a href="applications.html#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="applications.html#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidync)</span>
<span id="cb48-4"><a href="applications.html#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(obistools)</span>
<span id="cb48-5"><a href="applications.html#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ncdf4)</span>
<span id="cb48-6"><a href="applications.html#cb48-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse) <span class="co">#includes stringr</span></span>
<span id="cb48-7"><a href="applications.html#cb48-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lubridate)</span>
<span id="cb48-8"><a href="applications.html#cb48-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(maps)</span>
<span id="cb48-9"><a href="applications.html#cb48-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mapdata)</span></code></pre></div>
<div id="downloading-and-preprocessing-the-source-data" class="section level3 hasAnchor" number="2.6.1">
<h3><span class="header-section-number">2.6.1</span> Downloading and preprocessing the source data<a href="applications.html#downloading-and-preprocessing-the-source-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>See <a href="https://www.ncei.noaa.gov/archive/accession/0282699" class="uri">https://www.ncei.noaa.gov/archive/accession/0282699</a></p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="applications.html#cb49-1" aria-hidden="true" tabindex="-1"></a>url <span class="ot">=</span> <span class="st">&#39;https://www.nodc.noaa.gov/archive/arc0217/0282699/1.1/data/0-data/atn_45866_great-white-shark_trajectory_20090923-20091123.nc&#39;</span></span>
<span id="cb49-2"><a href="applications.html#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="applications.html#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="co">#fname = str_split_i(url, &quot;/&quot;, i=-1)</span></span>
<span id="cb49-4"><a href="applications.html#cb49-4" aria-hidden="true" tabindex="-1"></a>fname <span class="ot">=</span> <span class="fu">paste</span>(<span class="st">&#39;data/src/&#39;</span>,<span class="fu">basename</span>(url))</span>
<span id="cb49-5"><a href="applications.html#cb49-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-6"><a href="applications.html#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(url, fname, <span class="at">mode =</span> <span class="st">&quot;wb&quot;</span>)</span></code></pre></div>
<div id="open-the-netcdf-file" class="section level4 hasAnchor" number="2.6.1.1">
<h4><span class="header-section-number">2.6.1.1</span> Open the netCDF file<a href="applications.html#open-the-netcdf-file" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Once the file is opened, we print out the details of what the netCDF file contains.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="applications.html#cb50-1" aria-hidden="true" tabindex="-1"></a>atn <span class="ot">&lt;-</span> <span class="fu">nc_open</span>(fname)</span>
<span id="cb50-2"><a href="applications.html#cb50-2" aria-hidden="true" tabindex="-1"></a>atn</span></code></pre></div>
<pre><code>## File data/src/ atn_45866_great-white-shark_trajectory_20090923-20091123.nc (NC_FORMAT_NETCDF4):
## 
##      36 variables (excluding dimension variables):
##         string deploy_id[]   (Contiguous storage)  
##             long_name: id for this deployment. This is typically the tag ptt
##             comment: Friendly name given to the tag by the user. If no specific friendly name is given, this is the PTT id.
##             coordinates: time z lon lat
##             instrument: instrument_location
##             platform: animal
##             coverage_content_type: referenceInformation
##             _FillValue: -9999
##         double time[obs]   (Contiguous storage)  
##             units: seconds since 1990-01-01 00:00:00Z
##             standard_name: time
##             axis: T
##             _CoordinateAxisType: Time
##             calendar: standard
##             long_name: Time of the measurement, in seconds since 1990-01-01
##             actual_min: 2009-09-23T00:00:00Z
##             actual_max: 2009-11-23T05:12:00Z
##             ancillary_variables: qartod_time_flag qartod_rollup_flag qartod_speed_flag
##             instrument: instrument_location
##             platform: animal
##             coverage_content_type: coordinate
##             _FillValue: NaN
##         int z[obs]   (Contiguous storage)  
##             _FillValue: -9999
##             axis: Z
##             long_name: depth of measurement
##             positive: down
##             standard_name: depth
##             units: m
##             actual_min: 0
##             actual_max: 0
##             instrument: 
##             platform: animal
##             comment: This variable is synthetically generated to represent the depth of observations
##             coverage_content_type: coordinate
##         double lat[obs]   (Contiguous storage)  
##             axis: Y
##             _CoordinateAxisType: Lat
##             long_name: Latitude portion of location in decimal degrees North
##             standard_name: latitude
##             units: degrees_north
##             valid_max: 90
##             valid_min: -90
##             actual_min: 23.59
##             actual_max: 34.045
##             ancillary_variables: qartod_location_flag qartod_rollup_flag qartod_speed_flag error_radius semi_major_axis semi_minor_axis ellipse_orientation offset offset_orientation
##             instrument: instrument_location
##             platform: animal
##             coverage_content_type: coordinate
##             _FillValue: NaN
##         double lon[obs]   (Contiguous storage)  
##             axis: X
##             _CoordinateAxisType: Lon
##             long_name: Longitude portion of location in decimal degrees East
##             standard_name: longitude
##             units: degrees_east
##             valid_max: 180
##             valid_min: -180
##             actual_min: -166.18
##             actual_max: -118.504
##             ancillary_variables: qartod_location_flag qartod_rollup_flag qartod_speed_flag error_radius semi_major_axis semi_minor_axis ellipse_orientation offset offset_orientation
##             instrument: instrument_location
##             platform: animal
##             coverage_content_type: coordinate
##             _FillValue: NaN
##         int ptt[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: -9999
##             coordinates: time z lon lat
##             long_name: Platform Transmitter Terminal (PTT) id used for Argos transmissions
##             comment: PTT id for this deployment. PTT ids may be used on multiple deployments, but not concurrently. When combined with deployment dates, PTTs can uniquely identify a deployment.
##             coverage_content_type: referenceInformation
##             instrument: instrument_location
##             platform: animal
##         string instrument[obs]   (Contiguous storage)  
##             coordinates: time z lon lat
##             comment: Wildlife Computers instrument family. Variable may report manufacturer default values (e.g., Mk10) and may not match correctly defined instrument_location or instrument_tag variables and attributes.
##             long_name: Instrument family
##             instrument: instrument_location
##             platform: animal
##             coverage_content_type: referenceInformation
##         string type[obs]   (Contiguous storage)  
##             coordinates: time z lon lat
##             comment: Type of location: Argos, FastGPS or User
##             long_name: Type of location information - Argos, GPS satellite or user provided location
##             instrument: instrument_location
##             platform: animal
##             coverage_content_type: referenceInformation
##         string location_class[obs]   (Contiguous storage)  
##             coordinates: time z lon lat
##             standard_name: quality_flag
##             comment: Quality codes from the ARGOS satellite (in meters): G,3,2,1,0,A,B,Z. See http://www.argos-system.org/manual/3-location/34_location_classes.htm
##             long_name: Location Quality Code from ARGOS satellite system
##             code_values: G,3,2,1,0,A,B,Z
##             code_meanings: estimated error less than 100m and 1+ messages received per satellite pass, estimated error less than 250m and 4+ messages received per satellite pass, estimated error between 250m and 500m and 4+ messages per satellite pass, estimated error between 500m and 1500m and 4+ messages per satellite pass, estimated error greater than 1500m and 4+ messages received per satellite pass, no least squares estimated error or unbounded kalman filter estimated error and 3 messages received per satellite pass, no least squares estimated error or unbounded kalman filter estimated error and 1 or 2 messages received per satellite pass, invalid location (available for Service Plus or Auxilliary Location Processing)
##             instrument: instrument_location
##             platform: animal
##             ancillary_variables: lat lon
##             coverage_content_type: qualityInformation
##         int error_radius[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: -9999
##             coordinates: time z lon lat
##             long_name: Error radius
##             units: m
##             comment: If the position is best represented as a circle, this field gives the radius of that circle in meters.
##             instrument: instrument_location
##             platform: animal
##             ancillary_variables: lat lon offset offset_orientation
##             coverage_content_type: qualityInformation
##         int semi_major_axis[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: -9999
##             coordinates: time z lon lat
##             long_name: Error - ellipse semi-major axis
##             units: m
##             comment: If the estimated position error is best expressed as an ellipse, this field gives the length in meters of the semi-major elliptical axis (one half of the major axis).
##             instrument: instrument_location
##             platform: animal
##             ancillary_variables: lat lon ellipse_orientation offset offset_orientation
##             coverage_content_type: qualityInformation
##         int semi_minor_axis[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: -9999
##             coordinates: time z lon lat
##             long_name: Error - ellipse semi-minor axis
##             units: m
##             comment: If the estimated position error is best expressed as an ellipse, this field gives the length in meters of the semi-minor elliptical axis (one half of the minor axis).
##             instrument: instrument_location
##             platform: animal
##             ancillary_variables: lat lon ellipse_orientation offset offset_orientation
##             coverage_content_type: qualityInformation
##         int ellipse_orientation[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: -9999
##             coordinates: time z lon lat
##             long_name: Error - ellipse orientation in degrees clockwise from true north
##             units: degrees
##             comment: The angle in degrees of the ellipse from true north, proceeding clockwise (0 to 360). A blank field represents 0 degrees.
##             instrument: instrument_location
##             platform: animal
##             ancillary_variables: lat lon semi_major_axis semi_minor_axis offset offset_orientation
##             coverage_content_type: qualityInformation
##         int offset[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: -9999
##             coordinates: time z lon lat
##             long_name: Error - offset in meters to center of error ellipse or circle
##             units: m
##             comment: This field is non-zero if the circle or ellipse are not centered on the (Latitude, Longitude) values on this row. &quot;Offset&quot; gives the distance in meters from (Latitude, Longitude) to the center of the ellipse.
##             instrument: instrument_location
##             platform: animal
##             ancillary_variables: lat lon error_radius semi_major_axis semi_minor_axis offset_orientation
##             coverage_content_type: qualityInformation
##         int offset_orientation[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: -9999
##             coordinates: time z lon lat
##             long_name: Error - offset orientation angle to ellipse center
##             units: degrees
##             comment: If the &quot;Offset&quot; field is non-zero, this field is the angle in degrees from (Latitude, Longitude) to the center of the ellipse. Zero degrees is true north; a blank field represents 0 degrees.
##             instrument: instrument_location
##             platform: animal
##             ancillary_variables: lat lon error_radius semi_major_axis semi_minor_axis offset
##             coverage_content_type: qualityInformation
##         double gpe_msd[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             coordinates: time z lon lat
##             comment: Historical. No longer applicable.
##             long_name: 
##             units: 
##             instrument: instrument_location
##             platform: animal
##             coverage_content_type: auxillaryInformation
##             _FillValue: NaN
##         double gpe_u[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             coordinates: time z lon lat
##             comment: Historical. No longer applicable.
##             long_name: 
##             units: 
##             instrument: instrument_location
##             platform: animal
##             coverage_content_type: auxillaryInformation
##             _FillValue: NaN
##         int count[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: -9999
##             coordinates: time z lon lat
##             comment: Total number of times a particular data item was received, verified, and successfully decoded.
##             long_name: Count
##             units: count
##             instrument: instrument_location
##             platform: animal
##             coverage_content_type: auxillaryInformation
##         unsigned byte qartod_time_flag[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: 241
##             coordinates: time z lon lat
##             standard_name: gross_range_test_quality_flag
##             long_name: Time QC test - gross range test
##             implementation: https://github.com/ioos/ioos_qc/
##             flag_meanings: PASS NOT_EVALUATED SUSPECT FAIL MISSING
##             flag_values: 1
##              flag_values: 2
##              flag_values: 3
##              flag_values: 4
##              flag_values: 9
##             references: https://cdn.ioos.noaa.gov/media/2020/03/QARTOD_TS_Manual_Update2_200324_final.pdf
##             coverage_content_type: qualityInformation
##         unsigned byte qartod_speed_flag[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: 241
##             coordinates: time z lon lat
##             standard_name: gross_range_test_quality_flag
##             long_name: Speed QC test - gross range test
##             references: https://cdn.ioos.noaa.gov/media/2020/03/QARTOD_TS_Manual_Update2_200324_final.pdf
##             implementation: https://github.com/ioos/ioos_qc/
##             flag_meanings: PASS NOT_EVALUATED SUSPECT FAIL MISSING
##             flag_values: 1
##              flag_values: 2
##              flag_values: 3
##              flag_values: 4
##              flag_values: 9
##             coverage_content_type: qualityInformation
##         unsigned byte qartod_location_flag[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: 241
##             coordinates: time z lon lat
##             standard_name: location_test_quality_flag
##             long_name: Location QC test - Location test
##             implementation: https://github.com/ioos/ioos_qc/
##             flag_meanings: PASS NOT_EVALUATED SUSPECT FAIL MISSING
##             flag_values: 1
##              flag_values: 2
##              flag_values: 3
##              flag_values: 4
##              flag_values: 9
##             references: https://cdn.ioos.noaa.gov/media/2020/03/QARTOD_TS_Manual_Update2_200324_final.pdf
##             coverage_content_type: qualityInformation
##         unsigned byte qartod_rollup_flag[obs]   (Chunking: [29])  (Compression: shuffle,level 1)
##             _FillValue: 241
##             coordinates: time z lon lat
##             standard_name: aggregate_quality_flag
##             long_name: Aggregate QC value
##             implementation: https://github.com/ioos/ioos_qc/
##             flag_meanings: PASS NOT_EVALUATED SUSPECT FAIL MISSING
##             flag_values: 1
##              flag_values: 2
##              flag_values: 3
##              flag_values: 4
##              flag_values: 9
##             references: https://cdn.ioos.noaa.gov/media/2020/03/QARTOD_TS_Manual_Update2_200324_final.pdf
##             coverage_content_type: qualityInformation
##         int crs[]   (Contiguous storage)  
##             epsg_code: EPSG:4326
##             grid_mapping_name: latitude_longitude
##             inverse_flattening: 298.257223563
##             long_name: Coordinate Reference System - http://www.opengis.net/def/crs/EPSG/0/4326
##             semi_major_axis: 6378137
##             coverage_content_type: referenceInformation
##         string trajectory[]   (Contiguous storage)  
##             cf_role: trajectory_id
##             long_name: trajectory identifier
##         int animal_age[]   (Contiguous storage)  
##             _FillValue: -9999
##             units: 
##             long_name: age of the animal as measured or estimated at deployment
##             coverage_content_type: referenceInformation
##             animal_age: Not provided
##         string animal_life_stage[]   (Contiguous storage)  
##             animal_life_stage: juvenile
##             long_name: Lifestage of the animal at time of deployment 
##             coverage_content_type: referenceInformation
##         string animal_sex[]   (Contiguous storage)  
##             animal_sex: male
##             long_name: sex of the animal at time of tag deployment
##             coverage_content_type: referenceInformation
##         float animal_weight[]   (Contiguous storage)  
##             _FillValue: NaN
##             units: kg
##             long_name: mass of the animal as measured or estimated at deployment
##             animal_weight: Not provided
##             coverage_content_type: referenceInformation
##         float animal_length[]   (Contiguous storage)  
##             _FillValue: NaN
##             animal_length_type: total length
##             units: cm
##             animal_length: 213.0 (cm) total length
##             long_name: length of the animal as measured or estimated at deployment
##             coverage_content_type: referenceInformation
##         float animal_length_2[]   (Contiguous storage)  
##             _FillValue: NaN
##             animal_length_2_type: Not provided
##             units: 
##             animal_length_2: Not provided
##             long_name: length of the animal as measured or estimated at deployment
##             coverage_content_type: referenceInformation
##         string animal[]   (Contiguous storage)  
##             rank: Species
##             infraorder: 
##             scientificname: Carcharodon carcharias
##             long_name: tagged animal id
##             superdomain: Biota
##             order: Lamniformes
##             authority: (Linnaeus, 1758)
##             kingdom: Animalia
##             species: Carcharodon carcharias
##             genus: Carcharodon
##             megaclass: 
##             family: Lamnidae
##             taxonRankID: 220
##             class: Elasmobranchii
##             cf_role: trajectory_id
##             coverage_content_type: referenceInformation
##             subphylum: Vertebrata
##             phylum: Chordata
##             AphiaID: 105838
##             valid_name: Carcharodon carcharias
##             infraphylum: Gnathostomata
##             subclass: Neoselachii
##             suborder: 
##         string instrument_tag[]   (Contiguous storage)  
##             manufacturer: Wildlife Computers
##             make_model: SPOT5
##             serial_number: 07S0230
##             long_name: telemetry tag applied to animal
##             coverage_content_type: referenceInformation
##             calibration_date: Not Provided
##         string instrument_location[]   (Contiguous storage)  
##             manufacturer: Wildlife Computers
##             make_model: SPOT5
##             serial_number: 07S0230
##             long_name: Wildlife Computers SPOT5
##             location_type: argos / modeled
##             comment: Location
##             coverage_content_type: referenceInformation
##             calibration_date: Not Provided
##         string taxon_name[]   (Contiguous storage)  
##             standard_name: biological_taxon_name
##             long_name: most precise taxonomic classification for the tagged animal
##             coverage_content_type: referenceInformation
##             source: Froese, R. and D. Pauly. Editors. (2023). FishBase. Carcharodon carcharias (Linnaeus, 1758). Accessed through: World Register of Marine Species at: https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=105838 on 2023-08-16
##             url: https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=105838
##         string taxon_lsid[]   (Contiguous storage)  
##             standard_name: biological_taxon_lsid
##             long_name: Namespaced Taxon Identifier for the tagged animal
##             coverage_content_type: referenceInformation
##             source: Froese, R. and D. Pauly. Editors. (2023). FishBase. Carcharodon carcharias (Linnaeus, 1758). Accessed through: World Register of Marine Species at: https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=105838 on 2023-08-16
##             url: https://www.marinespecies.org/aphia.php?p=taxdetails&amp;id=105838
##         string comment[obs]   (Contiguous storage)  
##             long_name: Comment
##             comment: Optional text field
##             coordinates: time z lon lat
##             instrument: instrument_location
##             platform: animal
##             coverage_content_type: auxillaryInformation
## 
##      1 dimensions:
##         obs  Size:29 (no dimvar)
## 
##     89 global attributes:
##         date_created: 2023-08-16T20:00:00Z
##         featureType: trajectory
##         cdm_data_type: Trajectory
##         Conventions: CF-1.10, ACDD-1.3, IOOS-1.2
##         argos_program_number: 2414
##         creator_email: chris.lowe@csulb.edu
##         id: 5f0668a86321be13bc7ef628
##         tag_type: SPOT5
##         source: Service Argos
##         acknowledgement: NOAA IOOS, Axiom Data Science, Navy ONR, NOAA NMFS, Wildlife Computers, Argos, IOOS ATN
##         creator_name: Chris G. Lowe
##         creator_url: 
##         geospatial_lat_units: degrees_north
##         geospatial_lon_units: degrees_east
##         infoUrl: https://portal.atn.ioos.us/#metadata/6e2ba85c-2f61-4bc5-8c2b-34d6734155ed/project
##         institution: California State University Long Beach
##         keywords: EARTH SCIENCE &gt; AGRICULTURE &gt; ANIMAL SCIENCE &gt; ANIMAL ECOLOGY AND BEHAVIOR, EARTH SCIENCE &gt; BIOSPHERE &gt; ECOLOGICAL DYNAMICS &gt; SPECIES/POPULATION INTERACTIONS &gt; MIGRATORY RATES/ROUTES, EARTH SCIENCE &gt; OCEANS, EARTH SCIENCE &gt; CLIMATE INDICATORS &gt; BIOSPHERIC INDICATORS &gt; SPECIES MIGRATION, EARTH SCIENCE &gt; OCEANS, EARTH SCIENCE &gt; BIOLOGICAL CLASSIFICATION &gt; ANIMALS/VERTEBRATES, EARTH SCIENCE &gt; BIOSPHERE &gt; ECOSYSTEMS &gt; MARINE ECOSYSTEMS, PROVIDERS &gt; GOVERNMENT AGENCIES-U.S. FEDERAL AGENCIES &gt; DOC &gt; NOAA &gt; IOOS, PROVIDERS &gt; COMMERCIAL &gt; Axiom Data Science
##         license: These data may be used and redistributed for free, but are not intended for legal use, since they may contain inaccuracies. No person or group associated with these data makes any warranty, expressed or implied, including warranties of merchantability and fitness for a particular purpose, or assumes any legal liability for the accuracy, completeness or usefulness of this information. This disclaimer applies to both individual use of these data and aggregate use with other data. It is strongly recommended that users read and fully comprehend associated metadata prior to use. Please acknowledge the U.S. Animal Telemetry Network (ATN) or the specified citation as the source from which these data were obtained in any publications and/or representations of these data. Communication and collaboration with dataset authors are strongly encouraged.
##         metadata_link: 
##         naming_authority: com.wildlifecomputers
##         platform_category: animal
##         platform: fish
##         platform_vocabulary: https://vocab.nerc.ac.uk/collection/L06/current/
##         processing_level: NetCDF file created from position data obtained from Wildlife Computers API.
##         project: Project White Shark: Juvenile Satellite Biotelemetry, 2001-2020
##         publisher_email: atndata@ioos.us
##         publisher_institution: US Integrated Ocean Observing System Office
##         publisher_name: US Integrated Ocean Observing System (IOOS) Animal Telemetry Network (ATN)
##         publisher_url: https://atn.ioos.us
##         publisher_country: USA
##         standard_name_vocabulary: CF-v78
##         vendor: Wildlife Computers
##         geospatial_lat_min: 23.59
##         geospatial_lat_max: 34.045
##         geospatial_lon_min: -166.18
##         geospatial_lon_max: -118.504
##         geospatial_bbox: POLYGON ((-118.504 23.59, -118.504 34.045, -166.18 34.045, -166.18 23.59, -118.504 23.59))
##         geospatial_bounds: POLYGON ((-166.18 23.59, -118.581 34.038, -118.53 34.045, -118.504 33.989, -118.534 33.972, -119.75 33.517, -166.18 23.59))
##         geospatial_bounds_crs: EPSG:4326
##         time_coverage_start: 2009-09-23T00:00:00Z
##         time_coverage_end: 2009-11-23T05:12:00Z
##         time_coverage_duration: P61DT5H12M0S
##         time_coverage_resolution: P2DT2H39M43S
##         date_issued: 2023-08-16T20:00:00Z
##         date_modified: 2023-08-16T20:00:00Z
##         history: 2023-08-07T20:24:04Z - Created by the IOOS ATN DAC from the Wildlife Computers API
##         summary: Wildlife Computers SPOT5 tag (ptt id 45866) deployed on a great white shark (Carcharodon carcharias) by Chris G. Lowe in the North Pacific Ocean from 2009-09-23 to 2009-11-23
##         title: Great white shark (Carcharodon carcharias) location data from a satellite telemetry tag (ptt id 45866) deployed in the North Pacific Ocean from 2009-09-23 to 2009-11-23, deployment id 5f0668a86321be13bc7ef628
##         uuid: ff554ebf-bf4b-5a82-8a90-9c0ceb799d96
##         platform_name: Carcharodon carcharias
##         platform_id: 105838
##         vendor_id: 5f0668a86321be13bc7ef628
##         sea_name: North Pacific Ocean
##         arbitrary_keywords: ATN, Animal Telemetry Network, IOOS, Integrated Ocean Observing System, trajectory, satellite telemetry tag
##         contributor_role_vocabulary: https://vocab.nerc.ac.uk/collection/G04/current/
##         creator_role_vocabulary: https://vocab.nerc.ac.uk/collection/G04/current/
##         creator_sector_vocabulary: https://mmisw.org/ont/ioos/sector
##         creator_type: person
##         date_metadata_modified: 20230816
##         instrument: Satellite telemetry tag
##         instrument_vocabulary: 
##         keywords_vocabulary: GCMD Science Keywords v15.1
##         ncei_template_version: NCEI_NetCDF_Trajectory_Template_v2.0
##         product_version: 
##         program: IOOS Animal Telemetry Network
##         publisher_type: institution
##         references: 
##         animal_common_name: great white shark
##         animal_id: 09_13
##         animal_scientific_name: Carcharodon carcharias
##         deployment_id: 5f0668a86321be13bc7ef628
##         deployment_start_datetime: 2009-09-23T00:00:00Z
##         deployment_end_datetime: 2009-11-23T00:00:00Z
##         wmo_platform_code: 
##         comment: 09_13-45866
##         ptt_id: 45866
##         deployment_start_lat: 34.03
##         deployment_start_lon: -118.56
##         contributor_name: Thomas Farrugia
##         contributor_email: tjfarrugia@alaska.edu
##         contributor_role: collaborator
##         contributor_institution: California State University Long Beach
##         contributor_url: 
##         creator_role: principalInvestigator
##         creator_sector: academic
##         creator_country: USA
##         creator_institution: California State University Long Beach
##         creator_institution_url: https://www.csulb.edu/shark-lab
##         citation: Lowe, Chris G.; Farrugia, Thomas. (2023) great white shark (Carcharodon carcharias) location data from a satellite telemetry tag (ptt id 45866) deployed in the North Pacific Ocean from 2009-09-23 to 2009-11-23, deployment id 5f0668a86321be13bc7ef628. [Dataset]. US Integrated Ocean Observing System Office.</code></pre>
</div>
<div id="collect-all-the-metadata-from-the-netcdf-file." class="section level4 hasAnchor" number="2.6.1.2">
<h4><span class="header-section-number">2.6.1.2</span> Collect all the metadata from the netCDF file.<a href="applications.html#collect-all-the-metadata-from-the-netcdf-file." class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>This gathers not only the global attributes, but the variable level attributes as well. As you can see in the <strong>variable</strong> column the term <code>NC_GLOBAL</code> refers to global attributes.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="applications.html#cb52-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> ncmeta<span class="sc">::</span><span class="fu">nc_atts</span>(fname)</span>
<span id="cb52-2"><a href="applications.html#cb52-2" aria-hidden="true" tabindex="-1"></a>metadata</span></code></pre></div>
<pre><code>## # A tibble: 381 × 4
##       id name                  variable  value       
##    &lt;int&gt; &lt;chr&gt;                 &lt;chr&gt;     &lt;named list&gt;
##  1     0 long_name             deploy_id &lt;chr [1]&gt;   
##  2     1 comment               deploy_id &lt;chr [1]&gt;   
##  3     2 coordinates           deploy_id &lt;chr [1]&gt;   
##  4     3 instrument            deploy_id &lt;chr [1]&gt;   
##  5     4 platform              deploy_id &lt;chr [1]&gt;   
##  6     5 coverage_content_type deploy_id &lt;chr [1]&gt;   
##  7     6 _FillValue            deploy_id &lt;dbl [1]&gt;   
##  8     0 units                 time      &lt;chr [1]&gt;   
##  9     1 standard_name         time      &lt;chr [1]&gt;   
## 10     2 axis                  time      &lt;chr [1]&gt;   
## # ℹ 371 more rows</code></pre>
</div>
<div id="store-the-data-as-a-tibble" class="section level4 hasAnchor" number="2.6.1.3">
<h4><span class="header-section-number">2.6.1.3</span> Store the data as a tibble<a href="applications.html#store-the-data-as-a-tibble" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Collect the data dimensioned by <code>time</code> from the netCDF file as a tibble. Then, print the first ten rows.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="applications.html#cb54-1" aria-hidden="true" tabindex="-1"></a>atn <span class="ot">&lt;-</span> <span class="fu">tidync</span>(fname)</span>
<span id="cb54-2"><a href="applications.html#cb54-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-3"><a href="applications.html#cb54-3" aria-hidden="true" tabindex="-1"></a>atn_tbl <span class="ot">&lt;-</span> atn <span class="sc">%&gt;%</span> <span class="fu">hyper_tibble</span>(<span class="at">force=</span><span class="cn">TRUE</span>)</span>
<span id="cb54-4"><a href="applications.html#cb54-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-5"><a href="applications.html#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(atn_tbl, <span class="at">n=</span><span class="dv">4</span>)</span></code></pre></div>
<pre><code>## # A tibble: 4 × 23
##        time     z   lat   lon   ptt instrument type  location_class error_radius
##       &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt;                 &lt;int&gt;
## 1 622512000     0  34.0 -119. 45866 SPOT       User  nan                      NA
## 2 622708920     0  23.6 -166. 45866 SPOT       Argos A                        NA
## 3 622724940     0  34.0 -119. 45866 SPOT       Argos 1                        NA
## 4 622725060     0  34.0 -119. 45866 SPOT       Argos 0                        NA
## # ℹ 14 more variables: semi_major_axis &lt;int&gt;, semi_minor_axis &lt;int&gt;,
## #   ellipse_orientation &lt;int&gt;, offset &lt;int&gt;, offset_orientation &lt;int&gt;,
## #   gpe_msd &lt;dbl&gt;, gpe_u &lt;dbl&gt;, count &lt;int&gt;, qartod_time_flag &lt;int&gt;,
## #   qartod_speed_flag &lt;int&gt;, qartod_location_flag &lt;int&gt;,
## #   qartod_rollup_flag &lt;int&gt;, comment &lt;chr&gt;, obs &lt;int&gt;</code></pre>
</div>
<div id="dealing-with-time" class="section level4 hasAnchor" number="2.6.1.4">
<h4><span class="header-section-number">2.6.1.4</span> Dealing with time<a href="applications.html#dealing-with-time" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Notice the data in the <strong>time</strong> column aren’t formatted as times. We need to read the metadata associated with the time variable to understand what the units are. Below, we print a tibble of all the attributes from the <strong>time</strong> variable.</p>
<p>Notice the <em>units</em> attribute and it’s value of <code>seconds since 1990-01-01 00:00:00Z</code>. We need to use that information to convert the time variable to something useful that <code>ggplot</code> can handle.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="applications.html#cb56-1" aria-hidden="true" tabindex="-1"></a>time_attrs <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(variable <span class="sc">==</span> <span class="st">&quot;time&quot;</span>)</span>
<span id="cb56-2"><a href="applications.html#cb56-2" aria-hidden="true" tabindex="-1"></a>time_attrs</span></code></pre></div>
<pre><code>## # A tibble: 13 × 4
##       id name                  variable value       
##    &lt;int&gt; &lt;chr&gt;                 &lt;chr&gt;    &lt;named list&gt;
##  1     0 units                 time     &lt;chr [1]&gt;   
##  2     1 standard_name         time     &lt;chr [1]&gt;   
##  3     2 axis                  time     &lt;chr [1]&gt;   
##  4     3 _CoordinateAxisType   time     &lt;chr [1]&gt;   
##  5     4 calendar              time     &lt;chr [1]&gt;   
##  6     5 long_name             time     &lt;chr [1]&gt;   
##  7     6 actual_min            time     &lt;chr [1]&gt;   
##  8     7 actual_max            time     &lt;chr [1]&gt;   
##  9     8 ancillary_variables   time     &lt;chr [1]&gt;   
## 10     9 instrument            time     &lt;chr [1]&gt;   
## 11    10 platform              time     &lt;chr [1]&gt;   
## 12    11 coverage_content_type time     &lt;chr [1]&gt;   
## 13    12 _FillValue            time     &lt;dbl [1]&gt;</code></pre>
<p>So, we grab the value from the <code>units</code> attribute, split the string to collect the date information, and apply that to a time conversion function <code>as.POSIXct</code>.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="applications.html#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co">#library(stringr) - loaded with tidyverse</span></span>
<span id="cb58-2"><a href="applications.html#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="co"># grab origin date from time variable units attribute</span></span>
<span id="cb58-3"><a href="applications.html#cb58-3" aria-hidden="true" tabindex="-1"></a>tunit <span class="ot">&lt;-</span> time_attrs <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(name <span class="sc">==</span> <span class="st">&quot;units&quot;</span>)</span>
<span id="cb58-4"><a href="applications.html#cb58-4" aria-hidden="true" tabindex="-1"></a>lunit <span class="ot">&lt;-</span> <span class="fu">str_split</span>(tunit<span class="sc">$</span>value,<span class="st">&#39; &#39;</span>)[[<span class="dv">1</span>]]</span>
<span id="cb58-5"><a href="applications.html#cb58-5" aria-hidden="true" tabindex="-1"></a>atn_tbl<span class="sc">$</span>time <span class="ot">&lt;-</span> <span class="fu">as.POSIXct</span>(atn_tbl<span class="sc">$</span>time, <span class="at">origin=</span>lunit[<span class="dv">3</span>], <span class="at">tz=</span><span class="st">&quot;GMT&quot;</span>)</span>
<span id="cb58-6"><a href="applications.html#cb58-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-7"><a href="applications.html#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(atn_tbl)</span></code></pre></div>
<pre><code>## tibble [29 × 23] (S3: tbl_df/tbl/data.frame)
##  $ time                : POSIXct[1:29], format: &quot;2009-09-23 00:00:00&quot; &quot;2009-09-25 06:42:00&quot; ...
##  $ z                   : int [1:29] 0 0 0 0 0 0 0 0 0 0 ...
##  $ lat                 : num [1:29] 34 23.6 34 34 34 ...
##  $ lon                 : num [1:29] -119 -166 -119 -119 -119 ...
##  $ ptt                 : int [1:29] 45866 45866 45866 45866 45866 45866 45866 45866 45866 45866 ...
##  $ instrument          : chr [1:29] &quot;SPOT&quot; &quot;SPOT&quot; &quot;SPOT&quot; &quot;SPOT&quot; ...
##  $ type                : chr [1:29] &quot;User&quot; &quot;Argos&quot; &quot;Argos&quot; &quot;Argos&quot; ...
##  $ location_class      : chr [1:29] &quot;nan&quot; &quot;A&quot; &quot;1&quot; &quot;0&quot; ...
##  $ error_radius        : int [1:29] NA NA NA NA NA NA NA NA NA NA ...
##  $ semi_major_axis     : int [1:29] NA NA NA NA NA NA NA NA NA NA ...
##  $ semi_minor_axis     : int [1:29] NA NA NA NA NA NA NA NA NA NA ...
##  $ ellipse_orientation : int [1:29] NA NA NA NA NA NA NA NA NA NA ...
##  $ offset              : int [1:29] NA NA NA NA NA NA NA NA NA NA ...
##  $ offset_orientation  : int [1:29] NA NA NA NA NA NA NA NA NA NA ...
##  $ gpe_msd             : num [1:29] NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN ...
##  $ gpe_u               : num [1:29] NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN ...
##  $ count               : int [1:29] NA NA NA NA NA NA NA NA NA NA ...
##  $ qartod_time_flag    : int [1:29] 1 1 1 1 1 1 1 1 1 1 ...
##  $ qartod_speed_flag   : int [1:29] 2 4 4 4 1 1 1 1 1 1 ...
##  $ qartod_location_flag: int [1:29] 1 1 1 1 1 1 1 1 1 1 ...
##  $ qartod_rollup_flag  : int [1:29] 1 4 4 4 1 1 1 1 1 1 ...
##  $ comment             : chr [1:29] &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; ...
##  $ obs                 : int [1:29] 1 2 3 4 5 6 7 8 9 10 ...</code></pre>
<hr />
</div>
</div>
<div id="converting-to-darwin-core" class="section level3 hasAnchor" number="2.6.2">
<h3><span class="header-section-number">2.6.2</span> Converting to Darwin Core<a href="applications.html#converting-to-darwin-core" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now let’s work through converting this netCDF file to Darwin Core. Following the guidance published at <a href="https://github.com/tdwg/dwc-for-biologging/wiki/Data-guidelines" class="uri">https://github.com/tdwg/dwc-for-biologging/wiki/Data-guidelines</a> and <a href="https://github.com/ocean-tracking-network/biologging_standardization/tree/master/examples/braun-blueshark/darwincore-example" class="uri">https://github.com/ocean-tracking-network/biologging_standardization/tree/master/examples/braun-blueshark/darwincore-example</a></p>
<div id="occurrence-core" class="section level4 hasAnchor" number="2.6.2.1">
<h4><span class="header-section-number">2.6.2.1</span> Occurrence Core<a href="applications.html#occurrence-core" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Below is the mapping table from DarwinCore to the netCDF file.</p>
<table>
<colgroup>
<col width="42%" />
<col width="21%" />
<col width="36%" />
</colgroup>
<thead>
<tr class="header">
<th>DarwinCore Term</th>
<th>Status</th>
<th>netCDF source</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>occurrenceStatus</td>
<td>Required</td>
<td>hardcoded to <code>present</code>.</td>
</tr>
<tr class="even">
<td>basisOfRecord</td>
<td>Required</td>
<td>data contained in the <code>type</code> variable where <code>type</code> of <code>User</code> = <code>HumanObservation</code> and <code>Argos</code> = <code>MachineObservation</code>.</td>
</tr>
<tr class="odd">
<td>occurrenceID</td>
<td>Required</td>
<td><code>eventDate</code>, plus data contained in <code>z</code> variable, plus <code>animal_common_name</code> global attribute.</td>
</tr>
<tr class="even">
<td>organismID</td>
<td>Required</td>
<td><code>platform_id</code> global attribute plus the <code>animal_common_name</code> global attribute.</td>
</tr>
<tr class="odd">
<td>eventDate</td>
<td>Required</td>
<td>data contained in <code>time</code> variable. Converted to ISO8601.</td>
</tr>
<tr class="even">
<td>decimalLatitude &amp; decimalLongitude</td>
<td>Required</td>
<td>data in <code>lat</code> and <code>lon</code> variable, respectively.</td>
</tr>
<tr class="odd">
<td>geodeticDatum</td>
<td>Required</td>
<td>attribute <code>epsg_code</code> in the <code>crs</code> variable.</td>
</tr>
<tr class="even">
<td>scientificName</td>
<td>Required</td>
<td>data from the variable <code>taxon_name</code>.</td>
</tr>
<tr class="odd">
<td>scientificNameID</td>
<td></td>
<td>data from the variable <code>taxon_lsid</code>.</td>
</tr>
<tr class="even">
<td>eventID</td>
<td>Strongly recommended</td>
<td><code>animal_common_name</code> global attribute plus the <code>eventDate</code>.</td>
</tr>
<tr class="odd">
<td>samplingProtocol</td>
<td>Strongly recommended</td>
<td></td>
</tr>
<tr class="even">
<td>kingdom</td>
<td>Strongly recommended</td>
<td><code>kingdom</code> attribute in the <code>animal</code> variable.</td>
</tr>
<tr class="odd">
<td>taxonRank</td>
<td>Strongly recommended</td>
<td><code>rank</code> attribute in the <code>animal</code> variable.</td>
</tr>
<tr class="even">
<td>coordinateUncertaintyInMeters</td>
<td>Share if available</td>
<td>maximum value of the data from the variables <code>error_radius</code>, <code>semi_major_axis</code>, and <code>offset</code>.</td>
</tr>
<tr class="odd">
<td>lifeStage</td>
<td>Share if available</td>
<td>data from the variable <code>animal_life_stage</code>.</td>
</tr>
<tr class="even">
<td>sex</td>
<td>Share if available</td>
<td>data from the variable <code>animal_sex</code>.</td>
</tr>
</tbody>
</table>
<p>Now start working through the crosswalk. A few thoughts about some of the functions we use:</p>
<ol style="list-style-type: decimal">
<li><code>case_when</code> is a <a href="https://dplyr.tidyverse.org/reference/case_when.html">function from dplyr</a> that is essentially a ‘vectorized’ ifelse function. The take-home is that it plays nice with other <code>tidyverse</code> functions, like <code>mutate</code> and IMO is a bit more readable than a complex ifelse statement.</li>
<li><code>rename</code> is another nice dplyr function for renaming columns. It workes well following <code>mutate</code> because you can see the mutation applied to a column and then the column renamed, rather than a complex creation of a new column and dropping of the old column.</li>
</ol>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="applications.html#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Defined to grab attributes in subsequent code</span></span>
<span id="cb60-2"><a href="applications.html#cb60-2" aria-hidden="true" tabindex="-1"></a>nc <span class="ot">&lt;-</span> <span class="fu">nc_open</span>(fname)</span>
<span id="cb60-3"><a href="applications.html#cb60-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-4"><a href="applications.html#cb60-4" aria-hidden="true" tabindex="-1"></a>occurrencedf <span class="ot">&lt;-</span> atn_tbl <span class="sc">%&gt;%</span>  </span>
<span id="cb60-5"><a href="applications.html#cb60-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>( <span class="co"># Select desired columns</span></span>
<span id="cb60-6"><a href="applications.html#cb60-6" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-7"><a href="applications.html#cb60-7" aria-hidden="true" tabindex="-1"></a>        time, </span>
<span id="cb60-8"><a href="applications.html#cb60-8" aria-hidden="true" tabindex="-1"></a>        lat,</span>
<span id="cb60-9"><a href="applications.html#cb60-9" aria-hidden="true" tabindex="-1"></a>        lon,</span>
<span id="cb60-10"><a href="applications.html#cb60-10" aria-hidden="true" tabindex="-1"></a>        type,</span>
<span id="cb60-11"><a href="applications.html#cb60-11" aria-hidden="true" tabindex="-1"></a>        location_class,</span>
<span id="cb60-12"><a href="applications.html#cb60-12" aria-hidden="true" tabindex="-1"></a>        qartod_time_flag,</span>
<span id="cb60-13"><a href="applications.html#cb60-13" aria-hidden="true" tabindex="-1"></a>        qartod_speed_flag,</span>
<span id="cb60-14"><a href="applications.html#cb60-14" aria-hidden="true" tabindex="-1"></a>        qartod_location_flag,</span>
<span id="cb60-15"><a href="applications.html#cb60-15" aria-hidden="true" tabindex="-1"></a>        qartod_rollup_flag</span>
<span id="cb60-16"><a href="applications.html#cb60-16" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-17"><a href="applications.html#cb60-17" aria-hidden="true" tabindex="-1"></a>          ) <span class="sc">%&gt;%</span></span>
<span id="cb60-18"><a href="applications.html#cb60-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>( <span class="co"># add and mutate columns.</span></span>
<span id="cb60-19"><a href="applications.html#cb60-19" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-20"><a href="applications.html#cb60-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">type =</span> <span class="fu">case_when</span>(type <span class="sc">==</span> <span class="st">&#39;User&#39;</span> <span class="sc">~</span> <span class="st">&#39;HumanObservation&#39;</span>,</span>
<span id="cb60-21"><a href="applications.html#cb60-21" aria-hidden="true" tabindex="-1"></a>                         type <span class="sc">==</span> <span class="st">&#39;Argos&#39;</span> <span class="sc">~</span> <span class="st">&#39;MachineObservation&#39;</span>),</span>
<span id="cb60-22"><a href="applications.html#cb60-22" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-23"><a href="applications.html#cb60-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">time =</span> <span class="fu">format</span>(time, <span class="st">&#39;%Y-%m-%dT%H:%M:%SZ&#39;</span>),</span>
<span id="cb60-24"><a href="applications.html#cb60-24" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-25"><a href="applications.html#cb60-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">kingdom =</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(variable <span class="sc">==</span> <span class="st">&quot;animal&quot;</span> <span class="sc">&amp;</span> name <span class="sc">==</span> <span class="st">&quot;kingdom&quot;</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(value) <span class="sc">%&gt;%</span> <span class="fu">unlist</span>(<span class="at">use.names =</span> <span class="cn">FALSE</span>),</span>
<span id="cb60-26"><a href="applications.html#cb60-26" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-27"><a href="applications.html#cb60-27" aria-hidden="true" tabindex="-1"></a>        <span class="at">taxonRank =</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(variable <span class="sc">==</span> <span class="st">&quot;animal&quot;</span> <span class="sc">&amp;</span> name <span class="sc">==</span> <span class="st">&quot;rank&quot;</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(value) <span class="sc">%&gt;%</span> <span class="fu">unlist</span>(<span class="at">use.names =</span> <span class="cn">FALSE</span>),</span>
<span id="cb60-28"><a href="applications.html#cb60-28" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-29"><a href="applications.html#cb60-29" aria-hidden="true" tabindex="-1"></a>        <span class="at">occurrenceStatus =</span> <span class="st">&quot;present&quot;</span>,</span>
<span id="cb60-30"><a href="applications.html#cb60-30" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-31"><a href="applications.html#cb60-31" aria-hidden="true" tabindex="-1"></a>        <span class="at">sex =</span> <span class="fu">ncvar_get</span>( nc, <span class="st">&#39;animal_sex&#39;</span>),</span>
<span id="cb60-32"><a href="applications.html#cb60-32" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-33"><a href="applications.html#cb60-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">lifeStage =</span> <span class="fu">ncvar_get</span>( nc, <span class="st">&#39;animal_life_stage&#39;</span>),</span>
<span id="cb60-34"><a href="applications.html#cb60-34" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-35"><a href="applications.html#cb60-35" aria-hidden="true" tabindex="-1"></a>        <span class="at">scientificName =</span> <span class="fu">ncvar_get</span>( nc, <span class="st">&#39;taxon_name&#39;</span>),</span>
<span id="cb60-36"><a href="applications.html#cb60-36" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-37"><a href="applications.html#cb60-37" aria-hidden="true" tabindex="-1"></a>        <span class="at">scientificNameID =</span> <span class="fu">ncvar_get</span>( nc, <span class="st">&quot;taxon_lsid&quot;</span>)</span>
<span id="cb60-38"><a href="applications.html#cb60-38" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-39"><a href="applications.html#cb60-39" aria-hidden="true" tabindex="-1"></a>        ) <span class="sc">%&gt;%</span></span>
<span id="cb60-40"><a href="applications.html#cb60-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-41"><a href="applications.html#cb60-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rename</span>(  <span class="co"># rename columns to Darwin Core terms</span></span>
<span id="cb60-42"><a href="applications.html#cb60-42" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb60-43"><a href="applications.html#cb60-43" aria-hidden="true" tabindex="-1"></a>        <span class="at">basisOfRecord =</span> type,</span>
<span id="cb60-44"><a href="applications.html#cb60-44" aria-hidden="true" tabindex="-1"></a>        <span class="at">eventDate =</span> time,</span>
<span id="cb60-45"><a href="applications.html#cb60-45" aria-hidden="true" tabindex="-1"></a>        <span class="at">decimalLatitude =</span> lat,</span>
<span id="cb60-46"><a href="applications.html#cb60-46" aria-hidden="true" tabindex="-1"></a>        <span class="at">decimalLongitude =</span> lon) <span class="sc">%&gt;%</span> </span>
<span id="cb60-47"><a href="applications.html#cb60-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-48"><a href="applications.html#cb60-48" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(eventDate) <span class="co">#arrange by increasing date</span></span>
<span id="cb60-49"><a href="applications.html#cb60-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-50"><a href="applications.html#cb60-50" aria-hidden="true" tabindex="-1"></a><span class="co"># minimumDepthInMeters = z,</span></span>
<span id="cb60-51"><a href="applications.html#cb60-51" aria-hidden="true" tabindex="-1"></a>occurrencedf<span class="sc">$</span>minimumDepthInMeters <span class="ot">=</span> atn_tbl<span class="sc">$</span>z</span>
<span id="cb60-52"><a href="applications.html#cb60-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-53"><a href="applications.html#cb60-53" aria-hidden="true" tabindex="-1"></a><span class="co"># maximumDepthInMeters = z,</span></span>
<span id="cb60-54"><a href="applications.html#cb60-54" aria-hidden="true" tabindex="-1"></a>occurrencedf<span class="sc">$</span>maximumDepthInMeters <span class="ot">=</span> atn_tbl<span class="sc">$</span>z</span>
<span id="cb60-55"><a href="applications.html#cb60-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-56"><a href="applications.html#cb60-56" aria-hidden="true" tabindex="-1"></a><span class="co"># organismID - {platformID}_{common_name}</span></span>
<span id="cb60-57"><a href="applications.html#cb60-57" aria-hidden="true" tabindex="-1"></a>common_name_tbl <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(variable <span class="sc">==</span> <span class="st">&quot;NC_GLOBAL&quot;</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(name <span class="sc">==</span> <span class="st">&quot;animal_common_name&quot;</span>)</span>
<span id="cb60-58"><a href="applications.html#cb60-58" aria-hidden="true" tabindex="-1"></a>common_name <span class="ot">&lt;-</span> <span class="fu">chartr</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>, common_name_tbl<span class="sc">$</span>value)</span>
<span id="cb60-59"><a href="applications.html#cb60-59" aria-hidden="true" tabindex="-1"></a>platform_id_tbl <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(variable <span class="sc">==</span> <span class="st">&quot;NC_GLOBAL&quot;</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(name <span class="sc">==</span> <span class="st">&quot;platform_id&quot;</span>)</span>
<span id="cb60-60"><a href="applications.html#cb60-60" aria-hidden="true" tabindex="-1"></a>platform_id <span class="ot">&lt;-</span> <span class="fu">chartr</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>, platform_id_tbl<span class="sc">$</span>value)</span>
<span id="cb60-61"><a href="applications.html#cb60-61" aria-hidden="true" tabindex="-1"></a>occurrencedf<span class="sc">$</span>organismID <span class="ot">&lt;-</span> <span class="fu">paste</span>(platform_id , common_name, <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>) </span>
<span id="cb60-62"><a href="applications.html#cb60-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-63"><a href="applications.html#cb60-63" aria-hidden="true" tabindex="-1"></a><span class="co"># occurrenceID - {eventDate}_{depth}_{common_name}</span></span>
<span id="cb60-64"><a href="applications.html#cb60-64" aria-hidden="true" tabindex="-1"></a>occurrencedf<span class="sc">$</span>occurrenceID <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>, <span class="fu">paste</span>(occurrencedf<span class="sc">$</span>eventDate, atn_tbl<span class="sc">$</span>z, common_name, <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>))</span>
<span id="cb60-65"><a href="applications.html#cb60-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-66"><a href="applications.html#cb60-66" aria-hidden="true" tabindex="-1"></a><span class="co"># geodeticDatum</span></span>
<span id="cb60-67"><a href="applications.html#cb60-67" aria-hidden="true" tabindex="-1"></a>gd_tbl <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(variable <span class="sc">==</span> <span class="st">&quot;crs&quot;</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(name <span class="sc">==</span> <span class="st">&quot;epsg_code&quot;</span>)</span>
<span id="cb60-68"><a href="applications.html#cb60-68" aria-hidden="true" tabindex="-1"></a>occurrencedf<span class="sc">$</span>geodeticDatum <span class="ot">&lt;-</span> <span class="fu">paste</span>(gd_tbl<span class="sc">$</span>value)</span>
<span id="cb60-69"><a href="applications.html#cb60-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-70"><a href="applications.html#cb60-70" aria-hidden="true" tabindex="-1"></a><span class="co"># eventID</span></span>
<span id="cb60-71"><a href="applications.html#cb60-71" aria-hidden="true" tabindex="-1"></a><span class="co">#eventID - {common_name}_{dateTime}</span></span>
<span id="cb60-72"><a href="applications.html#cb60-72" aria-hidden="true" tabindex="-1"></a>cname <span class="ot">=</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(variable <span class="sc">==</span> <span class="st">&quot;NC_GLOBAL&quot;</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(name <span class="sc">==</span> <span class="st">&quot;animal_common_name&quot;</span>)</span>
<span id="cb60-73"><a href="applications.html#cb60-73" aria-hidden="true" tabindex="-1"></a>occurrencedf<span class="sc">$</span>eventID <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>, <span class="fu">paste0</span>(cname<span class="sc">$</span>value, <span class="st">&quot;_&quot;</span>, occurrencedf<span class="sc">$</span>eventDate))</span></code></pre></div>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="applications.html#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(occurrencedf)</span></code></pre></div>
<pre><code>## tibble [29 × 22] (S3: tbl_df/tbl/data.frame)
##  $ eventDate           : chr [1:29] &quot;2009-09-23T00:00:00Z&quot; &quot;2009-09-25T06:42:00Z&quot; &quot;2009-09-25T11:09:00Z&quot; &quot;2009-09-25T11:11:00Z&quot; ...
##  $ decimalLatitude     : num [1:29] 34 23.6 34 34 34 ...
##  $ decimalLongitude    : num [1:29] -119 -166 -119 -119 -119 ...
##  $ basisOfRecord       : chr [1:29] &quot;HumanObservation&quot; &quot;MachineObservation&quot; &quot;MachineObservation&quot; &quot;MachineObservation&quot; ...
##  $ location_class      : chr [1:29] &quot;nan&quot; &quot;A&quot; &quot;1&quot; &quot;0&quot; ...
##  $ qartod_time_flag    : int [1:29] 1 1 1 1 1 1 1 1 1 1 ...
##  $ qartod_speed_flag   : int [1:29] 2 4 4 4 1 1 1 1 1 1 ...
##  $ qartod_location_flag: int [1:29] 1 1 1 1 1 1 1 1 1 1 ...
##  $ qartod_rollup_flag  : int [1:29] 1 4 4 4 1 1 1 1 1 1 ...
##  $ kingdom             : chr [1:29] &quot;Animalia&quot; &quot;Animalia&quot; &quot;Animalia&quot; &quot;Animalia&quot; ...
##  $ taxonRank           : chr [1:29] &quot;Species&quot; &quot;Species&quot; &quot;Species&quot; &quot;Species&quot; ...
##  $ occurrenceStatus    : chr [1:29] &quot;present&quot; &quot;present&quot; &quot;present&quot; &quot;present&quot; ...
##  $ sex                 : chr [1:29] &quot;male&quot; &quot;male&quot; &quot;male&quot; &quot;male&quot; ...
##  $ lifeStage           : chr [1:29] &quot;juvenile&quot; &quot;juvenile&quot; &quot;juvenile&quot; &quot;juvenile&quot; ...
##  $ scientificName      : chr [1:29] &quot;Carcharodon carcharias&quot; &quot;Carcharodon carcharias&quot; &quot;Carcharodon carcharias&quot; &quot;Carcharodon carcharias&quot; ...
##  $ scientificNameID    : chr [1:29] &quot;urn:lsid:marinespecies.org:taxname:105838&quot; &quot;urn:lsid:marinespecies.org:taxname:105838&quot; &quot;urn:lsid:marinespecies.org:taxname:105838&quot; &quot;urn:lsid:marinespecies.org:taxname:105838&quot; ...
##  $ minimumDepthInMeters: int [1:29] 0 0 0 0 0 0 0 0 0 0 ...
##  $ maximumDepthInMeters: int [1:29] 0 0 0 0 0 0 0 0 0 0 ...
##  $ organismID          : chr [1:29] &quot;105838_great_white_shark&quot; &quot;105838_great_white_shark&quot; &quot;105838_great_white_shark&quot; &quot;105838_great_white_shark&quot; ...
##  $ occurrenceID        : chr [1:29] &quot;2009-09-23T00:00:00Z_0_great_white_shark&quot; &quot;2009-09-25T06:42:00Z_0_great_white_shark&quot; &quot;2009-09-25T11:09:00Z_0_great_white_shark&quot; &quot;2009-09-25T11:11:00Z_0_great_white_shark&quot; ...
##  $ geodeticDatum       : chr [1:29] &quot;EPSG:4326&quot; &quot;EPSG:4326&quot; &quot;EPSG:4326&quot; &quot;EPSG:4326&quot; ...
##  $ eventID             : chr [1:29] &quot;great_white shark_2009-09-23T00:00:00Z&quot; &quot;great_white shark_2009-09-25T06:42:00Z&quot; &quot;great_white shark_2009-09-25T11:09:00Z&quot; &quot;great_white shark_2009-09-25T11:11:00Z&quot; ...</code></pre>
<div id="add-coordinateuncertaintyinmeters-and-filter-by-location_class" class="section level5 hasAnchor" number="2.6.2.1.1">
<h5><span class="header-section-number">2.6.2.1.1</span> Add coordinateUncertaintyInMeters AND filter by location_class<a href="applications.html#add-coordinateuncertaintyinmeters-and-filter-by-location_class" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>When we add <code>coordinateUncertaintyInMeters</code> we are also filtering out where <code>location_class</code> == <code>A</code>,<code>B</code>,or <code>Z</code>.</p>
<p>In these data we also have additional information about the Location Quality Code from ARGOS satellite system. Below are the codes and those meanings.</p>
<table>
<colgroup>
<col width="46%" />
<col width="53%" />
</colgroup>
<thead>
<tr class="header">
<th>code_values</th>
<th>code meanings</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>G</td>
<td>estimated error less than 100m and 1+ messages received per satellite pass</td>
</tr>
<tr class="even">
<td>3</td>
<td>estimated error less than 250m and 4+ messages received per satellite pass</td>
</tr>
<tr class="odd">
<td>2</td>
<td>estimated error between 250m and 500m and 4+ messages per satellite pass</td>
</tr>
<tr class="even">
<td>1</td>
<td>estimated error between 500m and 1500m and 4+ messages per satellite pass</td>
</tr>
<tr class="odd">
<td>0</td>
<td>estimated error greater than 1500m and 4+ messages received per satellite pass</td>
</tr>
<tr class="even">
<td>A</td>
<td>no least squares estimated error or unbounded kalman filter estimated error and 3 messages received per satellite pass</td>
</tr>
<tr class="odd">
<td>B</td>
<td>no least squares estimated error or unbounded kalman filter estimated error and 1 or 2 messages received per satellite pass</td>
</tr>
<tr class="even">
<td>Z</td>
<td>invalid location (available for Service Plus or Auxilliary Location Processing)</td>
</tr>
</tbody>
</table>
<p>Since codes <code>A</code>, <code>B</code>, and <code>Z</code> are essentially bad values, I propose that we filter those out.</p>
<p>Also, create a mapping table for <code>coordinateUncertaintyInMeters</code> that corresponds to the ARGOS code maximum error as shown in the table below:</p>
<table>
<thead>
<tr class="header">
<th>code</th>
<th>coordinateUncertaintyInMeters</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>G</td>
<td>100</td>
</tr>
<tr class="even">
<td>3</td>
<td>250</td>
</tr>
<tr class="odd">
<td>2</td>
<td>500</td>
</tr>
<tr class="even">
<td>1</td>
<td>1500</td>
</tr>
<tr class="odd">
<td>0</td>
<td>10000 (<a href="https://github.com/ioos/bio_data_guide/issues/145#issuecomment-1805739244">ref</a>)</td>
</tr>
</tbody>
</table>
<p>Below we create a lookup table for the <code>location_class</code> values we agree are good, which contains the <code>coordinateUncertaintyInMeters</code> for the appropriate location class. When we merge that table with our raw data, the observations that don’t match the location_classes in our lookup table will not be transfered over (ie. they will be filtered out).</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="applications.html#cb63-1" aria-hidden="true" tabindex="-1"></a>occurrencedf <span class="ot">&lt;-</span> occurrencedf <span class="sc">%&gt;%</span></span>
<span id="cb63-2"><a href="applications.html#cb63-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(location_class <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&#39;nan&#39;</span>,<span class="st">&#39;G&#39;</span>,<span class="st">&#39;3&#39;</span>,<span class="st">&#39;2&#39;</span>,<span class="st">&#39;1&#39;</span>,<span class="st">&#39;0&#39;</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb63-3"><a href="applications.html#cb63-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(  <span class="co"># This returns NA for any other values than those defined below</span></span>
<span id="cb63-4"><a href="applications.html#cb63-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">coordinateUncertaintyInMeters =</span> <span class="fu">case_when</span>(location_class <span class="sc">==</span> <span class="st">&#39;nan&#39;</span> <span class="sc">~</span> <span class="dv">0</span>,</span>
<span id="cb63-5"><a href="applications.html#cb63-5" aria-hidden="true" tabindex="-1"></a>                                                     location_class <span class="sc">==</span> <span class="st">&#39;G&#39;</span> <span class="sc">~</span> <span class="dv">200</span>,</span>
<span id="cb63-6"><a href="applications.html#cb63-6" aria-hidden="true" tabindex="-1"></a>                                                     location_class <span class="sc">==</span> <span class="st">&#39;3&#39;</span> <span class="sc">~</span> <span class="dv">250</span>,</span>
<span id="cb63-7"><a href="applications.html#cb63-7" aria-hidden="true" tabindex="-1"></a>                                                     location_class <span class="sc">==</span> <span class="st">&#39;2&#39;</span> <span class="sc">~</span> <span class="dv">500</span>,</span>
<span id="cb63-8"><a href="applications.html#cb63-8" aria-hidden="true" tabindex="-1"></a>                                                     location_class <span class="sc">==</span> <span class="st">&#39;1&#39;</span> <span class="sc">~</span> <span class="dv">1500</span>,</span>
<span id="cb63-9"><a href="applications.html#cb63-9" aria-hidden="true" tabindex="-1"></a>                                                     location_class <span class="sc">==</span> <span class="st">&#39;0&#39;</span> <span class="sc">~</span> <span class="dv">10000</span>) <span class="co"># https://github.com/ioos/bio_data_guide/issues/145#issuecomment-1805739244</span></span>
<span id="cb63-10"><a href="applications.html#cb63-10" aria-hidden="true" tabindex="-1"></a>          ) <span class="sc">%&gt;%</span> </span>
<span id="cb63-11"><a href="applications.html#cb63-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(eventDate) <span class="co"># arrange by increasing date</span></span>
<span id="cb63-12"><a href="applications.html#cb63-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-13"><a href="applications.html#cb63-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-14"><a href="applications.html#cb63-14" aria-hidden="true" tabindex="-1"></a>occurrencedf</span></code></pre></div>
<pre><code>## # A tibble: 19 × 23
##    eventDate       decimalLatitude decimalLongitude basisOfRecord location_class
##    &lt;chr&gt;                     &lt;dbl&gt;            &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;         
##  1 2009-09-23T00:…            34.0            -119. HumanObserva… nan           
##  2 2009-09-25T11:…            34.0            -119. MachineObser… 1             
##  3 2009-09-25T11:…            34.0            -119. MachineObser… 0             
##  4 2009-09-27T17:…            34.0            -119. MachineObser… 1             
##  5 2009-10-08T20:…            34.0            -119. MachineObser… 2             
##  6 2009-10-15T11:…            34.0            -119. MachineObser… 0             
##  7 2009-10-17T06:…            34.0            -119. MachineObser… 0             
##  8 2009-10-17T09:…            34.0            -119. MachineObser… 2             
##  9 2009-10-17T10:…            34.0            -119. MachineObser… 3             
## 10 2009-10-18T08:…            34.0            -119. MachineObser… 1             
## 11 2009-10-18T10:…            34.0            -119. MachineObser… 2             
## 12 2009-10-18T11:…            34.0            -119. MachineObser… 0             
## 13 2009-10-23T23:…            34.0            -119. MachineObser… 2             
## 14 2009-10-24T00:…            34.0            -119. MachineObser… 0             
## 15 2009-10-26T10:…            34.0            -119. MachineObser… 3             
## 16 2009-10-27T16:…            34.0            -119. MachineObser… 1             
## 17 2009-10-27T16:…            34.0            -119. MachineObser… 2             
## 18 2009-10-29T11:…            34.0            -119. MachineObser… 2             
## 19 2009-10-31T21:…            34.0            -119. MachineObser… 0             
## # ℹ 18 more variables: qartod_time_flag &lt;int&gt;, qartod_speed_flag &lt;int&gt;,
## #   qartod_location_flag &lt;int&gt;, qartod_rollup_flag &lt;int&gt;, kingdom &lt;chr&gt;,
## #   taxonRank &lt;chr&gt;, occurrenceStatus &lt;chr&gt;, sex &lt;chr&gt;, lifeStage &lt;chr&gt;,
## #   scientificName &lt;chr&gt;, scientificNameID &lt;chr&gt;, minimumDepthInMeters &lt;int&gt;,
## #   maximumDepthInMeters &lt;int&gt;, organismID &lt;chr&gt;, occurrenceID &lt;chr&gt;,
## #   geodeticDatum &lt;chr&gt;, eventID &lt;chr&gt;, coordinateUncertaintyInMeters &lt;dbl&gt;</code></pre>
<p>Notice how we went from 29 rows down to 19 rows by only selecting specific the location_class.</p>
</div>
<div id="create-a-datageneralizations-column-to-describe-how-many-duplicates-were-found-for-each-deprecation-series" class="section level5 hasAnchor" number="2.6.2.1.2">
<h5><span class="header-section-number">2.6.2.1.2</span> Create a dataGeneralizations column to describe how many duplicates were found for each deprecation series<a href="applications.html#create-a-datageneralizations-column-to-describe-how-many-duplicates-were-found-for-each-deprecation-series" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Add a <code>dataGeneralizations</code> column containing a string like ‘first of # records’ to indicate there are more records in the raw dataset to be discovered by the super-curious.</p>
<p>The dataGeneralizations string is compiled by counting the number of consecutive duplicates and inserting that into a standard string. That string is “first of [n] records” which will make more sense once we’ve filtered down to keep the first occurrence of the hour.</p>
<p>The next step below this, we filter out only the first observation of the hour.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="applications.html#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by date</span></span>
<span id="cb65-2"><a href="applications.html#cb65-2" aria-hidden="true" tabindex="-1"></a>occurrencedf <span class="ot">&lt;-</span> occurrencedf <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(eventDate)</span>
<span id="cb65-3"><a href="applications.html#cb65-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-4"><a href="applications.html#cb65-4" aria-hidden="true" tabindex="-1"></a>occurrencedf <span class="ot">&lt;-</span> occurrencedf <span class="sc">%&gt;%</span></span>
<span id="cb65-5"><a href="applications.html#cb65-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">eventDateHrs =</span> <span class="fu">format</span>(<span class="fu">as.POSIXct</span>(eventDate, <span class="at">format=</span><span class="st">&quot;%Y-%m-%dT%H:%M:%SZ&quot;</span>),<span class="st">&quot;%Y-%m-%dT%H&quot;</span>)</span>
<span id="cb65-6"><a href="applications.html#cb65-6" aria-hidden="true" tabindex="-1"></a>           ) <span class="sc">%&gt;%</span></span>
<span id="cb65-7"><a href="applications.html#cb65-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">add_count</span>(eventDateHrs) <span class="sc">%&gt;%</span></span>
<span id="cb65-8"><a href="applications.html#cb65-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">dataGeneralizations =</span> <span class="fu">case_when</span>(n <span class="sc">==</span> <span class="dv">1</span> <span class="sc">~</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb65-9"><a href="applications.html#cb65-9" aria-hidden="true" tabindex="-1"></a>                                           <span class="cn">TRUE</span> <span class="sc">~</span> <span class="fu">paste</span>(<span class="st">&quot;first of &quot;</span>, n ,<span class="st">&quot;records&quot;</span>)</span>
<span id="cb65-10"><a href="applications.html#cb65-10" aria-hidden="true" tabindex="-1"></a>                                           )</span>
<span id="cb65-11"><a href="applications.html#cb65-11" aria-hidden="true" tabindex="-1"></a>           ) <span class="sc">%&gt;%</span></span>
<span id="cb65-12"><a href="applications.html#cb65-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span>n)</span>
<span id="cb65-13"><a href="applications.html#cb65-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-14"><a href="applications.html#cb65-14" aria-hidden="true" tabindex="-1"></a>occurrencedf</span></code></pre></div>
<pre><code>## # A tibble: 19 × 25
##    eventDate       decimalLatitude decimalLongitude basisOfRecord location_class
##    &lt;chr&gt;                     &lt;dbl&gt;            &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;         
##  1 2009-09-23T00:…            34.0            -119. HumanObserva… nan           
##  2 2009-09-25T11:…            34.0            -119. MachineObser… 1             
##  3 2009-09-25T11:…            34.0            -119. MachineObser… 0             
##  4 2009-09-27T17:…            34.0            -119. MachineObser… 1             
##  5 2009-10-08T20:…            34.0            -119. MachineObser… 2             
##  6 2009-10-15T11:…            34.0            -119. MachineObser… 0             
##  7 2009-10-17T06:…            34.0            -119. MachineObser… 0             
##  8 2009-10-17T09:…            34.0            -119. MachineObser… 2             
##  9 2009-10-17T10:…            34.0            -119. MachineObser… 3             
## 10 2009-10-18T08:…            34.0            -119. MachineObser… 1             
## 11 2009-10-18T10:…            34.0            -119. MachineObser… 2             
## 12 2009-10-18T11:…            34.0            -119. MachineObser… 0             
## 13 2009-10-23T23:…            34.0            -119. MachineObser… 2             
## 14 2009-10-24T00:…            34.0            -119. MachineObser… 0             
## 15 2009-10-26T10:…            34.0            -119. MachineObser… 3             
## 16 2009-10-27T16:…            34.0            -119. MachineObser… 1             
## 17 2009-10-27T16:…            34.0            -119. MachineObser… 2             
## 18 2009-10-29T11:…            34.0            -119. MachineObser… 2             
## 19 2009-10-31T21:…            34.0            -119. MachineObser… 0             
## # ℹ 20 more variables: qartod_time_flag &lt;int&gt;, qartod_speed_flag &lt;int&gt;,
## #   qartod_location_flag &lt;int&gt;, qartod_rollup_flag &lt;int&gt;, kingdom &lt;chr&gt;,
## #   taxonRank &lt;chr&gt;, occurrenceStatus &lt;chr&gt;, sex &lt;chr&gt;, lifeStage &lt;chr&gt;,
## #   scientificName &lt;chr&gt;, scientificNameID &lt;chr&gt;, minimumDepthInMeters &lt;int&gt;,
## #   maximumDepthInMeters &lt;int&gt;, organismID &lt;chr&gt;, occurrenceID &lt;chr&gt;,
## #   geodeticDatum &lt;chr&gt;, eventID &lt;chr&gt;, coordinateUncertaintyInMeters &lt;dbl&gt;,
## #   eventDateHrs &lt;chr&gt;, dataGeneralizations &lt;chr&gt;</code></pre>
</div>
<div id="decimate-occurrences-down-to-the-first-detectionlocation-per-hour" class="section level5 hasAnchor" number="2.6.2.1.3">
<h5><span class="header-section-number">2.6.2.1.3</span> Decimate occurrences down to the first detection/location per hour<a href="applications.html#decimate-occurrences-down-to-the-first-detectionlocation-per-hour" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Here we’ve done the decimation in Python: <a href="https://gist.github.com/MathewBiddle/d434ac2b538b2728aa80c6a7945f94be" class="uri">https://gist.github.com/MathewBiddle/d434ac2b538b2728aa80c6a7945f94be</a></p>
<p>Essentially we build a new colum that is the date plus the two digit hour. Then we find where that column has duplicates and keep the first entry.</p>
<p>In R, we do something slightly different as we only keep the distinct (ie. unique) rows and if there are duplicates, pick the first row of the duplicate.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="applications.html#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sort by date</span></span>
<span id="cb67-2"><a href="applications.html#cb67-2" aria-hidden="true" tabindex="-1"></a>occurrencedf_dec <span class="ot">&lt;-</span> occurrencedf <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(eventDate)</span>
<span id="cb67-3"><a href="applications.html#cb67-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-4"><a href="applications.html#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="co"># filter table to only unique date + hour and pick the first row.</span></span>
<span id="cb67-5"><a href="applications.html#cb67-5" aria-hidden="true" tabindex="-1"></a>occurrencedf_dec <span class="ot">&lt;-</span> <span class="fu">distinct</span>(occurrencedf_dec,eventDateHrs,<span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb67-6"><a href="applications.html#cb67-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span>eventDateHrs)</span>
<span id="cb67-7"><a href="applications.html#cb67-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-8"><a href="applications.html#cb67-8" aria-hidden="true" tabindex="-1"></a>occurrencedf_dec</span></code></pre></div>
<pre><code>## # A tibble: 17 × 24
##    eventDate       decimalLatitude decimalLongitude basisOfRecord location_class
##    &lt;chr&gt;                     &lt;dbl&gt;            &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;         
##  1 2009-09-23T00:…            34.0            -119. HumanObserva… nan           
##  2 2009-09-25T11:…            34.0            -119. MachineObser… 1             
##  3 2009-09-27T17:…            34.0            -119. MachineObser… 1             
##  4 2009-10-08T20:…            34.0            -119. MachineObser… 2             
##  5 2009-10-15T11:…            34.0            -119. MachineObser… 0             
##  6 2009-10-17T06:…            34.0            -119. MachineObser… 0             
##  7 2009-10-17T09:…            34.0            -119. MachineObser… 2             
##  8 2009-10-17T10:…            34.0            -119. MachineObser… 3             
##  9 2009-10-18T08:…            34.0            -119. MachineObser… 1             
## 10 2009-10-18T10:…            34.0            -119. MachineObser… 2             
## 11 2009-10-18T11:…            34.0            -119. MachineObser… 0             
## 12 2009-10-23T23:…            34.0            -119. MachineObser… 2             
## 13 2009-10-24T00:…            34.0            -119. MachineObser… 0             
## 14 2009-10-26T10:…            34.0            -119. MachineObser… 3             
## 15 2009-10-27T16:…            34.0            -119. MachineObser… 1             
## 16 2009-10-29T11:…            34.0            -119. MachineObser… 2             
## 17 2009-10-31T21:…            34.0            -119. MachineObser… 0             
## # ℹ 19 more variables: qartod_time_flag &lt;int&gt;, qartod_speed_flag &lt;int&gt;,
## #   qartod_location_flag &lt;int&gt;, qartod_rollup_flag &lt;int&gt;, kingdom &lt;chr&gt;,
## #   taxonRank &lt;chr&gt;, occurrenceStatus &lt;chr&gt;, sex &lt;chr&gt;, lifeStage &lt;chr&gt;,
## #   scientificName &lt;chr&gt;, scientificNameID &lt;chr&gt;, minimumDepthInMeters &lt;int&gt;,
## #   maximumDepthInMeters &lt;int&gt;, organismID &lt;chr&gt;, occurrenceID &lt;chr&gt;,
## #   geodeticDatum &lt;chr&gt;, eventID &lt;chr&gt;, coordinateUncertaintyInMeters &lt;dbl&gt;,
## #   dataGeneralizations &lt;chr&gt;</code></pre>
<p>Notice that we have gone from 19 rows to 17 rows. Removing rows observed on <code>2009-09-25T11:11:00Z</code> and <code>2009-10-27T16:22:00Z</code> as they were the second points within that specifc hour.</p>
</div>
<div id="filter-on-qartod-flags" class="section level5 hasAnchor" number="2.6.2.1.4">
<h5><span class="header-section-number">2.6.2.1.4</span> Filter on QARTOD flags?<a href="applications.html#filter-on-qartod-flags" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>We also have QARTOD flags and they are as follows:</p>
<table>
<thead>
<tr class="header">
<th>value</th>
<th>meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>PASS</td>
</tr>
<tr class="even">
<td>2</td>
<td>NOT_EVALUATED</td>
</tr>
<tr class="odd">
<td>3</td>
<td>SUSPECT</td>
</tr>
<tr class="even">
<td>4</td>
<td>FAIL</td>
</tr>
<tr class="odd">
<td>9</td>
<td>MISSING</td>
</tr>
</tbody>
</table>
<p>The QARTOD tests are:</p>
<table>
<thead>
<tr class="header">
<th>variable</th>
<th>long_name</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>qartod_time_flag</td>
<td>Time QC test - gross range test</td>
</tr>
<tr class="even">
<td>qartod_speed_flag</td>
<td>Speed QC test - gross range test</td>
</tr>
<tr class="odd">
<td>qartod_location_flag</td>
<td>Location QC test - Location test</td>
</tr>
<tr class="even">
<td>qartod_rollup_flag</td>
<td>Aggregate QC value</td>
</tr>
</tbody>
</table>
<p>I’m not sure what to do here. My preference would be to include all rows where <code>qartod_rollup_flag</code> == 1 and drop the rest. But I’m open to suggestions.</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="applications.html#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># perform filter but don&#39;t save it.</span></span>
<span id="cb69-2"><a href="applications.html#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>(occurrencedf_dec, qartod_rollup_flag <span class="sc">==</span> <span class="dv">1</span>)</span></code></pre></div>
<pre><code>## # A tibble: 16 × 24
##    eventDate       decimalLatitude decimalLongitude basisOfRecord location_class
##    &lt;chr&gt;                     &lt;dbl&gt;            &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;         
##  1 2009-09-23T00:…            34.0            -119. HumanObserva… nan           
##  2 2009-09-27T17:…            34.0            -119. MachineObser… 1             
##  3 2009-10-08T20:…            34.0            -119. MachineObser… 2             
##  4 2009-10-15T11:…            34.0            -119. MachineObser… 0             
##  5 2009-10-17T06:…            34.0            -119. MachineObser… 0             
##  6 2009-10-17T09:…            34.0            -119. MachineObser… 2             
##  7 2009-10-17T10:…            34.0            -119. MachineObser… 3             
##  8 2009-10-18T08:…            34.0            -119. MachineObser… 1             
##  9 2009-10-18T10:…            34.0            -119. MachineObser… 2             
## 10 2009-10-18T11:…            34.0            -119. MachineObser… 0             
## 11 2009-10-23T23:…            34.0            -119. MachineObser… 2             
## 12 2009-10-24T00:…            34.0            -119. MachineObser… 0             
## 13 2009-10-26T10:…            34.0            -119. MachineObser… 3             
## 14 2009-10-27T16:…            34.0            -119. MachineObser… 1             
## 15 2009-10-29T11:…            34.0            -119. MachineObser… 2             
## 16 2009-10-31T21:…            34.0            -119. MachineObser… 0             
## # ℹ 19 more variables: qartod_time_flag &lt;int&gt;, qartod_speed_flag &lt;int&gt;,
## #   qartod_location_flag &lt;int&gt;, qartod_rollup_flag &lt;int&gt;, kingdom &lt;chr&gt;,
## #   taxonRank &lt;chr&gt;, occurrenceStatus &lt;chr&gt;, sex &lt;chr&gt;, lifeStage &lt;chr&gt;,
## #   scientificName &lt;chr&gt;, scientificNameID &lt;chr&gt;, minimumDepthInMeters &lt;int&gt;,
## #   maximumDepthInMeters &lt;int&gt;, organismID &lt;chr&gt;, occurrenceID &lt;chr&gt;,
## #   geodeticDatum &lt;chr&gt;, eventID &lt;chr&gt;, coordinateUncertaintyInMeters &lt;dbl&gt;,
## #   dataGeneralizations &lt;chr&gt;</code></pre>
<p>Drop the quality flag columns to align with DarwinCore standard.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="applications.html#cb71-1" aria-hidden="true" tabindex="-1"></a>occurrencedf_dec <span class="ot">&lt;-</span> occurrencedf_dec <span class="sc">%&gt;%</span></span>
<span id="cb71-2"><a href="applications.html#cb71-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(</span>
<span id="cb71-3"><a href="applications.html#cb71-3" aria-hidden="true" tabindex="-1"></a>        <span class="sc">-</span><span class="fu">c</span>(location_class,</span>
<span id="cb71-4"><a href="applications.html#cb71-4" aria-hidden="true" tabindex="-1"></a>           qartod_time_flag,</span>
<span id="cb71-5"><a href="applications.html#cb71-5" aria-hidden="true" tabindex="-1"></a>           qartod_speed_flag,</span>
<span id="cb71-6"><a href="applications.html#cb71-6" aria-hidden="true" tabindex="-1"></a>           qartod_location_flag,</span>
<span id="cb71-7"><a href="applications.html#cb71-7" aria-hidden="true" tabindex="-1"></a>           qartod_rollup_flag</span>
<span id="cb71-8"><a href="applications.html#cb71-8" aria-hidden="true" tabindex="-1"></a>           ))</span>
<span id="cb71-9"><a href="applications.html#cb71-9" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb71-10"><a href="applications.html#cb71-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(occurrencedf_dec)</span></code></pre></div>
<pre><code>##  [1] &quot;eventDate&quot;                     &quot;decimalLatitude&quot;              
##  [3] &quot;decimalLongitude&quot;              &quot;basisOfRecord&quot;                
##  [5] &quot;kingdom&quot;                       &quot;taxonRank&quot;                    
##  [7] &quot;occurrenceStatus&quot;              &quot;sex&quot;                          
##  [9] &quot;lifeStage&quot;                     &quot;scientificName&quot;               
## [11] &quot;scientificNameID&quot;              &quot;minimumDepthInMeters&quot;         
## [13] &quot;maximumDepthInMeters&quot;          &quot;organismID&quot;                   
## [15] &quot;occurrenceID&quot;                  &quot;geodeticDatum&quot;                
## [17] &quot;eventID&quot;                       &quot;coordinateUncertaintyInMeters&quot;
## [19] &quot;dataGeneralizations&quot;</code></pre>
</div>
</div>
<div id="add-ncei-url-to-references" class="section level4 hasAnchor" number="2.6.2.2">
<h4><span class="header-section-number">2.6.2.2</span> Add NCEI URL to references<a href="applications.html#add-ncei-url-to-references" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>For full transparency, we will add the Darwin Core term references to provide the URL to the NCEI landing page, where the full dataset can be accessed by curious users.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="applications.html#cb73-1" aria-hidden="true" tabindex="-1"></a>occurrencedf_dec<span class="sc">$</span>references <span class="ot">&lt;-</span> <span class="st">&quot;https://www.ncei.noaa.gov/archive/accession/0282699&quot;</span></span>
<span id="cb73-2"><a href="applications.html#cb73-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-3"><a href="applications.html#cb73-3" aria-hidden="true" tabindex="-1"></a>occurrencedf_dec</span></code></pre></div>
<pre><code>## # A tibble: 17 × 20
##    eventDate    decimalLatitude decimalLongitude basisOfRecord kingdom taxonRank
##    &lt;chr&gt;                  &lt;dbl&gt;            &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;   &lt;chr&gt;    
##  1 2009-09-23T…            34.0            -119. HumanObserva… Animal… Species  
##  2 2009-09-25T…            34.0            -119. MachineObser… Animal… Species  
##  3 2009-09-27T…            34.0            -119. MachineObser… Animal… Species  
##  4 2009-10-08T…            34.0            -119. MachineObser… Animal… Species  
##  5 2009-10-15T…            34.0            -119. MachineObser… Animal… Species  
##  6 2009-10-17T…            34.0            -119. MachineObser… Animal… Species  
##  7 2009-10-17T…            34.0            -119. MachineObser… Animal… Species  
##  8 2009-10-17T…            34.0            -119. MachineObser… Animal… Species  
##  9 2009-10-18T…            34.0            -119. MachineObser… Animal… Species  
## 10 2009-10-18T…            34.0            -119. MachineObser… Animal… Species  
## 11 2009-10-18T…            34.0            -119. MachineObser… Animal… Species  
## 12 2009-10-23T…            34.0            -119. MachineObser… Animal… Species  
## 13 2009-10-24T…            34.0            -119. MachineObser… Animal… Species  
## 14 2009-10-26T…            34.0            -119. MachineObser… Animal… Species  
## 15 2009-10-27T…            34.0            -119. MachineObser… Animal… Species  
## 16 2009-10-29T…            34.0            -119. MachineObser… Animal… Species  
## 17 2009-10-31T…            34.0            -119. MachineObser… Animal… Species  
## # ℹ 14 more variables: occurrenceStatus &lt;chr&gt;, sex &lt;chr&gt;, lifeStage &lt;chr&gt;,
## #   scientificName &lt;chr&gt;, scientificNameID &lt;chr&gt;, minimumDepthInMeters &lt;int&gt;,
## #   maximumDepthInMeters &lt;int&gt;, organismID &lt;chr&gt;, occurrenceID &lt;chr&gt;,
## #   geodeticDatum &lt;chr&gt;, eventID &lt;chr&gt;, coordinateUncertaintyInMeters &lt;dbl&gt;,
## #   dataGeneralizations &lt;chr&gt;, references &lt;chr&gt;</code></pre>
<div id="write-decimated-occurrence-file-as-csv" class="section level5 hasAnchor" number="2.6.2.2.1">
<h5><span class="header-section-number">2.6.2.2.1</span> Write decimated occurrence file as csv<a href="applications.html#write-decimated-occurrence-file-as-csv" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="applications.html#cb75-1" aria-hidden="true" tabindex="-1"></a>tag_id <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(variable <span class="sc">==</span> <span class="st">&quot;NC_GLOBAL&quot;</span> <span class="sc">&amp;</span> name <span class="sc">==</span> <span class="st">&quot;ptt_id&quot;</span>)</span>
<span id="cb75-2"><a href="applications.html#cb75-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-3"><a href="applications.html#cb75-3" aria-hidden="true" tabindex="-1"></a>file_name_occur <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&#39;data/dwc/atn_&#39;</span>,tag_id<span class="sc">$</span>value,<span class="st">&#39;_occurrence.csv&#39;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb75-4"><a href="applications.html#cb75-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-5"><a href="applications.html#cb75-5" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(occurrencedf_dec, <span class="at">file=</span>file_name_occur, <span class="at">row.names=</span><span class="cn">FALSE</span>, <span class="at">fileEncoding=</span><span class="st">&quot;UTF-8&quot;</span>, <span class="at">quote=</span><span class="cn">TRUE</span>, <span class="at">na=</span><span class="st">&quot;&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="measurement-or-fact" class="section level4 hasAnchor" number="2.6.2.3">
<h4><span class="header-section-number">2.6.2.3</span> Measurement or Fact<a href="applications.html#measurement-or-fact" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Since we do have any additional observations, we can create a measurement or fact file to include those data. Might be worthwhile to include tag/device metadata, some of the animal measurements, and the detachment information. Each term should have a definition URI.</p>
<p>The measurementOrFact file will only contain information referencing the <code>basisOfRecord</code> = <code>HumanObservation</code> as these observations were made when the animal was directly tagged, in person (ie. when <code>basisOfRecord</code> == <code>HumanObservation</code>).</p>
<table>
<colgroup>
<col width="51%" />
<col width="25%" />
<col width="22%" />
</colgroup>
<thead>
<tr class="header">
<th>DarwinCore Term</th>
<th>Status</th>
<th>netCDF</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>organismID</td>
<td></td>
<td>The <code>platform_id</code> global attribute plus the <code>animal_common_name</code> global attribute.</td>
</tr>
<tr class="even">
<td>occurrenceID</td>
<td>Required</td>
<td><code>eventDate</code>, plus data contained in <code>z</code> variable, plus <code>animal_common_name</code> global attribute.</td>
</tr>
<tr class="odd">
<td>measurementType</td>
<td>Required</td>
<td><code>long_name</code> attribute of the <code>animal_weight</code>, <code>animal_length</code>, <code>animal_length_2</code> variables.</td>
</tr>
<tr class="even">
<td>measurementValue</td>
<td>Required</td>
<td>The data from the <code>animal_weight</code>, <code>animal_length</code>, <code>animal_length_2</code> variables.</td>
</tr>
<tr class="odd">
<td>eventID</td>
<td>Strongly Recommended</td>
<td><code>animal_common_name</code> global attribute plus the <code>eventDate</code>.</td>
</tr>
<tr class="even">
<td>measurementUnit</td>
<td>Strongly Recommended</td>
<td><code>unit</code> attribute of the <code>animal_weight</code>, <code>animal_length</code>, <code>animal_length_2</code> variables.</td>
</tr>
<tr class="odd">
<td>measurementMethod</td>
<td>Strongly Recommended</td>
<td><code>animal_weight</code>, <code>animal_length</code>, <code>animal_length_2</code> attributes of their respective variables.</td>
</tr>
<tr class="even">
<td>measurementTypeID</td>
<td>Strongly Recommended</td>
<td>mapping table somewhere?</td>
</tr>
<tr class="odd">
<td>measurementMethodID</td>
<td>Strongly Recommended</td>
<td>mapping table somewhere?</td>
</tr>
<tr class="even">
<td>measurementUnitID</td>
<td>Strongly Recommended</td>
<td>mapping table somewhere?</td>
</tr>
<tr class="odd">
<td>measurementAccuracy</td>
<td>Share if available</td>
<td></td>
</tr>
<tr class="even">
<td>measurementDeterminedDate</td>
<td>Share if available</td>
<td></td>
</tr>
<tr class="odd">
<td>measurementDeterminedBy</td>
<td>Share if available</td>
<td></td>
</tr>
<tr class="even">
<td>measurementRemarks</td>
<td>Share if available</td>
<td></td>
</tr>
<tr class="odd">
<td>measurementValueID</td>
<td>Share if available</td>
<td></td>
</tr>
</tbody>
</table>
<div id="extracting-variables-for-extended-measurement-or-fact-emof" class="section level5 hasAnchor" number="2.6.2.3.1">
<h5><span class="header-section-number">2.6.2.3.1</span> Extracting variables for Extended Measurement Or Fact (eMOF)<a href="applications.html#extracting-variables-for-extended-measurement-or-fact-emof" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Here there are two approaches to transforming a variable to the eMOF Darwin Core extension. The goal is to collapse the measurement name, value, unit, related identifiers and remarks into a generalized long format that can be linked to occurrences and events. For more info see:</p>
<ol style="list-style-type: decimal">
<li><a href="https://manual.obis.org/format_emof.html">The OBIS manual</a></li>
<li>The <a href="https://ioos.github.io/bio_mobilization_workshop/04-create-schema/index.html">Marine Biological Data Mobilization Workshop 2023</a> (SF:Not sure if it’s cool to reference the workshop like this)</li>
</ol>
<p>The first several lines of the below code show an example of pulling out the variable attributes and individually mapping them to the eMOF terms. However, this can be done more efficiently (although less readable) via this chunk of code:</p>
<pre><code># Supply vector of variable names
c(&quot;animal_length&quot;,
  &quot;animal_length_2&quot;,
  &quot;animal_weight&quot;) %&gt;%

      # Create a named list of the variable attributes and convert it into a data frame, for each name in the above vector.
      purrr::map_df(function(x) {
        0list(measurementValue = ncvar_get( nc, x),
             measurementType = ncatt_get( nc, x)$long_name,
             measurementUnit = ncatt_get( nc, x)$units,
             measurementMethod = ncatt_get( nc, x)[[paste0(x,&#39;_type&#39;)]])
        })</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="applications.html#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # Measurement or Fact extension</span></span>
<span id="cb77-2"><a href="applications.html#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="co"># # Need to find the occurrence where basisOfRecord == HumanObservation, then pull the organism.</span></span>
<span id="cb77-3"><a href="applications.html#cb77-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-4"><a href="applications.html#cb77-4" aria-hidden="true" tabindex="-1"></a>emof_data <span class="ot">&lt;-</span> <span class="co">#var_names %&gt;%</span></span>
<span id="cb77-5"><a href="applications.html#cb77-5" aria-hidden="true" tabindex="-1"></a>    <span class="co">#filter(str_starts(name, pattern = &quot;animal_[lw]e&quot;)) %&gt;% #example using regex to parse names</span></span>
<span id="cb77-6"><a href="applications.html#cb77-6" aria-hidden="true" tabindex="-1"></a>    <span class="co">#    pull(name) %&gt;%</span></span>
<span id="cb77-7"><a href="applications.html#cb77-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-8"><a href="applications.html#cb77-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Example using vector of variables</span></span>
<span id="cb77-9"><a href="applications.html#cb77-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">c</span>(<span class="st">&quot;animal_length&quot;</span>,</span>
<span id="cb77-10"><a href="applications.html#cb77-10" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;animal_length_2&quot;</span>,</span>
<span id="cb77-11"><a href="applications.html#cb77-11" aria-hidden="true" tabindex="-1"></a>      <span class="st">&quot;animal_weight&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb77-12"><a href="applications.html#cb77-12" aria-hidden="true" tabindex="-1"></a>    purrr<span class="sc">::</span><span class="fu">map_df</span>(<span class="cf">function</span>(x) {</span>
<span id="cb77-13"><a href="applications.html#cb77-13" aria-hidden="true" tabindex="-1"></a>        <span class="fu">list</span>(<span class="at">measurementValue =</span> <span class="fu">ncvar_get</span>( nc, x),</span>
<span id="cb77-14"><a href="applications.html#cb77-14" aria-hidden="true" tabindex="-1"></a>             <span class="at">measurementType =</span> <span class="fu">ncatt_get</span>( nc, x)<span class="sc">$</span>long_name,</span>
<span id="cb77-15"><a href="applications.html#cb77-15" aria-hidden="true" tabindex="-1"></a>             <span class="at">measurementUnit =</span> <span class="fu">ncatt_get</span>( nc, x)<span class="sc">$</span>units,</span>
<span id="cb77-16"><a href="applications.html#cb77-16" aria-hidden="true" tabindex="-1"></a>             <span class="at">measurementMethod =</span> <span class="fu">ncatt_get</span>( nc, x)[[<span class="fu">paste0</span>(x,<span class="st">&#39;_type&#39;</span>)]])</span>
<span id="cb77-17"><a href="applications.html#cb77-17" aria-hidden="true" tabindex="-1"></a>    }) <span class="sc">%&gt;%</span></span>
<span id="cb77-18"><a href="applications.html#cb77-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb77-19"><a href="applications.html#cb77-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(measurementValue <span class="sc">!=</span> <span class="st">&quot;NaN&quot;</span>)</span>
<span id="cb77-20"><a href="applications.html#cb77-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-21"><a href="applications.html#cb77-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-22"><a href="applications.html#cb77-22" aria-hidden="true" tabindex="-1"></a>emofdf <span class="ot">&lt;-</span> occurrencedf <span class="sc">%&gt;%</span></span>
<span id="cb77-23"><a href="applications.html#cb77-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(basisOfRecord <span class="sc">==</span> <span class="st">&#39;HumanObservation&#39;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb77-24"><a href="applications.html#cb77-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(organismID, eventID, occurrenceID) <span class="sc">%&gt;%</span></span>
<span id="cb77-25"><a href="applications.html#cb77-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind</span>(emof_data)</span>
<span id="cb77-26"><a href="applications.html#cb77-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-27"><a href="applications.html#cb77-27" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(emofdf)</span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    1 obs. of  7 variables:
##  $ organismID       : chr &quot;105838_great_white_shark&quot;
##  $ eventID          : chr &quot;great_white shark_2009-09-23T00:00:00Z&quot;
##  $ occurrenceID     : chr &quot;2009-09-23T00:00:00Z_0_great_white_shark&quot;
##  $ measurementValue : num 213
##  $ measurementType  : chr &quot;length of the animal as measured or estimated at deployment&quot;
##  $ measurementUnit  : chr &quot;cm&quot;
##  $ measurementMethod: chr &quot;total length&quot;</code></pre>
</div>
<div id="write-emof-file-as-csv" class="section level5 hasAnchor" number="2.6.2.3.2">
<h5><span class="header-section-number">2.6.2.3.2</span> Write emof file as csv<a href="applications.html#write-emof-file-as-csv" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="applications.html#cb79-1" aria-hidden="true" tabindex="-1"></a>tag_id <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(variable <span class="sc">==</span> <span class="st">&quot;NC_GLOBAL&quot;</span>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(name <span class="sc">==</span> <span class="st">&quot;ptt_id&quot;</span>)</span>
<span id="cb79-2"><a href="applications.html#cb79-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-3"><a href="applications.html#cb79-3" aria-hidden="true" tabindex="-1"></a>file_name_emof <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&#39;data/dwc/atn_&#39;</span>,tag_id<span class="sc">$</span>value,<span class="st">&#39;_emof.csv&#39;</span>, <span class="at">sep =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb79-4"><a href="applications.html#cb79-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-5"><a href="applications.html#cb79-5" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(emofdf, <span class="at">file=</span>file_name_emof, <span class="at">row.names=</span><span class="cn">FALSE</span>, <span class="at">fileEncoding=</span><span class="st">&quot;UTF-8&quot;</span>, <span class="at">quote=</span><span class="cn">TRUE</span>, <span class="at">na=</span><span class="st">&quot;&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="metadata-creation" class="section level4 hasAnchor" number="2.6.2.4">
<h4><span class="header-section-number">2.6.2.4</span> Metadata creation<a href="applications.html#metadata-creation" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Now that we know our data are aligned to Darwin Core, we can start collecting metadata. Using the R package <a href="https://docs.ropensci.org/EML/">EML</a> we can create the EML metadata to associate with the data above.</p>
<p>Some good sources to help identify what requirements we need in the EML metadata can be found at:</p>
<ul>
<li><p><a href="https://github.com/gbif/ipt/wiki/GMPHowToGuide" class="uri">https://github.com/gbif/ipt/wiki/GMPHowToGuide</a></p></li>
<li><p><a href="https://github.com/gbif/ipt/wiki/GMPHowToGuide#dataset-resource" class="uri">https://github.com/gbif/ipt/wiki/GMPHowToGuide#dataset-resource</a></p></li>
</ul>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="applications.html#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># library(EML)</span></span></code></pre></div>
<p>The first thing we need to do is collect all of the relevant pieces of metadata for our EML record.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="applications.html#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># me &lt;- list(individualName = list(givenName = &quot;Matt&quot;, surName = &quot;Biddle&quot;))</span></span>
<span id="cb81-2"><a href="applications.html#cb81-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-3"><a href="applications.html#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="co"># my_eml &lt;- list(dataset = list(</span></span>
<span id="cb81-4"><a href="applications.html#cb81-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-5"><a href="applications.html#cb81-5" aria-hidden="true" tabindex="-1"></a><span class="co">#                           title = &quot;A Minimal Valid EML Dataset&quot;,</span></span>
<span id="cb81-6"><a href="applications.html#cb81-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-7"><a href="applications.html#cb81-7" aria-hidden="true" tabindex="-1"></a><span class="co">#                           creator = me,</span></span>
<span id="cb81-8"><a href="applications.html#cb81-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-9"><a href="applications.html#cb81-9" aria-hidden="true" tabindex="-1"></a><span class="co">#                           contact = me</span></span>
<span id="cb81-10"><a href="applications.html#cb81-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-11"><a href="applications.html#cb81-11" aria-hidden="true" tabindex="-1"></a><span class="co">#                             )</span></span>
<span id="cb81-12"><a href="applications.html#cb81-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-13"><a href="applications.html#cb81-13" aria-hidden="true" tabindex="-1"></a><span class="co">#                 )</span></span></code></pre></div>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="applications.html#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># geographicDescription &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;sea_name&quot;)</span></span>
<span id="cb82-2"><a href="applications.html#cb82-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-3"><a href="applications.html#cb82-3" aria-hidden="true" tabindex="-1"></a><span class="co"># west &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;geospatial_lon_min&quot;)</span></span>
<span id="cb82-4"><a href="applications.html#cb82-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-5"><a href="applications.html#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="co"># east &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;geospatial_lon_max&quot;)</span></span>
<span id="cb82-6"><a href="applications.html#cb82-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-7"><a href="applications.html#cb82-7" aria-hidden="true" tabindex="-1"></a><span class="co"># north &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;geospatial_lat_max&quot;)</span></span>
<span id="cb82-8"><a href="applications.html#cb82-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-9"><a href="applications.html#cb82-9" aria-hidden="true" tabindex="-1"></a><span class="co"># south &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;geospatial_lat_min&quot;)</span></span>
<span id="cb82-10"><a href="applications.html#cb82-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-11"><a href="applications.html#cb82-11" aria-hidden="true" tabindex="-1"></a><span class="co"># altitudeMin &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;geospatial_vertical_min&quot;)</span></span>
<span id="cb82-12"><a href="applications.html#cb82-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-13"><a href="applications.html#cb82-13" aria-hidden="true" tabindex="-1"></a><span class="co"># altitudeMax &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;geospatial_vertical_max&quot;)</span></span>
<span id="cb82-14"><a href="applications.html#cb82-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-15"><a href="applications.html#cb82-15" aria-hidden="true" tabindex="-1"></a><span class="co"># altitudeUnits &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;geospatial_vertical_units&quot;)</span></span>
<span id="cb82-16"><a href="applications.html#cb82-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-17"><a href="applications.html#cb82-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-18"><a href="applications.html#cb82-18" aria-hidden="true" tabindex="-1"></a><span class="co"># coverage &lt;-</span></span>
<span id="cb82-19"><a href="applications.html#cb82-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-20"><a href="applications.html#cb82-20" aria-hidden="true" tabindex="-1"></a><span class="co">#   set_coverage(begin = format(min(atn_tbl$time),&#39;%Y-%m-%d&#39;), end = format(max(atn_tbl$time), &#39;%Y-%m-%d&#39;),</span></span>
<span id="cb82-21"><a href="applications.html#cb82-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-22"><a href="applications.html#cb82-22" aria-hidden="true" tabindex="-1"></a><span class="co">#                sci_names = RNetCDF::var.get.nc(RNetCDF::open.nc(&quot;atn_trajectory_template.nc&quot;), &quot;taxon_name&quot;),</span></span>
<span id="cb82-23"><a href="applications.html#cb82-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-24"><a href="applications.html#cb82-24" aria-hidden="true" tabindex="-1"></a><span class="co">#                geographicDescription = paste(geographicDescription$value),</span></span>
<span id="cb82-25"><a href="applications.html#cb82-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-26"><a href="applications.html#cb82-26" aria-hidden="true" tabindex="-1"></a><span class="co">#                west = paste(west$value),</span></span>
<span id="cb82-27"><a href="applications.html#cb82-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-28"><a href="applications.html#cb82-28" aria-hidden="true" tabindex="-1"></a><span class="co">#                east = paste(east$value) ,</span></span>
<span id="cb82-29"><a href="applications.html#cb82-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-30"><a href="applications.html#cb82-30" aria-hidden="true" tabindex="-1"></a><span class="co">#                north = paste(north$value) ,</span></span>
<span id="cb82-31"><a href="applications.html#cb82-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-32"><a href="applications.html#cb82-32" aria-hidden="true" tabindex="-1"></a><span class="co">#                south = paste(south$value) ,</span></span>
<span id="cb82-33"><a href="applications.html#cb82-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-34"><a href="applications.html#cb82-34" aria-hidden="true" tabindex="-1"></a><span class="co">#                altitudeMin = paste(altitudeMin$value),</span></span>
<span id="cb82-35"><a href="applications.html#cb82-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-36"><a href="applications.html#cb82-36" aria-hidden="true" tabindex="-1"></a><span class="co">#                altitudeMaximum = paste(altitudeMax$value),</span></span>
<span id="cb82-37"><a href="applications.html#cb82-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-38"><a href="applications.html#cb82-38" aria-hidden="true" tabindex="-1"></a><span class="co">#                altitudeUnits = ifelse (paste(altitudeUnits$value) == &#39;m&#39;, &quot;meter&quot;, &quot;?&quot;))</span></span></code></pre></div>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="applications.html#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co"># creator_name &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;creator_name&quot;)</span></span>
<span id="cb83-2"><a href="applications.html#cb83-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-3"><a href="applications.html#cb83-3" aria-hidden="true" tabindex="-1"></a><span class="co"># creator_email &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;creator_email&quot;)</span></span>
<span id="cb83-4"><a href="applications.html#cb83-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-5"><a href="applications.html#cb83-5" aria-hidden="true" tabindex="-1"></a><span class="co"># creator_sector &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;creator_sector&quot;)</span></span>
<span id="cb83-6"><a href="applications.html#cb83-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-7"><a href="applications.html#cb83-7" aria-hidden="true" tabindex="-1"></a><span class="co"># creator &lt;- eml$creator(</span></span>
<span id="cb83-8"><a href="applications.html#cb83-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-9"><a href="applications.html#cb83-9" aria-hidden="true" tabindex="-1"></a><span class="co">#             eml$individualName(</span></span>
<span id="cb83-10"><a href="applications.html#cb83-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-11"><a href="applications.html#cb83-11" aria-hidden="true" tabindex="-1"></a><span class="co">#                 givenName = paste(creator_name$value),</span></span>
<span id="cb83-12"><a href="applications.html#cb83-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-13"><a href="applications.html#cb83-13" aria-hidden="true" tabindex="-1"></a><span class="co">#                 surName = paste(creator_name$value)</span></span>
<span id="cb83-14"><a href="applications.html#cb83-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-15"><a href="applications.html#cb83-15" aria-hidden="true" tabindex="-1"></a><span class="co">#                 ),</span></span>
<span id="cb83-16"><a href="applications.html#cb83-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-17"><a href="applications.html#cb83-17" aria-hidden="true" tabindex="-1"></a><span class="co">#             position = paste(creator_sector$value),</span></span>
<span id="cb83-18"><a href="applications.html#cb83-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-19"><a href="applications.html#cb83-19" aria-hidden="true" tabindex="-1"></a><span class="co">#             electronicMailAddress = paste(creator_email$value)</span></span>
<span id="cb83-20"><a href="applications.html#cb83-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-21"><a href="applications.html#cb83-21" aria-hidden="true" tabindex="-1"></a><span class="co">#             )</span></span></code></pre></div>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="applications.html#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co"># #contact_name = metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;contact_name&quot;)</span></span>
<span id="cb84-2"><a href="applications.html#cb84-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-3"><a href="applications.html#cb84-3" aria-hidden="true" tabindex="-1"></a><span class="co"># contact &lt;- eml$contact(</span></span>
<span id="cb84-4"><a href="applications.html#cb84-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-5"><a href="applications.html#cb84-5" aria-hidden="true" tabindex="-1"></a><span class="co">#             eml$individualName(</span></span>
<span id="cb84-6"><a href="applications.html#cb84-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-7"><a href="applications.html#cb84-7" aria-hidden="true" tabindex="-1"></a><span class="co">#             givenName = paste(creator_name$value),</span></span>
<span id="cb84-8"><a href="applications.html#cb84-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-9"><a href="applications.html#cb84-9" aria-hidden="true" tabindex="-1"></a><span class="co">#             surName = paste(creator_name$value)),</span></span>
<span id="cb84-10"><a href="applications.html#cb84-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-11"><a href="applications.html#cb84-11" aria-hidden="true" tabindex="-1"></a><span class="co">#             position = paste(creator_sector$value),</span></span>
<span id="cb84-12"><a href="applications.html#cb84-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-13"><a href="applications.html#cb84-13" aria-hidden="true" tabindex="-1"></a><span class="co">#             electronicMailAddress = paste(creator_email$value)</span></span>
<span id="cb84-14"><a href="applications.html#cb84-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-15"><a href="applications.html#cb84-15" aria-hidden="true" tabindex="-1"></a><span class="co">#             )</span></span></code></pre></div>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="applications.html#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># #metadata_name</span></span>
<span id="cb85-2"><a href="applications.html#cb85-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-3"><a href="applications.html#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="co"># metadataProvider &lt;- eml$metadataProvider(</span></span>
<span id="cb85-4"><a href="applications.html#cb85-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-5"><a href="applications.html#cb85-5" aria-hidden="true" tabindex="-1"></a><span class="co">#             eml$individualName(</span></span>
<span id="cb85-6"><a href="applications.html#cb85-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-7"><a href="applications.html#cb85-7" aria-hidden="true" tabindex="-1"></a><span class="co">#                 givenName = paste(creator_name$value),</span></span>
<span id="cb85-8"><a href="applications.html#cb85-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-9"><a href="applications.html#cb85-9" aria-hidden="true" tabindex="-1"></a><span class="co">#                 surName = paste(creator_name$value)),</span></span>
<span id="cb85-10"><a href="applications.html#cb85-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-11"><a href="applications.html#cb85-11" aria-hidden="true" tabindex="-1"></a><span class="co">#             position = paste(creator_sector$value),</span></span>
<span id="cb85-12"><a href="applications.html#cb85-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-13"><a href="applications.html#cb85-13" aria-hidden="true" tabindex="-1"></a><span class="co">#             electronicMailAddress = paste(creator_email$value)</span></span>
<span id="cb85-14"><a href="applications.html#cb85-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-15"><a href="applications.html#cb85-15" aria-hidden="true" tabindex="-1"></a><span class="co">#             )</span></span></code></pre></div>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="applications.html#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ## these are the entries in contributor, need to iterate since comma separated list.</span></span>
<span id="cb86-2"><a href="applications.html#cb86-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-3"><a href="applications.html#cb86-3" aria-hidden="true" tabindex="-1"></a><span class="co"># contrib_name &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;contributor_name&quot;)</span></span>
<span id="cb86-4"><a href="applications.html#cb86-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-5"><a href="applications.html#cb86-5" aria-hidden="true" tabindex="-1"></a><span class="co"># contrib_position &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;contributor_role&quot;)</span></span>
<span id="cb86-6"><a href="applications.html#cb86-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-7"><a href="applications.html#cb86-7" aria-hidden="true" tabindex="-1"></a><span class="co"># contrib_email &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;contributor_email&quot;)</span></span></code></pre></div>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="applications.html#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># associatedParty &lt;- eml$associatedParty(</span></span>
<span id="cb87-2"><a href="applications.html#cb87-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-3"><a href="applications.html#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="co">#                     eml$individualName(</span></span>
<span id="cb87-4"><a href="applications.html#cb87-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-5"><a href="applications.html#cb87-5" aria-hidden="true" tabindex="-1"></a><span class="co">#                     givenName = paste(contrib_name$value),</span></span>
<span id="cb87-6"><a href="applications.html#cb87-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-7"><a href="applications.html#cb87-7" aria-hidden="true" tabindex="-1"></a><span class="co">#                     surName = paste(contrib_name$value)),</span></span>
<span id="cb87-8"><a href="applications.html#cb87-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-9"><a href="applications.html#cb87-9" aria-hidden="true" tabindex="-1"></a><span class="co">#                     position = paste(contrib_position$value),</span></span>
<span id="cb87-10"><a href="applications.html#cb87-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-11"><a href="applications.html#cb87-11" aria-hidden="true" tabindex="-1"></a><span class="co">#                     electronicMailAddress = paste(contrib_email$value)</span></span>
<span id="cb87-12"><a href="applications.html#cb87-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-13"><a href="applications.html#cb87-13" aria-hidden="true" tabindex="-1"></a><span class="co">#                     )</span></span></code></pre></div>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="applications.html#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co"># abstract &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;summary&quot;)</span></span></code></pre></div>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="applications.html#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # keywords</span></span>
<span id="cb89-2"><a href="applications.html#cb89-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-3"><a href="applications.html#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="co"># keywords &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;keywords&quot;)</span></span>
<span id="cb89-4"><a href="applications.html#cb89-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-5"><a href="applications.html#cb89-5" aria-hidden="true" tabindex="-1"></a><span class="co"># kw_vocab &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;keywords_vocabulary&quot;)</span></span>
<span id="cb89-6"><a href="applications.html#cb89-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-7"><a href="applications.html#cb89-7" aria-hidden="true" tabindex="-1"></a><span class="co"># keywordSet &lt;- list(</span></span>
<span id="cb89-8"><a href="applications.html#cb89-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-9"><a href="applications.html#cb89-9" aria-hidden="true" tabindex="-1"></a><span class="co">#     list(</span></span>
<span id="cb89-10"><a href="applications.html#cb89-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-11"><a href="applications.html#cb89-11" aria-hidden="true" tabindex="-1"></a><span class="co">#         keywordThesaurus = kw_vocab$value$keywords_vocabulary,</span></span>
<span id="cb89-12"><a href="applications.html#cb89-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-13"><a href="applications.html#cb89-13" aria-hidden="true" tabindex="-1"></a><span class="co">#         keyword = as.list(strsplit(keywords$value$keywords, &quot;, &quot;))</span></span>
<span id="cb89-14"><a href="applications.html#cb89-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-15"><a href="applications.html#cb89-15" aria-hidden="true" tabindex="-1"></a><span class="co">#         ))</span></span></code></pre></div>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="applications.html#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># title &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;title&quot;)</span></span></code></pre></div>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="applications.html#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co"># methods &lt;- &quot;NEED TO MAP FROM NCFILE&quot;</span></span></code></pre></div>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="applications.html#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co"># license &lt;- metadata %&gt;% dplyr::filter(variable == &quot;NC_GLOBAL&quot;) %&gt;% dplyr::filter(name == &quot;license&quot;)</span></span></code></pre></div>
<p>Now build the eml file.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="applications.html#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co"># library(uuid)</span></span>
<span id="cb93-2"><a href="applications.html#cb93-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-3"><a href="applications.html#cb93-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-4"><a href="applications.html#cb93-4" aria-hidden="true" tabindex="-1"></a><span class="co"># physical &lt;- set_physical(file_name_occur)</span></span>
<span id="cb93-5"><a href="applications.html#cb93-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-6"><a href="applications.html#cb93-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-7"><a href="applications.html#cb93-7" aria-hidden="true" tabindex="-1"></a><span class="co"># # attributeList &lt;-</span></span>
<span id="cb93-8"><a href="applications.html#cb93-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-9"><a href="applications.html#cb93-9" aria-hidden="true" tabindex="-1"></a><span class="co"># #   set_attributes(attributes,</span></span>
<span id="cb93-10"><a href="applications.html#cb93-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-11"><a href="applications.html#cb93-11" aria-hidden="true" tabindex="-1"></a><span class="co"># #                  factors,</span></span>
<span id="cb93-12"><a href="applications.html#cb93-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-13"><a href="applications.html#cb93-13" aria-hidden="true" tabindex="-1"></a><span class="co"># #                  col_classes = c(&quot;character&quot;,</span></span>
<span id="cb93-14"><a href="applications.html#cb93-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-15"><a href="applications.html#cb93-15" aria-hidden="true" tabindex="-1"></a><span class="co"># #                                  &quot;Date&quot;,</span></span>
<span id="cb93-16"><a href="applications.html#cb93-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-17"><a href="applications.html#cb93-17" aria-hidden="true" tabindex="-1"></a><span class="co"># #                                  &quot;Date&quot;,</span></span>
<span id="cb93-18"><a href="applications.html#cb93-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-19"><a href="applications.html#cb93-19" aria-hidden="true" tabindex="-1"></a><span class="co"># #                                  &quot;Date&quot;,</span></span>
<span id="cb93-20"><a href="applications.html#cb93-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-21"><a href="applications.html#cb93-21" aria-hidden="true" tabindex="-1"></a><span class="co"># #                                  &quot;factor&quot;,</span></span>
<span id="cb93-22"><a href="applications.html#cb93-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-23"><a href="applications.html#cb93-23" aria-hidden="true" tabindex="-1"></a><span class="co"># #                                  &quot;factor&quot;,</span></span>
<span id="cb93-24"><a href="applications.html#cb93-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-25"><a href="applications.html#cb93-25" aria-hidden="true" tabindex="-1"></a><span class="co"># #                                  &quot;factor&quot;,</span></span>
<span id="cb93-26"><a href="applications.html#cb93-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-27"><a href="applications.html#cb93-27" aria-hidden="true" tabindex="-1"></a><span class="co"># #                                  &quot;numeric&quot;))</span></span>
<span id="cb93-28"><a href="applications.html#cb93-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-29"><a href="applications.html#cb93-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-30"><a href="applications.html#cb93-30" aria-hidden="true" tabindex="-1"></a><span class="co"># my_eml &lt;- eml$eml(</span></span>
<span id="cb93-31"><a href="applications.html#cb93-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-32"><a href="applications.html#cb93-32" aria-hidden="true" tabindex="-1"></a><span class="co">#            packageId = paste(uuid_tbl$value),</span></span>
<span id="cb93-33"><a href="applications.html#cb93-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-34"><a href="applications.html#cb93-34" aria-hidden="true" tabindex="-1"></a><span class="co">#            system = &quot;uuid&quot;,</span></span>
<span id="cb93-35"><a href="applications.html#cb93-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-36"><a href="applications.html#cb93-36" aria-hidden="true" tabindex="-1"></a><span class="co">#            dataset = eml$dataset(</span></span>
<span id="cb93-37"><a href="applications.html#cb93-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-38"><a href="applications.html#cb93-38" aria-hidden="true" tabindex="-1"></a><span class="co">#                alternateIdentifier = UUIDgenerate(use.time = TRUE),</span></span>
<span id="cb93-39"><a href="applications.html#cb93-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-40"><a href="applications.html#cb93-40" aria-hidden="true" tabindex="-1"></a><span class="co">#                title = title$value,</span></span>
<span id="cb93-41"><a href="applications.html#cb93-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-42"><a href="applications.html#cb93-42" aria-hidden="true" tabindex="-1"></a><span class="co">#                creator = creator,</span></span>
<span id="cb93-43"><a href="applications.html#cb93-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-44"><a href="applications.html#cb93-44" aria-hidden="true" tabindex="-1"></a><span class="co">#                metadataProvider = metadataProvider,</span></span>
<span id="cb93-45"><a href="applications.html#cb93-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-46"><a href="applications.html#cb93-46" aria-hidden="true" tabindex="-1"></a><span class="co">#                #associatedParty = associatedParty,</span></span>
<span id="cb93-47"><a href="applications.html#cb93-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-48"><a href="applications.html#cb93-48" aria-hidden="true" tabindex="-1"></a><span class="co">#                contact = contact,</span></span>
<span id="cb93-49"><a href="applications.html#cb93-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-50"><a href="applications.html#cb93-50" aria-hidden="true" tabindex="-1"></a><span class="co">#                pubDate = format(Sys.time(),&#39;%Y-%m-%d&#39;),</span></span>
<span id="cb93-51"><a href="applications.html#cb93-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-52"><a href="applications.html#cb93-52" aria-hidden="true" tabindex="-1"></a><span class="co">#                language = &quot;English&quot;,</span></span>
<span id="cb93-53"><a href="applications.html#cb93-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-54"><a href="applications.html#cb93-54" aria-hidden="true" tabindex="-1"></a><span class="co">#                intellectualRights = eml$intellectualRights(</span></span>
<span id="cb93-55"><a href="applications.html#cb93-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-56"><a href="applications.html#cb93-56" aria-hidden="true" tabindex="-1"></a><span class="co">#                                     para = &quot;To the extent possible under law, the publisher has waived all rights to these data and has dedicated them to the &lt;ulink url=\&quot;http://creativecommons.org/publicdomain/zero/1.0/legalcode\&quot;&gt;&lt;citetitle&gt;Public Domain (CC0 1.0)&lt;/citetitle&gt;&lt;/ulink&gt;. Users may copy, modify, distribute and use the work, including for commercial purposes, without restriction.&quot;</span></span>
<span id="cb93-57"><a href="applications.html#cb93-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-58"><a href="applications.html#cb93-58" aria-hidden="true" tabindex="-1"></a><span class="co">#                                     #para = paste(license$value),</span></span>
<span id="cb93-59"><a href="applications.html#cb93-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-60"><a href="applications.html#cb93-60" aria-hidden="true" tabindex="-1"></a><span class="co">#                                                            ),</span></span>
<span id="cb93-61"><a href="applications.html#cb93-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-62"><a href="applications.html#cb93-62" aria-hidden="true" tabindex="-1"></a><span class="co">#                abstract = eml$abstract(</span></span>
<span id="cb93-63"><a href="applications.html#cb93-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-64"><a href="applications.html#cb93-64" aria-hidden="true" tabindex="-1"></a><span class="co">#                                para = abstract$value$summary,</span></span>
<span id="cb93-65"><a href="applications.html#cb93-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-66"><a href="applications.html#cb93-66" aria-hidden="true" tabindex="-1"></a><span class="co">#                                        ),</span></span>
<span id="cb93-67"><a href="applications.html#cb93-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-68"><a href="applications.html#cb93-68" aria-hidden="true" tabindex="-1"></a><span class="co">#                keywordSet = keywordSet,</span></span>
<span id="cb93-69"><a href="applications.html#cb93-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-70"><a href="applications.html#cb93-70" aria-hidden="true" tabindex="-1"></a><span class="co">#                coverage = coverage,</span></span>
<span id="cb93-71"><a href="applications.html#cb93-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-72"><a href="applications.html#cb93-72" aria-hidden="true" tabindex="-1"></a><span class="co"># #                license = eml$license(</span></span>
<span id="cb93-73"><a href="applications.html#cb93-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-74"><a href="applications.html#cb93-74" aria-hidden="true" tabindex="-1"></a><span class="co"># #                            licenseName = &quot;CC0 1.0&quot;,</span></span>
<span id="cb93-75"><a href="applications.html#cb93-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-76"><a href="applications.html#cb93-76" aria-hidden="true" tabindex="-1"></a><span class="co"># #                            #licenseName = paste(license$value),</span></span>
<span id="cb93-77"><a href="applications.html#cb93-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-78"><a href="applications.html#cb93-78" aria-hidden="true" tabindex="-1"></a><span class="co"># #                            ),</span></span>
<span id="cb93-79"><a href="applications.html#cb93-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-80"><a href="applications.html#cb93-80" aria-hidden="true" tabindex="-1"></a><span class="co">#                #dataTable = eml$dataTable(</span></span>
<span id="cb93-81"><a href="applications.html#cb93-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-82"><a href="applications.html#cb93-82" aria-hidden="true" tabindex="-1"></a><span class="co">#                #  entityName = file_name_occur,</span></span>
<span id="cb93-83"><a href="applications.html#cb93-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-84"><a href="applications.html#cb93-84" aria-hidden="true" tabindex="-1"></a><span class="co">#                #  entityDescription = &quot;Occurrences&quot;,</span></span>
<span id="cb93-85"><a href="applications.html#cb93-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-86"><a href="applications.html#cb93-86" aria-hidden="true" tabindex="-1"></a><span class="co">#                #  physical = physical)</span></span>
<span id="cb93-87"><a href="applications.html#cb93-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-88"><a href="applications.html#cb93-88" aria-hidden="true" tabindex="-1"></a><span class="co">#                ))</span></span></code></pre></div>
<p>Validate EML</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="applications.html#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co"># val &lt;- eml_validate(my_eml)</span></span>
<span id="cb94-2"><a href="applications.html#cb94-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-3"><a href="applications.html#cb94-3" aria-hidden="true" tabindex="-1"></a><span class="co"># attr(val,&quot;errors&quot;)</span></span></code></pre></div>
<p>Write eml to file.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="applications.html#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="co"># file_name_eml &lt;- &#39;eml.xml&#39;</span></span>
<span id="cb95-2"><a href="applications.html#cb95-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-3"><a href="applications.html#cb95-3" aria-hidden="true" tabindex="-1"></a><span class="co"># write_eml(my_eml, file_name_eml)</span></span></code></pre></div>
<p>Raw EML</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="applications.html#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># my_eml</span></span></code></pre></div>
<div id="create-meta.xml" class="section level5 hasAnchor" number="2.6.2.4.1">
<h5><span class="header-section-number">2.6.2.4.1</span> Create meta.xml<a href="applications.html#create-meta.xml" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Below is an example of the contents of meta.xml:</p>
<pre><code>
&lt;archive xmlns=&quot;http://rs.tdwg.org/dwc/text/&quot; metadata=&quot;eml.xml&quot;&gt;

  &lt;core encoding=&quot;UTF-8&quot; fieldsTerminatedBy=&quot;\t&quot; linesTerminatedBy=&quot;\n&quot; fieldsEnclosedBy=&quot;&quot; ignoreHeaderLines=&quot;1&quot; rowType=&quot;http://rs.tdwg.org/dwc/terms/Occurrence&quot;&gt;

    &lt;files&gt;

      &lt;location&gt;occurrence.txt&lt;/location&gt;

    &lt;/files&gt;

    &lt;id index=&quot;0&quot; /&gt;

    &lt;field index=&quot;1&quot; term=&quot;http://rs.tdwg.org/dwc/terms/datasetID&quot;/&gt;

    &lt;field index=&quot;2&quot; term=&quot;http://rs.tdwg.org/dwc/terms/institutionCode&quot;/&gt;

    &lt;field index=&quot;3&quot; term=&quot;http://rs.tdwg.org/dwc/terms/collectionCode&quot;/&gt;

    &lt;field index=&quot;4&quot; term=&quot;http://rs.tdwg.org/dwc/terms/basisOfRecord&quot;/&gt;

    &lt;field index=&quot;5&quot; term=&quot;http://rs.tdwg.org/dwc/terms/occurrenceID&quot;/&gt;

    &lt;field index=&quot;6&quot; term=&quot;http://rs.tdwg.org/dwc/terms/catalogNumber&quot;/&gt;

    &lt;field index=&quot;7&quot; term=&quot;http://rs.tdwg.org/dwc/terms/occurrenceRemarks&quot;/&gt;

    &lt;field index=&quot;8&quot; term=&quot;http://rs.tdwg.org/dwc/terms/individualCount&quot;/&gt;

    &lt;field index=&quot;9&quot; term=&quot;http://rs.tdwg.org/dwc/terms/sex&quot;/&gt;

    &lt;field index=&quot;10&quot; term=&quot;http://rs.tdwg.org/dwc/terms/occurrenceStatus&quot;/&gt;

    &lt;field index=&quot;11&quot; term=&quot;http://rs.tdwg.org/dwc/terms/eventDate&quot;/&gt;

    &lt;field index=&quot;12&quot; term=&quot;http://rs.tdwg.org/dwc/terms/year&quot;/&gt;

    &lt;field index=&quot;13&quot; term=&quot;http://rs.tdwg.org/dwc/terms/decimalLatitude&quot;/&gt;

    &lt;field index=&quot;14&quot; term=&quot;http://rs.tdwg.org/dwc/terms/decimalLongitude&quot;/&gt;

    &lt;field index=&quot;15&quot; term=&quot;http://rs.tdwg.org/dwc/terms/coordinateUncertaintyInMeters&quot;/&gt;

    &lt;field index=&quot;16&quot; term=&quot;http://rs.tdwg.org/dwc/terms/scientificNameID&quot;/&gt;

    &lt;field index=&quot;17&quot; term=&quot;http://rs.tdwg.org/dwc/terms/scientificName&quot;/&gt;

  &lt;/core&gt;

  &lt;extension encoding=&quot;UTF-8&quot; fieldsTerminatedBy=&quot;\t&quot; linesTerminatedBy=&quot;\n&quot; fieldsEnclosedBy=&quot;&quot; ignoreHeaderLines=&quot;1&quot; rowType=&quot;http://rs.iobis.org/obis/terms/ExtendedMeasurementOrFact&quot;&gt;

    &lt;files&gt;

      &lt;location&gt;extendedmeasurementorfact.txt&lt;/location&gt;

    &lt;/files&gt;

    &lt;coreid index=&quot;0&quot; /&gt;

    &lt;field index=&quot;1&quot; term=&quot;http://rs.tdwg.org/dwc/terms/occurrenceID&quot;/&gt;

    &lt;field index=&quot;2&quot; term=&quot;http://rs.tdwg.org/dwc/terms/measurementType&quot;/&gt;

    &lt;field index=&quot;3&quot; term=&quot;http://rs.tdwg.org/dwc/terms/measurementValue&quot;/&gt;

    &lt;field index=&quot;4&quot; term=&quot;http://rs.tdwg.org/dwc/terms/measurementUnit&quot;/&gt;

    &lt;field index=&quot;5&quot; term=&quot;http://rs.iobis.org/obis/terms/measurementUnitID&quot;/&gt;

    &lt;field index=&quot;6&quot; term=&quot;http://rs.tdwg.org/dwc/terms/measurementDeterminedDate&quot;/&gt;

  &lt;/extension&gt;

&lt;/archive&gt;
</code></pre>
<p>Checkout <a href="https://cran.r-project.org/web/packages/XML/index.html">XML package for R</a>.</p>
<p><code>conda install -c conda-forge r-xml</code></p>
<p>Another example in this <a href="https://github.com/EDIorg/ecocomDP/blob/372c293c9e597c64db22c027debe18528d44f53b/inst/extdata/dwca_event_core/meta.xml">github repository</a>.</p>
<p>Or use the <a href="http://tools.gbif.org/dwca-assistant/">gui here</a> to create meta.xml.</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="applications.html#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co"># library(XML)</span></span>
<span id="cb98-2"><a href="applications.html#cb98-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-3"><a href="applications.html#cb98-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-4"><a href="applications.html#cb98-4" aria-hidden="true" tabindex="-1"></a><span class="co"># doc = newXMLDoc()</span></span>
<span id="cb98-5"><a href="applications.html#cb98-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-6"><a href="applications.html#cb98-6" aria-hidden="true" tabindex="-1"></a><span class="co"># archiveNode = newXMLNode(&quot;archive&quot;, attrs = c(metadata=file_name_eml), namespaceDefinitions=c(&quot;http://rs.tdwg.org/dwc/text/&quot;), doc=doc )</span></span>
<span id="cb98-7"><a href="applications.html#cb98-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-8"><a href="applications.html#cb98-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-9"><a href="applications.html#cb98-9" aria-hidden="true" tabindex="-1"></a><span class="co"># ## For the core occurrence</span></span>
<span id="cb98-10"><a href="applications.html#cb98-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-11"><a href="applications.html#cb98-11" aria-hidden="true" tabindex="-1"></a><span class="co"># coreNode = newXMLNode(&quot;core&quot;, attrs = c(encoding=&quot;UTF-8&quot;, linesTerminatedBy=&quot;\\r\\n&quot;, fieldsTerminatedBy=&quot;,&quot;, fieldsEnclosedBy=&#39;\&quot;&#39;, ignoreHeaderLines=&quot;1&quot;, rowType=&quot;http://rs.tdwg.org/dwc/terms/Occurrence&quot;), parent = archiveNode)</span></span>
<span id="cb98-12"><a href="applications.html#cb98-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-13"><a href="applications.html#cb98-13" aria-hidden="true" tabindex="-1"></a><span class="co"># filesNode = newXMLNode(&quot;files&quot;, parent = coreNode)</span></span>
<span id="cb98-14"><a href="applications.html#cb98-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-15"><a href="applications.html#cb98-15" aria-hidden="true" tabindex="-1"></a><span class="co"># locationNode = newXMLNode(&quot;location&quot;, file_name_occur, parent = filesNode)</span></span>
<span id="cb98-16"><a href="applications.html#cb98-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-17"><a href="applications.html#cb98-17" aria-hidden="true" tabindex="-1"></a><span class="co"># idnode = newXMLNode(&quot;id&quot;, attrs = c(index=&quot;9&quot;), parent = coreNode)</span></span>
<span id="cb98-18"><a href="applications.html#cb98-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-19"><a href="applications.html#cb98-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-20"><a href="applications.html#cb98-20" aria-hidden="true" tabindex="-1"></a><span class="co"># # iterate over the columns in occurrence file to create field elements</span></span>
<span id="cb98-21"><a href="applications.html#cb98-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-22"><a href="applications.html#cb98-22" aria-hidden="true" tabindex="-1"></a><span class="co"># i=0</span></span>
<span id="cb98-23"><a href="applications.html#cb98-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-24"><a href="applications.html#cb98-24" aria-hidden="true" tabindex="-1"></a><span class="co"># for (col in colnames(occurrencedf))</span></span>
<span id="cb98-25"><a href="applications.html#cb98-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-26"><a href="applications.html#cb98-26" aria-hidden="true" tabindex="-1"></a><span class="co">#     {</span></span>
<span id="cb98-27"><a href="applications.html#cb98-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-28"><a href="applications.html#cb98-28" aria-hidden="true" tabindex="-1"></a><span class="co">#     termstr = paste(&quot;http://rs.tdwg.org/dwc/terms/&quot;,col, sep=&quot;&quot;)</span></span>
<span id="cb98-29"><a href="applications.html#cb98-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-30"><a href="applications.html#cb98-30" aria-hidden="true" tabindex="-1"></a><span class="co">#     i=i+1</span></span>
<span id="cb98-31"><a href="applications.html#cb98-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-32"><a href="applications.html#cb98-32" aria-hidden="true" tabindex="-1"></a><span class="co">#     fieldnode = newXMLNode(&quot;field&quot;, attrs = c(index=i, term=termstr), parent=coreNode)</span></span>
<span id="cb98-33"><a href="applications.html#cb98-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-34"><a href="applications.html#cb98-34" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb98-35"><a href="applications.html#cb98-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-36"><a href="applications.html#cb98-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-37"><a href="applications.html#cb98-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-38"><a href="applications.html#cb98-38" aria-hidden="true" tabindex="-1"></a><span class="co"># ## for the extensions</span></span>
<span id="cb98-39"><a href="applications.html#cb98-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-40"><a href="applications.html#cb98-40" aria-hidden="true" tabindex="-1"></a><span class="co"># extensionNode = newXMLNode(&quot;extension&quot;, attrs = c(encoding=&quot;UTF-8&quot;, linesTerminatedBy=&quot;\\r\\n&quot;, fieldsTerminatedBy=&quot;,&quot;, fieldsEnclosedBy=&#39;\&quot;&#39;, ignoreHeaderLines=&quot;1&quot;, rowType=&quot;http://rs.tdwg.org/dwc/terms/Event&quot;), parent = archiveNode)</span></span>
<span id="cb98-41"><a href="applications.html#cb98-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-42"><a href="applications.html#cb98-42" aria-hidden="true" tabindex="-1"></a><span class="co"># filesNode = newXMLNode(&quot;files&quot;, parent = extensionNode)</span></span>
<span id="cb98-43"><a href="applications.html#cb98-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-44"><a href="applications.html#cb98-44" aria-hidden="true" tabindex="-1"></a><span class="co"># locationNode = newXMLNode(&quot;location&quot;, file_name_event, parent = filesNode)</span></span>
<span id="cb98-45"><a href="applications.html#cb98-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-46"><a href="applications.html#cb98-46" aria-hidden="true" tabindex="-1"></a><span class="co"># idnode = newXMLNode(&quot;id&quot;, attrs = c(index=&quot;0&quot;), parent = extensionNode)</span></span>
<span id="cb98-47"><a href="applications.html#cb98-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-48"><a href="applications.html#cb98-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-49"><a href="applications.html#cb98-49" aria-hidden="true" tabindex="-1"></a><span class="co"># # iterate over the columns in occurrence file to create field elements</span></span>
<span id="cb98-50"><a href="applications.html#cb98-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-51"><a href="applications.html#cb98-51" aria-hidden="true" tabindex="-1"></a><span class="co"># i=0</span></span>
<span id="cb98-52"><a href="applications.html#cb98-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-53"><a href="applications.html#cb98-53" aria-hidden="true" tabindex="-1"></a><span class="co"># for (col in colnames(eventdf))</span></span>
<span id="cb98-54"><a href="applications.html#cb98-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-55"><a href="applications.html#cb98-55" aria-hidden="true" tabindex="-1"></a><span class="co">#     {</span></span>
<span id="cb98-56"><a href="applications.html#cb98-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-57"><a href="applications.html#cb98-57" aria-hidden="true" tabindex="-1"></a><span class="co">#     if (col == &#39;modified&#39;){</span></span>
<span id="cb98-58"><a href="applications.html#cb98-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-59"><a href="applications.html#cb98-59" aria-hidden="true" tabindex="-1"></a><span class="co">#         termstr = paste(&quot;http://purl.org/dc/terms/&quot;, col, sep=&quot;&quot;)</span></span>
<span id="cb98-60"><a href="applications.html#cb98-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-61"><a href="applications.html#cb98-61" aria-hidden="true" tabindex="-1"></a><span class="co">#     } else {</span></span>
<span id="cb98-62"><a href="applications.html#cb98-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-63"><a href="applications.html#cb98-63" aria-hidden="true" tabindex="-1"></a><span class="co">#         termstr = paste(&quot;http://rs.tdwg.org/dwc/terms/&quot;,col, sep=&quot;&quot;)</span></span>
<span id="cb98-64"><a href="applications.html#cb98-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-65"><a href="applications.html#cb98-65" aria-hidden="true" tabindex="-1"></a><span class="co">#         }</span></span>
<span id="cb98-66"><a href="applications.html#cb98-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-67"><a href="applications.html#cb98-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-68"><a href="applications.html#cb98-68" aria-hidden="true" tabindex="-1"></a><span class="co">#     i=i+1</span></span>
<span id="cb98-69"><a href="applications.html#cb98-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-70"><a href="applications.html#cb98-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-71"><a href="applications.html#cb98-71" aria-hidden="true" tabindex="-1"></a><span class="co">#     fieldnode = newXMLNode(&quot;field&quot;, attrs = c(index=i, term=termstr), parent=extensionNode)</span></span>
<span id="cb98-72"><a href="applications.html#cb98-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-73"><a href="applications.html#cb98-73" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span>
<span id="cb98-74"><a href="applications.html#cb98-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-75"><a href="applications.html#cb98-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-76"><a href="applications.html#cb98-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-77"><a href="applications.html#cb98-77" aria-hidden="true" tabindex="-1"></a><span class="co"># print(doc)</span></span>
<span id="cb98-78"><a href="applications.html#cb98-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-79"><a href="applications.html#cb98-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-80"><a href="applications.html#cb98-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-81"><a href="applications.html#cb98-81" aria-hidden="true" tabindex="-1"></a><span class="co"># saveXML(doc, file=&quot;meta.xml&quot;)</span></span></code></pre></div>
</div>
</div>
<div id="build-the-darwincore-archive-zip-package" class="section level4 hasAnchor" number="2.6.2.5">
<h4><span class="header-section-number">2.6.2.5</span> Build the DarwinCore-Archive zip package<a href="applications.html#build-the-darwincore-archive-zip-package" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="applications.html#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># library(zip)</span></span>
<span id="cb99-2"><a href="applications.html#cb99-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-3"><a href="applications.html#cb99-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-4"><a href="applications.html#cb99-4" aria-hidden="true" tabindex="-1"></a><span class="co"># files = c(file_name_occur, file_name_event, file_name_eml, &quot;meta.xml&quot;)</span></span>
<span id="cb99-5"><a href="applications.html#cb99-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-6"><a href="applications.html#cb99-6" aria-hidden="true" tabindex="-1"></a><span class="co"># zip::zip(</span></span>
<span id="cb99-7"><a href="applications.html#cb99-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-8"><a href="applications.html#cb99-8" aria-hidden="true" tabindex="-1"></a><span class="co">#     &quot;atn.zip&quot;,</span></span>
<span id="cb99-9"><a href="applications.html#cb99-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-10"><a href="applications.html#cb99-10" aria-hidden="true" tabindex="-1"></a><span class="co">#     files,</span></span>
<span id="cb99-11"><a href="applications.html#cb99-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-12"><a href="applications.html#cb99-12" aria-hidden="true" tabindex="-1"></a><span class="co">#     root = &quot;.&quot;,</span></span>
<span id="cb99-13"><a href="applications.html#cb99-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-14"><a href="applications.html#cb99-14" aria-hidden="true" tabindex="-1"></a><span class="co">#     mode = &quot;mirror&quot;,</span></span>
<span id="cb99-15"><a href="applications.html#cb99-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-16"><a href="applications.html#cb99-16" aria-hidden="true" tabindex="-1"></a><span class="co"># )</span></span>
<span id="cb99-17"><a href="applications.html#cb99-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-18"><a href="applications.html#cb99-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-19"><a href="applications.html#cb99-19" aria-hidden="true" tabindex="-1"></a><span class="co"># zip_list(&quot;atn.zip&quot;)</span></span></code></pre></div>
</div>
</div>
<div id="sessioninfo" class="section level3 hasAnchor" number="2.6.3">
<h3><span class="header-section-number">2.6.3</span> sessionInfo()<a href="applications.html#sessioninfo" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="applications.html#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.1.1 (2021-08-10)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] mapdata_2.3.1   maps_3.4.1      ncdf4_1.21      obistools_0.1.0
##  [5] tidync_0.3.0    worrms_0.4.3    magrittr_2.0.3  knitr_1.42     
##  [9] here_1.0.1      lubridate_1.9.3 forcats_1.0.0   stringr_1.5.0  
## [13] dplyr_1.1.2     purrr_1.0.1     readr_2.1.5     tidyr_1.3.0    
## [17] tibble_3.2.1    ggplot2_3.4.2   tidyverse_2.0.0
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.10       rprojroot_2.0.4   digest_0.6.31     utf8_1.2.3       
##  [5] R6_2.5.1          cellranger_1.1.0  evaluate_0.21     pillar_1.9.0     
##  [9] rlang_1.1.1       curl_5.0.0        readxl_1.4.3      ncmeta_0.3.5     
## [13] rstudioapi_0.16.0 jquerylib_0.1.4   rmarkdown_2.21    urltools_1.7.3   
## [17] htmlwidgets_1.6.2 bit_4.0.5         triebeard_0.4.1   munsell_0.5.0    
## [21] compiler_4.1.1    xfun_0.39         pkgconfig_2.0.3   htmltools_0.5.5  
## [25] tidyselect_1.2.1  httpcode_0.3.0    bookdown_0.40     fansi_1.0.4      
## [29] crayon_1.5.2      tzdb_0.3.0        withr_2.5.0       rappdirs_0.3.3   
## [33] crul_1.5.0        grid_4.1.1        jsonlite_1.8.4    gtable_0.3.3     
## [37] lifecycle_1.0.3   scales_1.2.1      cli_3.6.1         stringi_1.7.12   
## [41] vroom_1.6.3       cachem_1.0.8      leaflet_2.1.2     xml2_1.3.6       
## [45] bslib_0.4.2       generics_0.1.3    vctrs_0.6.5       data.tree_1.1.0  
## [49] tools_4.1.1       bit64_4.0.5       glue_1.6.2        RNetCDF_2.6-2    
## [53] crosstalk_1.2.0   hms_1.1.3         parallel_4.1.1    fastmap_1.1.1    
## [57] yaml_2.3.7        timechange_0.3.0  colorspace_2.1-0  sass_0.4.6</code></pre>

</div>
</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="intro.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="frequently-asked-questions.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": true,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": {}
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/ioos/bio_data_guide/edit/main/bio-data-guide/02-application.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/ioos/bio_data_guide/blob/main/bio-data-guide/02-application.Rmd",
"text": null
},
"download": ["darwin-core-guide.pdf", "darwin-core-guide.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
},
"info": false
});
});
</script>

</body>

</html>
