% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{book}
\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Standardizing-Marine-Biological-Data},
  pdfauthor={Standardizing Marine Biological Data Working Group},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newlength{\cslentryspacingunit} % times entry-spacing
\setlength{\cslentryspacingunit}{\parskip}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
  \let\oldpar\par
  \def\par{\hangindent=\cslhangindent\oldpar}
  \fi
  % set entry spacing
  \setlength{\parskip}{#2\cslentryspacingunit}
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}
\usepackage{booktabs}
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi

\title{Standardizing-Marine-Biological-Data}
\author{\href{https://github.com/ioos/bio_data_guide/graphs/contributors}{Standardizing Marine Biological Data Working Group}}
\date{2021-10-29}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{1}
\tableofcontents
}
\hypertarget{preface}{%
\chapter*{Preface}\label{preface}}
\addcontentsline{toc}{chapter}{Preface}

Biological data structures, definitions, measurements, and linkages are neccessarily as diverse as the systems they represent. This presents a real challenge when integrating data across biological research domains such as ecology, oceanography, fisheries, and climate sciences.

\hypertarget{intro}{%
\chapter{Introduction}\label{intro}}

The world of standardizing marine biological data can seem complex for the naive oceanographer, biologist, scientist, or programmer.
Transforming and integrating data is about combining the right standards for your desired interoperability with other data types.
For example, interoperating fish biology measurements with climate level variables.
There are a few concepts necessary to make this possible such as standard data structures, controlled vocabularies and knowledge representations, along with metadata standards to facilitate data discovery. This will permit the inclusion of more data and broader access to better ecosystem based models. Many scientific domains data handling practices are currently being reshaped in light of recent advances in computing power, technology, and data science.

\hypertarget{data-structures}{%
\section{Data Structures}\label{data-structures}}

The OBIS-ENV Darwin Core Archive Data Structure.

\href{https://obis.org/manual/}{OBIS manual}

\hypertarget{ontologies}{%
\section{Ontologies}\label{ontologies}}

An ontology is a classification system for establishing a hierarchically related set of concepts. Concepts are often terms from controlled vocabularies.

From Marine Metadata: \# TODO: add link

``Ontologies can include all of the following, but are not required to include them, depending on which perspective from above you adhere to:

Classes (general things, types of things)
Instances (individual things)
Relationships among things
Properties of things
Functions, processes, constraints, and rules relating to things''

TODO: Research Unified Modelling Language?

\href{http://www.obofoundry.org/ontology/envo.html}{Environment Ontology (EnvO)} EnvO is a community ontology for the concise, controlled description of environments.

\hypertarget{controlled-vocabularies}{%
\section{Controlled Vocabularies}\label{controlled-vocabularies}}

There are a number of controlled vocabularies that are used to describe parameters commonly used in specific research domains. This allows for greater interoperability of data sets within the domain, and ideally between domains. Here, we strive to document a number of relevant examples.

\begin{itemize}
\item
  \href{http://cfconventions.org/standard-names.html}{Climate and Format (CF) Standard Names} The purpose of the standard\_name attribute is to provide a succinct and distinguishing description of a variable, in a way that encourages interoperability. These terms are typically for physical observations, however, there have been advancements in aligning biological taxa into the CF standard names (see \href{https://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html\#taxon-names-and-identifiers}{here}).
\item
  \href{http://vocab.nerc.ac.uk/}{NERC Vocabulary Server} The NVS gives access to standardised and hierarchically-organized vocabularies.

  \begin{itemize}
  \item
    \href{http://vocab.nerc.ac.uk/collection/L05/current/}{Device categories} using the SeaDataNet device categories
  \item
    \href{http://vocab.nerc.ac.uk/collection/L22/current/}{Device make/model using the SeaVoX Device Catalogue}
  \item
    \href{http://vocab.nerc.ac.uk/collection/L06/current/}{Platform categories using SeaVoX Platform Categories}
  \item
    \href{http://vocab.nerc.ac.uk/collection/C17/current/}{Platform instances using the ICES Platform Codes}
  \item
    \href{http://vocab.nerc.ac.uk/collection/P06/current/}{Unit of measure}
  \end{itemize}
\item
  \href{https://wiki.earthdata.nasa.gov/display/CMR/GCMD+Keyword+Access}{GCMD Keywords (NASA)}
\item
  \href{http://vocab.nerc.ac.uk/collection/C19/current/}{Geographic Domain/Features of Interest}
\item
  \href{http://schema.geolink.org/1.0/base/main.html}{GeoLink base ontology} was part of the \href{http://www.geolink.org/}{EarthCube GeoLink Project}
\end{itemize}

Note: To describe a measurement or fact of a biological specimen that conforms to Darwin Core standards, it's necessary to use the `Biological entity described elsewhere' method rather than taxon specific.

\hypertarget{taxonomy}{%
\subsection{Taxonomy}\label{taxonomy}}

\begin{itemize}
\tightlist
\item
  \href{https://www.marinespecies.org/}{The World Registry of Marine Species (WoRMS)} The aim of a World Register of Marine Species (WoRMS) is to provide an authoritative and comprehensive list of names of marine organisms, including information on synonymy. While the highest priority goes to valid names, other names in use are included so that this register can serve as a guide to interpret taxonomic literature.
\end{itemize}

\hypertarget{resources}{%
\subsection{Resources}\label{resources}}

\hypertarget{oceanography}{%
\subsection{Oceanography}\label{oceanography}}

\href{http://www.bco-dmo.org/}{Biological and Chemcial Oceanography Data Management Office}

\href{https://mmisw.org/ont/\#/}{Marine metadata interoperability vocab resources}

\hypertarget{biology}{%
\subsection{Biology}\label{biology}}

\href{http://bioportal.bioontology.org/ontologies/ECSO}{BioPortal Ecosystem Ontology}

\hypertarget{nerc-search-interfaces}{%
\subsection{NERC Search Interfaces}\label{nerc-search-interfaces}}

\begin{itemize}
\item
  \href{http://seadatanet.maris2.nl/v_bodc_vocab_v2/welcome.asp}{SeaDatanet Common Vocab Search Interface:}
\item
  \href{https://www.seadatanet.org/Standards/Common-Vocabularies/}{SeaDataNet Common Vocabularies:}
\item
  \href{http://seadatanet.maris2.nl/v_bodc_vocab_v2/vocab_relations.asp?lib=P08}{SeaDataNet Vocab Library}
\item
  \href{https://mof.obis.org/}{Measurement Types in OBIS}
\end{itemize}

\hypertarget{geosciences}{%
\subsection{Geosciences}\label{geosciences}}

\href{https://www.unidata.ucar.edu/software/udunits/}{UDUNITS}are more common unit measurements in geosciences

\hypertarget{ecoenvo}{%
\subsection{Eco/EnvO}\label{ecoenvo}}

\href{http://www.obofoundry.org/ontology/envo.html}{Environment Ontology} including genomics.

\hypertarget{wild-cards}{%
\subsection{Wild Cards}\label{wild-cards}}

Question: Not sure use case for this.

\href{https://www.bodc.ac.uk/resources/vocabularies/vocabulary_builder/biomodel/}{P01 Biological Entity Parameter Code Builder}

\hypertarget{technologies}{%
\section{Technologies}\label{technologies}}

\hypertarget{erddap}{%
\subsection{ERDDAP}\label{erddap}}

\href{https://coastwatch.pfeg.noaa.gov/erddap/index.html}{ERDDAP} can be thought of as a data server. It provides `easier access to scientific data' by providing a consistent interface that aggregates many disparate data sources. It does this by providing translation services between many common file types for gridded arrarys (`net CDF' files) and tabular data (spreadsheets). Data access is also made easier because it unifies different types of data servers and access protocols. \href{https://github.com/HakaiInstitute/erddap-basic}{Here} is a basic erddap installation that walks you through how to load a data set.

\hypertarget{notes-on-integrating-obis-darwin-core-as-it-relates-to-ooss}{%
\section{Notes on Integrating OBIS, Darwin Core as it relates to OOS's}\label{notes-on-integrating-obis-darwin-core-as-it-relates-to-ooss}}

\hypertarget{metadata}{%
\section{Metadata}\label{metadata}}

OBIS uses the \href{http://rs.gbif.org/schema/eml-gbif-profile/1.1/eml-gbif-profile.xsd}{GBIF EML profile} (version 1.1). In case data providers use ISO19115/ISO19139, there is a mapping available here: \url{http://rs.gbif.org/schema/eml-gbif-profile/1.1/eml2iso19139.xsl} This will be important for integrating OBIS datasets to other CIOOS and IOOS metadata profiles.

\hypertarget{data-qc}{%
\section{Data QC}\label{data-qc}}

There are a number of tools available to check the quality of data or check your data format against the expected standard.

\href{https://obis.org/manual/processing/}{OBIS Datatools} shows some great R packages for this.

\hypertarget{compliance-checking}{%
\subsection{Compliance Checking}\label{compliance-checking}}

LifeWatch Belgium provides a number of tools to check your data against.
Specifically you can test OBIS data format and see a map of your sample locations to check if they are on land.
See \url{http://www.lifewatch.be/data-services/}

There's also the \href{https://www.gbif.org/tools/data-validator}{GBIF data validator} which allows anyone with a GBIF-relevant dataset to receive a report on the syntactical correctness and the validity of the content contained within the dataset.

\hypertarget{semantic-web-and-darwin-core}{%
\subsection{Semantic Web and Darwin Core}\label{semantic-web-and-darwin-core}}

\href{http://www.semantic-web-journal.net/system/files/swj1093.pdf}{Lessons learned from adapting the Darwin Core vocabulary standard for use in RDF}

\hypertarget{resource-description-framework}{%
\subsection{Resource Description Framework}\label{resource-description-framework}}

\href{https://dwc.tdwg.org/rdf/}{Darwin Core Resource Description Framework Guide}

\hypertarget{applications}{%
\chapter{Applications}\label{applications}}

Some \emph{significant} applications are demonstrated in this chapter.

\hypertarget{salmon-ocean-ecology-data}{%
\section{Salmon Ocean Ecology Data}\label{salmon-ocean-ecology-data}}

\hypertarget{intro-1}{%
\subsection{Intro}\label{intro-1}}

One of the goals of the Hakai Institute and the Canadian Integrated Ocean Observing System (CIOOS) is to facilitate Open Science and FAIR (findable, accessible, interoperable, reusable) ecological and oceanographic data. In a concerted effort to adopt or establish how best to do that, several Hakai and CIOOS staff attended an International Ocean Observing System (IOOS) Code Sprint in Ann Arbour, Michigan between October 7--11, 2019, to discuss how to implement FAIR data principles for biological data collected in the marine environment.

The \href{https://dwc.tdwg.org}{Darwin Core} is a highly structured data format that standardizes data table relations, vocabularies, and defines field names. The Darwin Core defines three table types: \texttt{event}, \texttt{occurrence}, and \texttt{measurementOrFact}. This intuitively captures the way most ecologists conduct their research. Typically, a survey (event) is conducted and measurements, counts, or observations (collectively measurementOrFacts) are made regarding a specific habitat or species (occurrence).

In the following script I demonstrate how I go about converting a subset of the data collected from the Hakai Institute Juvenile Salmon Program and discuss challenges, solutions, pros and cons, and when and what's worthwhile to convert to Darwin Core.

The conversion of a dataset to Darwin Core is much easier if your data are already tidy (normalized) in which you represent your data in separate tables that reflect the hierarchical and related nature of your observations. If your data are not already in a consistent and structured format, the conversion would likely be very arduous and not intuitive.

\hypertarget{event}{%
\subsection{event}\label{event}}

The first step is to consider what you will define as an event in your data set. I defined the capture of fish using a purse seine net as the \texttt{event}. Therefore, each row in the \texttt{event} table is one deployment of a seine net and is assigned a unique \texttt{eventID}.

My process for conversion was to make a new table called \texttt{event} and map the standard Darwin Core column names to pre-existing columns that serve the same purpose in my original \texttt{seine\_data} table and populate the other required fields.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\#TODO: Include abiotic measurements (YSI temp and salinity from 0 and 1 m) to hang off eventID in the eMoF table}

\NormalTok{event }\OtherTok{\textless{}{-}} \FunctionTok{tibble}\NormalTok{(}\AttributeTok{datasetName =} \StringTok{"Hakai Institute Juvenile Salmon Program"}\NormalTok{,}
                \AttributeTok{eventID =}\NormalTok{ survey\_seines}\SpecialCharTok{$}\NormalTok{seine\_id,}
                \AttributeTok{eventDate =} \FunctionTok{date}\NormalTok{(survey\_seines}\SpecialCharTok{$}\NormalTok{survey\_date),}
                \AttributeTok{eventTime =} \FunctionTok{paste0}\NormalTok{(survey\_seines}\SpecialCharTok{$}\NormalTok{set\_time, }\StringTok{"{-}0700"}\NormalTok{),}
                \AttributeTok{eventRemarks =} \FunctionTok{paste3}\NormalTok{(survey\_seines}\SpecialCharTok{$}\NormalTok{survey\_comments, survey\_seines}\SpecialCharTok{$}\NormalTok{seine\_comments),}
                \AttributeTok{decimalLatitude =}\NormalTok{ survey\_seines}\SpecialCharTok{$}\NormalTok{lat,}
                \AttributeTok{decimalLongitude =}\NormalTok{ survey\_seines}\SpecialCharTok{$}\NormalTok{long,}
                \AttributeTok{locationID =}\NormalTok{ survey\_seines}\SpecialCharTok{$}\NormalTok{site\_id,}
                \AttributeTok{coordinatePrecision =} \FloatTok{0.00001}\NormalTok{,}
                \AttributeTok{coordinateUncertaintyInMeters =} \DecValTok{10}\NormalTok{,}
                \AttributeTok{country =} \StringTok{"Canada"}\NormalTok{,}
                \AttributeTok{countryCode =} \StringTok{"CA"}\NormalTok{,}
                \AttributeTok{stateProvince =} \StringTok{"British Columbia"}\NormalTok{,}
                \AttributeTok{habitat =} \StringTok{"Nearshore marine"}\NormalTok{,}
                \AttributeTok{geodeticDatum =} \StringTok{"EPSG:4326 WGS84"}\NormalTok{,}
                \AttributeTok{minimumDepthInMeters =} \DecValTok{0}\NormalTok{,}
                \AttributeTok{maximumDepthInMeters =} \DecValTok{9}\NormalTok{, }\CommentTok{\# seine depth is 9 m}
                \AttributeTok{samplingProtocol =} \StringTok{"http://dx.doi.org/10.21966/1.566666"}\NormalTok{, }\CommentTok{\# This is the DOI for the Hakai Salmon Data Package that contains the smnpling protocol, as well as the complete data package}
                \AttributeTok{language =} \StringTok{"en"}\NormalTok{,}
                \AttributeTok{license =} \StringTok{"http://creativecommons.org/licenses/by/4.0/legalcode"}\NormalTok{,}
                \AttributeTok{bibliographicCitation =} \StringTok{"Johnson, B.T., J.C.L. Gan, S.C. Godwin, M. Krkosek, B.P.V. Hunt. 2020. Hakai Juvenile Salmon Program Time Series. Hakai Institute, Quadra Island Ecological Observatory, Heriot Bay, British Columbia, Canada. v\#.\#.\#, http://dx.doi.org/10.21966/1.566666"}\NormalTok{,}
                \AttributeTok{references =} \StringTok{"https://github.com/HakaiInstitute/jsp{-}data"}\NormalTok{,}
                \AttributeTok{institutionID =} \StringTok{"https://www.gbif.org/publisher/55897143{-}3f69{-}42f1{-}810d{-}ae94b55fde24, https://oceanexpert.org/institution/20121, https://edmo.seadatanet.org/report/5148"}\NormalTok{,}
                \AttributeTok{institutionCode =} \StringTok{"Hakai"}
\NormalTok{               ) }
\end{Highlighting}
\end{Shaded}

\hypertarget{occurrence}{%
\subsection{occurrence}\label{occurrence}}

Next you'll want to determine what constitutes an occurrence for your data set. Because each event captures fish, I consider each fish to be an occurrence. Therefore, the unit of observation (each row) in the occurrence table is a fish. To link each occurrence to an event you need to include the \texttt{eventID} column for every occurrence so that you know what seine (event) each fish (occurrence) came from. You must also provide a globally unique identifier for each occurrence. I already have a locally unique identifier for each fish in the original \texttt{fish\_data} table called \texttt{ufn}. To make it globally unique I pre-pend the organization and research program metadata to the \texttt{ufn} column.

Not every fish is actually collected and given a Universal Fish Number (UFN) in our fish data tables, so in our field data sheets we record the total number of fish captured and the total number retained. So to get an occurrence row for every fish captured I create a row for every fish caught (minus the number taken) and create a generic numeric id (ie hakai-jsp-1) in one table and then join that to the fish table that includes a row for every fish retained that already has a UFN.

\begin{Shaded}
\begin{Highlighting}[]
\DocumentationTok{\#\# make table long first}
\NormalTok{seines\_total\_long }\OtherTok{\textless{}{-}}\NormalTok{ survey\_seines }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(seine\_id, so\_total, pi\_total, cu\_total, co\_total, he\_total, ck\_total) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{pivot\_longer}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{seine\_id, }\AttributeTok{names\_to =} \StringTok{"scientificName"}\NormalTok{, }\AttributeTok{values\_to =} \StringTok{"n"}\NormalTok{)}

\NormalTok{seines\_total\_long}\SpecialCharTok{$}\NormalTok{scientificName }\OtherTok{\textless{}{-}} \FunctionTok{recode}\NormalTok{(seines\_total\_long}\SpecialCharTok{$}\NormalTok{scientificName, }\AttributeTok{so\_total =} \StringTok{"Oncorhynchus nerka"}\NormalTok{, }\AttributeTok{pi\_total =} \StringTok{"Oncorhynchus gorbuscha"}\NormalTok{, }\AttributeTok{cu\_total =} \StringTok{"Oncorhynchus keta"}\NormalTok{, }\AttributeTok{co\_total =} \StringTok{"Oncorhynchus kisutch"}\NormalTok{, }\AttributeTok{ck\_total =} \StringTok{"Oncorhynchus tshawytscha"}\NormalTok{, }\AttributeTok{he\_total =} \StringTok{"Clupea pallasii"}\NormalTok{) }

\NormalTok{seines\_taken\_long }\OtherTok{\textless{}{-}}\NormalTok{ survey\_seines }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(seine\_id, so\_taken, pi\_taken, cu\_taken, co\_taken, he\_taken, ck\_taken) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{pivot\_longer}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{seine\_id, }\AttributeTok{names\_to =} \StringTok{"scientificName"}\NormalTok{, }\AttributeTok{values\_to =} \StringTok{"n\_taken"}\NormalTok{) }

\NormalTok{seines\_taken\_long}\SpecialCharTok{$}\NormalTok{scientificName }\OtherTok{\textless{}{-}} \FunctionTok{recode}\NormalTok{(seines\_taken\_long}\SpecialCharTok{$}\NormalTok{scientificName, }\AttributeTok{so\_taken =} \StringTok{"Oncorhynchus nerka"}\NormalTok{, }\AttributeTok{pi\_taken =} \StringTok{"Oncorhynchus gorbuscha"}\NormalTok{, }\AttributeTok{cu\_taken =} \StringTok{"Oncorhynchus keta"}\NormalTok{, }\AttributeTok{co\_taken =} \StringTok{"Oncorhynchus kisutch"}\NormalTok{, }\AttributeTok{ck\_taken =} \StringTok{"Oncorhynchus tshawytscha"}\NormalTok{, }\AttributeTok{he\_taken =} \StringTok{"Clupea pallasii"}\NormalTok{) }

\DocumentationTok{\#\# remove records that have already been assigned an ID because they were actually retained}
\NormalTok{seines\_long }\OtherTok{\textless{}{-}}  \FunctionTok{full\_join}\NormalTok{(seines\_total\_long, seines\_taken\_long, }\AttributeTok{by =} \FunctionTok{c}\NormalTok{(}\StringTok{"seine\_id"}\NormalTok{, }\StringTok{"scientificName"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{drop\_na}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{n\_not\_taken =}\NormalTok{ n }\SpecialCharTok{{-}}\NormalTok{ n\_taken) }\SpecialCharTok{\%\textgreater{}\%} \CommentTok{\#so\_total includes the number taken so I subtract n\_taken to get n\_not\_taken}
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{n\_taken, }\SpecialCharTok{{-}}\NormalTok{n) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{filter}\NormalTok{(n\_not\_taken }\SpecialCharTok{\textgreater{}} \DecValTok{0}\NormalTok{)}

\NormalTok{all\_fish\_not\_retained }\OtherTok{\textless{}{-}}
\NormalTok{  seines\_long[}\FunctionTok{rep}\NormalTok{(}\FunctionTok{seq.int}\NormalTok{(}\DecValTok{1}\NormalTok{, }\FunctionTok{nrow}\NormalTok{(seines\_long)), seines\_long}\SpecialCharTok{$}\NormalTok{n\_not\_taken), }\DecValTok{1}\SpecialCharTok{:}\DecValTok{3}\NormalTok{] }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{n\_not\_taken) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{prefix =} \StringTok{"hakai{-}jsp{-}"}\NormalTok{,}
         \AttributeTok{suffix =} \DecValTok{1}\SpecialCharTok{:}\FunctionTok{nrow}\NormalTok{(.),}
         \AttributeTok{occurrenceID =} \FunctionTok{paste0}\NormalTok{(prefix, suffix)}
\NormalTok{  ) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{prefix, }\SpecialCharTok{{-}}\NormalTok{suffix)}

\CommentTok{\#}

\CommentTok{\# Change species names to full Scientific names }
\NormalTok{latin }\OtherTok{\textless{}{-}} \FunctionTok{fct\_recode}\NormalTok{(fish\_data}\SpecialCharTok{$}\NormalTok{species, }\StringTok{"Oncorhynchus nerka"} \OtherTok{=} \StringTok{"SO"}\NormalTok{, }\StringTok{"Oncorhynchus gorbuscha"} \OtherTok{=} \StringTok{"PI"}\NormalTok{, }\StringTok{"Oncorhynchus keta"} \OtherTok{=} \StringTok{"CU"}\NormalTok{, }\StringTok{"Oncorhynchus kisutch"} \OtherTok{=} \StringTok{"CO"}\NormalTok{, }\StringTok{"Clupea pallasii"} \OtherTok{=} \StringTok{"HE"}\NormalTok{, }\StringTok{"Oncorhynchus tshawytscha"} \OtherTok{=} \StringTok{"CK"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{as.character}\NormalTok{()}

\NormalTok{fish\_retained\_data }\OtherTok{\textless{}{-}}\NormalTok{ fish\_data }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{scientificName =}\NormalTok{ latin) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{species) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{prefix =} \StringTok{"hakai{-}jsp{-}"}\NormalTok{,}
         \AttributeTok{occurrenceID =} \FunctionTok{paste0}\NormalTok{(prefix, ufn)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(seine\_id, scientificName, occurrenceID)}

\NormalTok{occurrence }\OtherTok{\textless{}{-}} \FunctionTok{bind\_rows}\NormalTok{(all\_fish\_not\_retained, fish\_retained\_data) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{rename}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ seine\_id) }\SpecialCharTok{\%\textgreater{}\%}  \CommentTok{\# rename = dplyr::rename; vs plyr::rename}
  \FunctionTok{mutate}\NormalTok{(}\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}} \OtherTok{=} \StringTok{"juvenile"}\NormalTok{)}

\NormalTok{unique\_taxa }\OtherTok{\textless{}{-}} \FunctionTok{unique}\NormalTok{(occurrence}\SpecialCharTok{$}\NormalTok{scientificName)  }
\NormalTok{worms\_names }\OtherTok{\textless{}{-}} \FunctionTok{wm\_records\_names}\NormalTok{(unique\_taxa) }\CommentTok{\# library(worrms)}
\NormalTok{df\_worms\_names }\OtherTok{\textless{}{-}} \FunctionTok{bind\_rows}\NormalTok{(worms\_names) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\AttributeTok{scientificName =}\NormalTok{ scientificname,}
         \AttributeTok{scientificNameAuthorship =}\NormalTok{ authority,}
         \AttributeTok{taxonRank =}\NormalTok{ rank,}
         \AttributeTok{scientificNameID =}\NormalTok{ lsid}
\NormalTok{         )}

\CommentTok{\#include bycatch species}

\NormalTok{unique\_bycatch }\OtherTok{\textless{}{-}} \FunctionTok{unique}\NormalTok{(bycatch}\SpecialCharTok{$}\NormalTok{scientificName) }\SpecialCharTok{\%\textgreater{}\%}  \FunctionTok{glimpse}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##  chr [1:29] "Oncorhynchus nerka" "Oncorhynchus tshawytscha" ...
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{by\_worms\_names }\OtherTok{\textless{}{-}} \FunctionTok{wm\_records\_names}\NormalTok{(unique\_bycatch) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{bind\_rows}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\AttributeTok{scientificName =}\NormalTok{ scientificname,}
         \AttributeTok{scientificNameAuthorship =}\NormalTok{ authority,}
         \AttributeTok{taxonRank =}\NormalTok{ rank,}
         \AttributeTok{scientificNameID =}\NormalTok{ lsid}
\NormalTok{         )}

\NormalTok{bycatch\_occurrence }\OtherTok{\textless{}{-}}\NormalTok{ bycatch }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\AttributeTok{eventID =}\NormalTok{ seine\_id, occurrenceID, scientificName, }\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}} \OtherTok{=}\NormalTok{ bm\_ageclass) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{filter}\NormalTok{(scientificName }\SpecialCharTok{!=} \StringTok{"unknown"}\NormalTok{)}

\NormalTok{bycatch\_occurrence}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}}\NormalTok{[bycatch\_occurrence}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}} \SpecialCharTok{==} \StringTok{"J"}\NormalTok{] }\OtherTok{\textless{}{-}} \StringTok{"juvenile"}
\NormalTok{bycatch\_occurrence}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}}\NormalTok{[bycatch\_occurrence}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}} \SpecialCharTok{==} \StringTok{"A"}\NormalTok{] }\OtherTok{\textless{}{-}} \StringTok{"adult"}
\NormalTok{bycatch\_occurrence}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}}\NormalTok{[bycatch\_occurrence}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}} \SpecialCharTok{==} \StringTok{"Y"}\NormalTok{] }\OtherTok{\textless{}{-}} \StringTok{"Young of year"}

\NormalTok{combined\_worms\_names }\OtherTok{\textless{}{-}} \FunctionTok{bind\_rows}\NormalTok{(by\_worms\_names, df\_worms\_names) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{distinct}\NormalTok{(scientificName, }\AttributeTok{.keep\_all =} \ConstantTok{TRUE}\NormalTok{)}

\NormalTok{occurrence }\OtherTok{\textless{}{-}} \FunctionTok{bind\_rows}\NormalTok{(bycatch\_occurrence, occurrence)}

\NormalTok{occurrence }\OtherTok{\textless{}{-}} \FunctionTok{left\_join}\NormalTok{(occurrence, combined\_worms\_names) }\SpecialCharTok{\%\textgreater{}\%} 
    \FunctionTok{mutate}\NormalTok{(}\AttributeTok{basisOfRecord =} \StringTok{"HumanObservation"}\NormalTok{,}
        \AttributeTok{occurrenceStatus =} \StringTok{"present"}\NormalTok{)}

\FunctionTok{write\_csv}\NormalTok{(occurrence,}\StringTok{"../datasets/hakai\_salmon\_data/raw\_data/occurrence.csv"}\NormalTok{) }\CommentTok{\# here::here("..", "datasets", "hakai\_salmon\_data", "raw\_data",   "occurrence.csv"))}

\CommentTok{\# This removes events that didn\textquotesingle{}t result in any occurrences}
\NormalTok{event }\OtherTok{\textless{}{-}}\NormalTok{ dplyr}\SpecialCharTok{::}\FunctionTok{semi\_join}\NormalTok{(event, occurrence, }\AttributeTok{by =} \StringTok{\textquotesingle{}eventID\textquotesingle{}}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{coordinateUncertaintyInMeters =} \FunctionTok{ifelse}\NormalTok{(}\FunctionTok{is.na}\NormalTok{(decimalLatitude), }\DecValTok{1852}\NormalTok{, coordinateUncertaintyInMeters))}

\NormalTok{simple\_sites }\OtherTok{\textless{}{-}}\NormalTok{ sites }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(site\_id, ocgy\_std\_lat, ocgy\_std\_lon)}

\NormalTok{event }\OtherTok{\textless{}{-}}\NormalTok{ dplyr}\SpecialCharTok{::}\FunctionTok{left\_join}\NormalTok{(event, simple\_sites, }\AttributeTok{by =} \FunctionTok{c}\NormalTok{(}\StringTok{"locationID"} \OtherTok{=} \StringTok{"site\_id"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{decimalLatitude =} \FunctionTok{coalesce}\NormalTok{(decimalLatitude, ocgy\_std\_lat),}
         \AttributeTok{decimalLongitude =} \FunctionTok{coalesce}\NormalTok{(decimalLongitude, ocgy\_std\_lon)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\FunctionTok{c}\NormalTok{(ocgy\_std\_lat, ocgy\_std\_lon))}

\FunctionTok{write\_csv}\NormalTok{(event,}\StringTok{"../datasets/hakai\_salmon\_data/raw\_data/event.csv"}\NormalTok{) }\CommentTok{\# here::here("..", "datasets", "hakai\_salmon\_data", "raw\_data",   "event.csv"))}
\end{Highlighting}
\end{Shaded}

\hypertarget{measurementorfact}{%
\subsection{measurementOrFact}\label{measurementorfact}}

To convert all your measurements or facts from your normal format to Darwin Core you essentially need to put all your measurements into one column called measurementType and a corresponding column called MeasurementValue. This standardizes the column names are in the \texttt{measurementOrFact} table. There are a number of predefined \texttt{measurementType}s listed on the \href{https://www.bodc.ac.uk/resources/vocabularies/}{NERC} database that should be used where possible. I found it difficult to navigate this page to find the correct \texttt{measurementType}.

Here I convert length, and weight measurements that relate to an event and an occurrence and call those \texttt{measurementTypes} as \texttt{length} and \texttt{weight}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mof\_types }\OtherTok{\textless{}{-}} \FunctionTok{read\_csv}\NormalTok{(}\StringTok{"https://raw.githubusercontent.com/HakaiInstitute/jsp{-}data/master/OBIS\_data/mof\_type\_units\_id.csv"}\NormalTok{)}

\NormalTok{fish\_data}\SpecialCharTok{$}\NormalTok{weight }\OtherTok{\textless{}{-}} \FunctionTok{coalesce}\NormalTok{(fish\_data}\SpecialCharTok{$}\NormalTok{weight, fish\_data}\SpecialCharTok{$}\NormalTok{weight\_field)}
\NormalTok{fish\_data}\SpecialCharTok{$}\NormalTok{fork\_length }\OtherTok{\textless{}{-}} \FunctionTok{coalesce}\NormalTok{(fish\_data}\SpecialCharTok{$}\NormalTok{fork\_length, fish\_data}\SpecialCharTok{$}\NormalTok{fork\_length\_field)}
\NormalTok{fish\_data}\SpecialCharTok{$}\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}} \OtherTok{\textless{}{-}} \StringTok{"juvenile"}



\NormalTok{measurementOrFact }\OtherTok{\textless{}{-}}\NormalTok{ fish\_data }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{occurrenceID =} \FunctionTok{paste0}\NormalTok{(}\StringTok{"hakai{-}jsp{-}"}\NormalTok{, ufn)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{select}\NormalTok{(occurrenceID, }\AttributeTok{eventID =}\NormalTok{ seine\_id, }\StringTok{"Length (fork length)"} \OtherTok{=}\NormalTok{ fork\_length,}
         \StringTok{"Standard length"} \OtherTok{=}\NormalTok{ standard\_length, }\StringTok{"Weight"} \OtherTok{=}\NormalTok{ weight, }\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{pivot\_longer}\NormalTok{(}\StringTok{\textasciigrave{}}\AttributeTok{Length (fork length)}\StringTok{\textasciigrave{}}\SpecialCharTok{:}\StringTok{\textasciigrave{}}\AttributeTok{Life stage}\StringTok{\textasciigrave{}}\NormalTok{,}
               \AttributeTok{names\_to =} \StringTok{"measurementType"}\NormalTok{,}
               \AttributeTok{values\_to =} \StringTok{"measurementValue"}\NormalTok{,}
               \AttributeTok{values\_transform =} \FunctionTok{list}\NormalTok{(}\AttributeTok{measurementValue =}\NormalTok{ as.character)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(measurementValue }\SpecialCharTok{!=} \StringTok{"NA"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{left\_join}\NormalTok{(mof\_types,}\AttributeTok{by =} \FunctionTok{c}\NormalTok{(}\StringTok{"measurementType"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{measurementValueID =} \FunctionTok{case\_when}\NormalTok{(measurementValue }\SpecialCharTok{==} \StringTok{"juvenile"} \SpecialCharTok{\textasciitilde{}} \StringTok{"http://vocab.nerc.ac.uk/collection/S11/current/S1127/"}\NormalTok{),}
         \AttributeTok{measurementID =} \FunctionTok{paste}\NormalTok{(eventID, measurementType, occurrenceID, }\AttributeTok{sep =} \StringTok{"{-}"}\NormalTok{))}


\FunctionTok{write\_csv}\NormalTok{(measurementOrFact,}\StringTok{"../datasets/hakai\_salmon\_data/raw\_data/extendedMeasurementOrFact.csv"}\NormalTok{) }\CommentTok{\# here::here("..", "datasets", "hakai\_salmon\_data", "raw\_data",   "extendedMeasurementOrFact.csv"))}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\#Check that every eventID in Occurrence occurs in event table}
\NormalTok{no\_keys }\OtherTok{\textless{}{-}} \FunctionTok{dm}\NormalTok{(event, occurrence, measurementOrFact)}
\NormalTok{only\_pk }\OtherTok{\textless{}{-}}\NormalTok{ no\_keys }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{dm\_add\_pk}\NormalTok{(event, eventID) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{dm\_add\_pk}\NormalTok{(occurrence, occurrenceID) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{dm\_add\_pk}\NormalTok{(measurementOrFact, measurementID)}
\FunctionTok{dm\_examine\_constraints}\NormalTok{(only\_pk)}

\NormalTok{model }\OtherTok{\textless{}{-}}\NormalTok{ only\_pk }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{dm\_add\_fk}\NormalTok{(occurrence, eventID, event) }\SpecialCharTok{\%\textgreater{}\%} 
  \FunctionTok{dm\_add\_fk}\NormalTok{(measurementOrFact, occurrenceID, occurrence)}
\FunctionTok{dm\_examine\_constraints}\NormalTok{(model)}

\CommentTok{\#TODO: Fix bookdown issues so that dm\_draw shows data model html output. Perhaps add  \textasciigrave{}always\_allow\_html: true\textasciigrave{} to yaml front matter}
\CommentTok{\# dm\_draw(model, view\_type = "all") }
\end{Highlighting}
\end{Shaded}

\hypertarget{example-two}{%
\section{Example Two}\label{example-two}}

Add another example here, perhaps zooplankton?

\hypertarget{final-words}{%
\chapter{Final Words}\label{final-words}}

We have finished a nice book.

\hypertarget{tools}{%
\chapter{Tools}\label{tools}}

Below are some of the tools and packages used in workflows. R and Python package ``Type'' is BIO for packages specifically for biological applications, and GEN for generic packages.

\hypertarget{r}{%
\section{R}\label{r}}

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.28}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.39}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.33}}@{}}
\toprule
\begin{minipage}[b]{\linewidth}\raggedright
Package
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Type
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Description
\end{minipage} \\
\midrule
\endhead
\href{https://bdverse.org/}{bdveRse} & BIO & A family of R packages for biodiversity data. \\
\href{https://cran.r-project.org/web/packages/ecocomDP/index.html}{ecocomDP} & BIO & Work with the Ecological Community Data Design Pattern. `ecocomDP' is a flexible data model for harmonizing ecological community surveys, in a research question agnostic format, from source data published across repositories, and with methods that keep the derived data up-to-date as the underlying sources change. \\
\href{https://ediorg.github.io/EMLassemblyline/}{EDIorg/EMLasseblyline} & BIO & For scientists and data managers to create high quality EML metadata for dataset publication. \\
\href{https://cran.r-project.org/web/packages/finch/index.html}{finch} & BIO & Parse Darwin Core Files \\
\href{https://iobis.github.io/obistools/}{iobis/obistools} & BIO & Tools for data enhancement and quality control. \\
\href{https://cran.r-project.org/web/packages/robis/index.html}{robis} & BIO & R client for the OBIS API \\
\href{https://docs.ropensci.org/EML/}{ropensci/EML} & BIO & Provides support for the serializing and parsing of all low-level EML concepts \\
\href{https://cran.r-project.org/web/packages/taxize/index.html}{taxize} & BIO & Interacts with a suite of web `APIs' for taxonomic tasks, such as getting database specific taxonomic identifiers, verifying species names, getting taxonomic hierarchies, fetching downstream and upstream taxonomic names, getting taxonomic synonyms, converting scientific to common names and vice versa, and more. \\
\href{https://cran.r-project.org/web/packages/worrms/index.html}{worrms} & BIO & Client for \href{http://www.marinespecies.org/}{World Register of Marine Species}. Includes functions for each of the API methods, including searching for names by name, date and common names, searching using external identifiers, fetching synonyms, as well as fetching taxonomic children and taxonomic classification. \\
\href{https://www.rdocumentation.org/packages/Hmisc/versions/4.6-0}{Hmisc} & GEN & Contains many functions useful for data analysis, high-level graphics, utility operations, functions for computing sample size and power, simulation, importing and annotating datasets, imputing missing values, advanced table making, variable clustering, character string manipulation, conversion of R objects to LaTeX and html code, and recoding variables. Particularly check out the \href{https://www.rdocumentation.org/packages/Hmisc/versions/4.6-0/topics/describe}{describe()} function. \\
\href{https://cran.r-project.org/web/packages/lubridate/index.html}{lubridate} & GEN & Functions to work with date-times and time-spans: fast and user friendly parsing of date-time data, extraction and updating of components of a date-time (years, months, days, hours, minutes, and seconds), algebraic manipulation on date-time and time-span objects. \\
\href{https://cran.r-project.org/web/packages/stringr/index.html}{stringr} & GEN & Simple, Consistent Wrappers for Common String Operations \\
\href{https://cran.r-project.org/web/packages/tidyverse/index.html}{tidyverse} & GEN & The `tidyverse' is a set of packages that work in harmony because they share common data representations and `API' design. This package is designed to make it easy to install and load multiple `tidyverse' packages in a single step. \\
\href{https://cran.r-project.org/web/packages/uuid/index.html}{uuid} & GEN & Tools for generating and handling of UUIDs (Universally Unique Identifiers). \\
\bottomrule
\end{longtable}

\hypertarget{python}{%
\section{Python}\label{python}}

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.28}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.39}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.33}}@{}}
\toprule
\begin{minipage}[b]{\linewidth}\raggedright
Package
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Type
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Description
\end{minipage} \\
\midrule
\endhead
\href{https://pypi.org/project/metapype/}{metapype} & BIO & A lightweight Python 3 library for generating EML metadata \\
\href{https://python-dwca-reader.readthedocs.io/en/latest/index.html}{python-dwca-reader} & BIO & A simple Python package to read and parse Darwin Core Archive (DwC-A) files, as produced by the GBIF website, the IPT and many other biodiversity informatics tools. \\
\href{https://github.com/iobis/pyworms}{pyworms} & BIO & Python client for the World Register of Marine Species (WoRMS) REST service. \\
\href{https://numpy.org/}{numpy} & GEN & NumPy (Numerical Python) is an open source Python library that's used in almost every field of science and engineering. It's the universal standard for working with numerical data in Python, and it's at the core of the scientific Python and PyData ecosystems. \\
\href{https://pandas.pydata.org/}{pandas} & GEN & pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language. Super helpful when manipulating tabular data! \\
\href{https://docs.python.org/3/library/uuid.html}{uuid} & GEN & This module provides immutable UUID objects (class UUID) and the functions uuid1(), uuid3(), uuid4(), uuid5() for generating version 1, 3, 4, and 5 UUIDs as specified in RFC 4122. Built in -- part of the Python standard library. \\
\bottomrule
\end{longtable}

\hypertarget{google-sheets}{%
\section{Google Sheets}\label{google-sheets}}

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 2\tabcolsep) * \real{0.40}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 2\tabcolsep) * \real{0.60}}@{}}
\toprule
\begin{minipage}[b]{\linewidth}\raggedright
Package
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Description
\end{minipage} \\
\midrule
\endhead
\href{https://dwcaassistant.com/}{Google Sheet DarwinCore Archive Assistant add-on} & Google Sheet add-on which assists the creation of Darwin Core Archives (DwCA) and publising to Zenodo. DwCA's are stored into user's Google Drive and can be downloaded for upload into IPT installations or other software which is able to read DwC-archives. \\
\bottomrule
\end{longtable}

\hypertarget{validators}{%
\section{Validators}\label{validators}}

\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 2\tabcolsep) * \real{0.29}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 2\tabcolsep) * \real{0.71}}@{}}
\toprule
\begin{minipage}[b]{\linewidth}\raggedright
Name
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
Description
\end{minipage} \\
\midrule
\endhead
\href{https://tools.gbif.org/dwca-validator/}{Darwin Core Archive Validator} & This validator verifies the structural integrity of a Darwin Core Archive. It does not check the data values, such as coordinates, dates or scientific names. \\
\href{https://www.gbif.org/tools/data-validator}{GBIF DATA VALIDATOR} & The GBIF data validator is a service that allows anyone with a GBIF-relevant dataset to receive a report on the syntactical correctness and the validity of the content contained within the dataset. \\
\href{https://www.lifewatch.be/data-services/}{LifeWatch Belgium} & Through this interactive section of the LifeWatch.be portal users can upload their own data using a standard data format, and choose from several web services, models and applications to process the data. \\
\bottomrule
\end{longtable}

\hypertarget{references}{%
\chapter*{References}\label{references}}
\addcontentsline{toc}{chapter}{References}

\hypertarget{refs}{}
\begin{CSLReferences}{1}{0}
\leavevmode\vadjust pre{\hypertarget{ref-R-rmarkdown}{}}%
Allaire, JJ, Yihui Xie, Jonathan McPherson, Javier Luraschi, Kevin Ushey, Aron Atkins, Hadley Wickham, Joe Cheng, Winston Chang, and Richard Iannone. 2021. \emph{Rmarkdown: Dynamic Documents for r}. \url{https://CRAN.R-project.org/package=rmarkdown}.

\leavevmode\vadjust pre{\hypertarget{ref-Barbier2017}{}}%
Barbier, Edward B. 2017. {``Marine Ecosystem Services.''} \emph{Current Biology}. Cell Press. \url{https://doi.org/10.1016/j.cub.2017.03.020}.

\leavevmode\vadjust pre{\hypertarget{ref-Benson2018}{}}%
Benson, Abigail, Cassandra M. Brooks, Gabrielle Canonico, Emmett Duffy, Frank Muller-Karger, Heidi M. Sosik, Patricia Miloslavich, and Eduardo Klein. 2018. {``Integrated Observations and Informatics Improve Understanding of Changing Marine Ecosystems.''} \emph{Frontiers in Marine Science}. Frontiers Media S.A. \url{https://doi.org/10.3389/fmars.2018.00428}.

\leavevmode\vadjust pre{\hypertarget{ref-benson_lascala-gruenewald_mcguinn_satterthwaite_beaulieu_biddle_dewitt_mckinzie_montes_moustahfid_etal_2021}{}}%
Benson, Abigail, Diana LaScala-Gruenewald, Robert McGuinn, Erin Satterthwaite, Stace Beaulieu, Mathew Biddle, Lynn deWitt, et al. 2021. {``Biological Observation Data Standardization - a Primer for Data Managers.''} ESIP. \url{https://doi.org/10.6084/m9.figshare.16806712.v1}.

\leavevmode\vadjust pre{\hypertarget{ref-article}{}}%
Benson, Abigail, Tylar Murray, Gabrielle Canonico, Enrique Montes, Frank Muller-Karger, Maria T. Kavanaugh, Joaquin Trinanes, and Lynn M. deWitt. 2021. {``Data Management and Interactive Visualizations for the Evolving Marine Biodiversity Observation Network.''} \emph{Oceanography}. \url{https://doi.org/10.5670/oceanog.2021.220}.

\leavevmode\vadjust pre{\hypertarget{ref-Borer2009}{}}%
Borer, Elizabeth T., Eric W. Seabloom, Matthew B. Jones, and Mark Schildhauer. 2009. {``Some Simple Guidelines for Effective Data Management.''} \emph{Bulletin of the Ecological Society of America} 90 (April). \url{https://doi.org/10.1890/0012-9623-90.2.205}.

\leavevmode\vadjust pre{\hypertarget{ref-Canonico2019}{}}%
Canonico, Gabrielle, Pier Luigi Buttigieg, Enrique Montes, Frank E. Muller-Karger, Carol Stepien, Dawn Wright, Abigail Benson, et al. 2019. {``Global Observational Needs and Resources for Marine Biodiversity.''} \emph{Frontiers in Marine Science}. Frontiers Media S.A. \url{https://doi.org/10.3389/fmars.2019.00367}.

\leavevmode\vadjust pre{\hypertarget{ref-Crystal-Ornela2021}{}}%
Crystal-Ornelas, Robert, Charuleka Varadharajan, Ben Bond-Lamberty, Kristin Boye, Madison Burrus, Shreyas Cholia, Michael Crow, et al. 2021. {``A Guide to Using GitHub for Developing and Versioning Data Standards and Reporting Formats.''} \emph{Earth and Space Science}, July, e2021EA001797. \url{https://doi.org/10.1029/2021EA001797}.

\leavevmode\vadjust pre{\hypertarget{ref-Davies2021}{}}%
Davies, Neil, John Deck, Eric C Kansa, Sarah Whitcher Kansa, John Kunze, Christopher Meyer, Thomas Orrell, et al. 2021. {``Internet of Samples (iSamples): Toward an Interdisciplinary Cyberinfrastructure for Material Samples.''} \emph{GigaScience} 10: 1--5. \url{https://doi.org/10.1093/gigascience/giab028}.

\leavevmode\vadjust pre{\hypertarget{ref-Djurhuus2020}{}}%
Djurhuus, Anni, Collin J. Closek, Ryan P. Kelly, Kathleen J. Pitz, Reiko P. Michisaki, Hilary A. Starks, Kristine R. Walz, et al. 2020. {``Environmental DNA Reveals Seasonal Shifts and Potential Interactions in a Marine Community.''} \emph{Nature Communications} 11 (December): 1--9. \url{https://doi.org/10.1038/s41467-019-14105-1}.

\leavevmode\vadjust pre{\hypertarget{ref-Duffy2013}{}}%
Duffy, J. Emmett, Linda A. Amaral-Zettler, Daphne G. Fautin, Gustav Paulay, Tatiana A. Rynearson, Heidi M. Sosik, and John J. Stachowicz. 2013. {``Envisioning a Marine Biodiversity Observation Network.''} \emph{BioScience} 63 (May): 350--61. \url{https://doi.org/10.1525/bio.2013.63.5.8}.

\leavevmode\vadjust pre{\hypertarget{ref-Fornwall2012}{}}%
Fornwall, M, R Gisiner, S E Simmons, H Moustahfid, G Canonico, P Halpin, P Goldstein, et al. 2012. {``Expanding Biological Data Standards Development Processes for US IOOS: Visual Line Transect Observing Community for Mammal, Bird, and Turtle Data.''} IOOS. \url{https://www.researchgate.net/publication/255681522}.

\leavevmode\vadjust pre{\hypertarget{ref-Hardisty2019}{}}%
Hardisty, Alex R., William K. Michener, Donat Agosti, Enrique Alonso Garca, Lucy Bastin, Lee Belbin, Anne Bowser, et al. 2019. {``The Bari Manifesto: An Interoperability Framework for Essential Biodiversity Variables.''} \emph{Ecological Informatics} 49 (January): 22--31. \url{https://doi.org/10.1016/j.ecoinf.2018.11.003}.

\leavevmode\vadjust pre{\hypertarget{ref-Heberling2021}{}}%
Heberling, J Mason, Joseph T Miller, Daniel Noesgaard, Scott B Weingart C, Dmitry Schigel, and Douglas E Soltis. 2021. {``Data Integration Enables Global Biodiversity Synthesis.''} \emph{Proceedings of the National Academy of Sciences of the United States of America}. \url{https://doi.org/10.1073/pnas.2018093118/-/DCSupplemental}.

\leavevmode\vadjust pre{\hypertarget{ref-Hare2016PLOS}{}}%
Jonathan A. Hare, Mark W. Nelson, Wendy E. Morrison. n.d. {``A Vulnerability Assessment of Fish and Invertebrates to Climate Change on the Northeast u.s. Continental Shelf.''} \emph{PLoS ONE} 11 (2): e0146756. \url{https://doi.org/10.1371/journal.pone.0146756}.

\leavevmode\vadjust pre{\hypertarget{ref-Jones2006}{}}%
Jones, Matthew B., Mark P. Schildhauer, O. J. Reichman, and Shawn Bowers. 2006. {``The New Bioinformatics: Integrating Ecological Data from the Gene to the Biosphere.''} \emph{Annual Review of Ecology, Evolution, and Systematics}. \url{https://doi.org/10.1146/annurev.ecolsys.37.091305.110031}.

\leavevmode\vadjust pre{\hypertarget{ref-Kavanaugh2016}{}}%
Kavanaugh, Maria T., Matthew J. Oliver, Francisco P. Chavez, Ricardo M. Letelier, Frank E. Muller-Karger, and Scott C. Doney. 2016. {``Seascapes as a New Vernacular for Pelagic Ocean Monitoring, Management and Conservation.''} \emph{ICES Journal of Marine Science} 73 (July): 1839--50. \url{https://doi.org/10.1093/icesjms/fsw086}.

\leavevmode\vadjust pre{\hypertarget{ref-Kot2010}{}}%
Kot, Connie Y., Ei Fujioka, Lucie J. Hazen, Benjamin D. Best, Andrew J. Read, and Patrick N. Halpin. 2010. {``Spatio-Temporal Gap Analysis of OBIS-SEAMAP Project Data: Assessment and Way Forward.''} \emph{PLoS ONE} 5 (September): 12990. \url{https://doi.org/10.1371/journal.pone.0012990}.

\leavevmode\vadjust pre{\hypertarget{ref-OceanAdapt}{}}%
Lab, Malin Pinksy. n.d. {``OceanAdapt.''} \emph{GitHub}. \url{https://oceanadapt.rutgers.edu/}.

\leavevmode\vadjust pre{\hypertarget{ref-Lamprecht2019}{}}%
Lamprecht, Anna-Lena, Leyla Garcia, Mateusz Kuzak, Carlos Martinez, Ricardo Arcila, Eva Martin Del Pico, Victoria Dominguez Del Angel, et al. 2019. {``Towards FAIR Principles for Research Software.''} Edited by Paul Groth. \emph{Data Science}, 1--23. \url{https://doi.org/10.3233/DS-190026}.

\leavevmode\vadjust pre{\hypertarget{ref-Dornelas2014Science}{}}%
Maria Dornelas, Brian McGill, Nicholas J. Gotelli. 2014. {``Assemblage Time Series Reveal Biodiversity Change but Not Systematic Loss.''} \emph{Science} 344 (6181): 296--99. \url{https://doi.org/10.1126/science.1248484}.

\leavevmode\vadjust pre{\hypertarget{ref-McKenna2021}{}}%
McKenna, Megan F., Simone Baumann-Pickering, Annebelle C. M. Kok, William K. Oestreich, Jeffrey D. Adams, Jack Barkowski, Kurt M. Fristrup, et al. 2021. {``Advancing the Interpretation of Shallow Water Marine Soundscapes.''} \emph{Frontiers in Marine Science} 0 (September): 1426. \url{https://doi.org/10.3389/FMARS.2021.719258}.

\leavevmode\vadjust pre{\hypertarget{ref-Miloslavich2018}{}}%
Miloslavich, Patricia, Nicholas J. Bax, Samantha E. Simmons, Eduardo Klein, Ward Appeltans, Octavio Aburto-Oropeza, Melissa Andersen Garcia, et al. 2018. {``Essential Ocean Variables for Global Sustained Observations of Biodiversity and Ecosystem Changes.''} \emph{Global Change Biology} 24 (June): 2416--33. \url{https://doi.org/10.1111/gcb.14108}.

\leavevmode\vadjust pre{\hypertarget{ref-Montes2020}{}}%
Montes, Enrique, Anni Djurhuus, Frank E. Muller-Karger, Daniel Otis, Christopher R. Kelble, and Maria T. Kavanaugh. 2020. {``Dynamic Satellite Seascapes as a Biogeographic Framework for Understanding Phytoplankton Assemblages in the Florida Keys National Marine Sanctuary, United States.''} \emph{Frontiers in Marine Science} 7 (July): 575. \url{https://doi.org/10.3389/fmars.2020.00575}.

\leavevmode\vadjust pre{\hypertarget{ref-Moustahfid2014}{}}%
Moustahfid, Hassan, and Philip Goldstein. 2014. {``IOOS Biological Data Services Enrollment Procedures.''}

\leavevmode\vadjust pre{\hypertarget{ref-Moustahfid2011}{}}%
Moustahfid, Hassan, Jim Potemra, Philip Goldstein, Roy Mendelssohn, and Annette Desrochers. 2011. {``Making United States Integrated Ocean Observing System (u.s. IOOS) Inclusive of Marine Biological Resources.''} \url{https://www.researchgate.net/publication/254013004}.

\leavevmode\vadjust pre{\hypertarget{ref-Muller-Karger2018b}{}}%
Muller-Karger, Frank E., Erin Hestir, Christiana Ade, Kevin Turpie, Dar A. Roberts, David Siegel, Robert J. Miller, et al. 2018. {``Satellite Sensor Requirements for Monitoring Essential Biodiversity Variables of Coastal Ecosystems.''} \emph{Ecological Applications} 28 (April): 749--60. \url{https://doi.org/10.1002/eap.1682}.

\leavevmode\vadjust pre{\hypertarget{ref-Muller-Karger2018a}{}}%
Muller-Karger, Frank E., Patricia Miloslavich, Nicholas J. Bax, Samantha Simmons, Mark J. Costello, Isabel Sousa Pinto, Gabrielle Canonico, et al. 2018. {``Advancing Marine Biological Observations and Data Requirements of the Complementary Essential Ocean Variables (EOVs) and Essential Biodiversity Variables (EBVs) Frameworks.''} \emph{Frontiers in Marine Science}. Frontiers Media S.A. \url{https://doi.org/10.3389/fmars.2018.00211}.

\leavevmode\vadjust pre{\hypertarget{ref-Neeley2021}{}}%
Neeley, Aimee, Stace E. Beaulieu, Chris Proctor, Ivona Cetini, Joe Futrelle, Inia Soto Ramos, Heidi M. Sosik, et al. 2021. {``Standards and Practices for Reporting Plankton and Other Particle Observations from Images.''} \url{https://doi.org/10.1575/1912/27377}.

\leavevmode\vadjust pre{\hypertarget{ref-Obrien2021}{}}%
O'Brien, Margaret, Colin A. Smith, Eric R. Sokol, Corinna Gries, Nina Lany, Sydne Record, and Max C. N. Castorani. 2021. {``ecocomDP: A Flexible Data Design Pattern for Ecological Community Survey Data.''} \emph{Ecological Informatics} 64 (September): 101374. \url{https://doi.org/10.1016/J.ECOINF.2021.101374}.

\leavevmode\vadjust pre{\hypertarget{ref-Pooter2017}{}}%
Pooter, Daphnis De, Ward Appeltans, Nicolas Bailly, Sky Bristol, Klaas Deneudt, Menash Eliezer, Ei Fujioka, et al. 2017. {``Toward a New Data Standard for Combined Marine Biological and Environmental Datasets - Expanding OBIS Beyond Species Occurrences.''} \emph{Biodiversity Data Journal} 5 (January): 10989. \url{https://doi.org/10.3897/BDJ.5.e10989}.

\leavevmode\vadjust pre{\hypertarget{ref-iobis_ebsa}{}}%
Provoost, Pieter. n.d. {``Iobis/Ebsa.''} \emph{GitHub}. \url{https://github.com/iobis/ebsa}.

\leavevmode\vadjust pre{\hypertarget{ref-R-base}{}}%
R Core Team. 2021. \emph{R: A Language and Environment for Statistical Computing}. Vienna, Austria: R Foundation for Statistical Computing. \url{https://www.R-project.org/}.

\leavevmode\vadjust pre{\hypertarget{ref-Ruxfccknagel2015}{}}%
Rcknagel, J., P. Vierkant, R. Ulrich, G. Kloska, E. Schnepf, D. Fichtmller, E. Reuter, et al. 2015. {``Metadata Schema for the Description of Research Data Repositories. Version 3.0.''} \url{https://doi.org/10.2312/re3.008}.

\leavevmode\vadjust pre{\hypertarget{ref-Rule2019}{}}%
Rule, Adam, Amanda Birmingham, Cristal Zuniga, Ilkay Altintas, Shih-Cheng Huang, Rob Knight, Niema Moshiri, et al. 2019. {``Ten Simple Rules for Writing and Sharing Computational Analyses in Jupyter Notebooks.''} \emph{PLOS Computational Biology} 15 (July): e1007007. \url{https://doi.org/10.1371/JOURNAL.PCBI.1007007}.

\leavevmode\vadjust pre{\hypertarget{ref-Santora2017}{}}%
Santora, Jarrod A., Elliott L. Hazen, Isaac D. Schroeder, Steven J. Bograd, Keith M. Sakuma, and John C. Field. 2017. {``Impacts of Ocean Climate Variability on Biodiversity of Pelagic Forage Species in an Upwelling Ecosystem.''} \emph{Marine Ecology Progress Series} 580 (September): 205--20. \url{https://doi.org/10.3354/meps12278}.

\leavevmode\vadjust pre{\hypertarget{ref-Schmid_Daprano_Jacobson_Sullivan_Briseuxf1o-Avena_Luo_Cowen_2021}{}}%
Schmid, Moritz S, Dominic Daprano, Kyler M Jacobson, Christopher Sullivan, Christian Briseo-Avena, Jessica Y Luo, and Robert K Cowen. 2021. \emph{A Convolutional Neural Network Based High-Throughput Image Classification Pipeline - Code and Documentation to Process Plankton Underwater Imagery Using Local HPC Infrastructure and NSF's XSEDE}. Zenodo. \url{https://doi.org/10.5281/ZENODO.4641158}.

\leavevmode\vadjust pre{\hypertarget{ref-Taylor2012}{}}%
Taylor, Gordon T., Frank E. Muller-Karger, Robert C. Thunell, Mary I. Scranton, Yrene Astor, Ramon Varela, Luis Troccoli Ghinaglia, et al. 2012. {``Ecosystem Responses in the Southern Caribbean Sea to Global Climate Change.''} \emph{Proceedings of the National Academy of Sciences of the United States of America} 109 (November): 19315--20. \url{https://doi.org/10.1073/pnas.1207514109}.

\leavevmode\vadjust pre{\hypertarget{ref-tittensor2010global}{}}%
Tittensor, Derek P, Camilo Mora, Walter Jetz, Heike K Lotze, Daniel Ricard, Edward Vanden Berghe, and Boris Worm. 2010. {``Global Patterns and Predictors of Marine Biodiversity Across Taxa.''} \emph{Nature} 466 (7310): 1098.

\leavevmode\vadjust pre{\hypertarget{ref-Warren2018}{}}%
Warren, R., J. Price, E. Graham, N. Forstenhaeusler, and J. VanDerWal. 2018. {``The Projected Effect on Insects, Vertebrates, and Plants of Limiting Global Warming to 1.5c Rather Than 2c.''} \emph{Science} 360 (May): 791--95. \url{https://doi.org/10.1126/science.aar3646}.

\leavevmode\vadjust pre{\hypertarget{ref-Wieczorek2012}{}}%
Wieczorek, John, David Bloom, Robert Guralnick, Stan Blum, Markus Dring, Renato Giovanni, Tim Robertson, and David Vieglais. 2012. {``Darwin Core: An Evolving Community-Developed Biodiversity Data Standard.''} \emph{PLoS ONE} 7 (January): 29715. \url{https://doi.org/10.1371/journal.pone.0029715}.

\leavevmode\vadjust pre{\hypertarget{ref-Wilkinson2016}{}}%
Wilkinson, Mark D., Michel Dumontier, IJsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, et al. 2016. {``The FAIR Guiding Principles for Scientific Data Management and Stewardship.''} \emph{Scientific Data} 3 (March): 1--9. \url{https://doi.org/10.1038/sdata.2016.18}.

\leavevmode\vadjust pre{\hypertarget{ref-knitr2014}{}}%
Xie, Yihui. 2014. {``Knitr: A Comprehensive Tool for Reproducible Research in {R}.''} In \emph{Implementing Reproducible Computational Research}, edited by Victoria Stodden, Friedrich Leisch, and Roger D. Peng. Chapman; Hall/CRC. \url{http://www.crcpress.com/product/isbn/9781466561595}.

\leavevmode\vadjust pre{\hypertarget{ref-xie2015}{}}%
---------. 2015b. \emph{Dynamic Documents with {R} and Knitr}. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. \url{http://yihui.name/knitr/}.

\leavevmode\vadjust pre{\hypertarget{ref-knitr2015}{}}%
---------. 2015a. \emph{Dynamic Documents with {R} and Knitr}. 2nd ed. Boca Raton, Florida: Chapman; Hall/CRC. \url{https://yihui.org/knitr/}.

\leavevmode\vadjust pre{\hypertarget{ref-bookdown2016}{}}%
---------. 2016. \emph{Bookdown: Authoring Books and Technical Documents with {R} Markdown}. Boca Raton, Florida: Chapman; Hall/CRC. \url{https://bookdown.org/yihui/bookdown}.

\leavevmode\vadjust pre{\hypertarget{ref-R-bookdown}{}}%
---------. 2021a. \emph{Bookdown: Authoring Books and Technical Documents with r Markdown}. \url{https://CRAN.R-project.org/package=bookdown}.

\leavevmode\vadjust pre{\hypertarget{ref-R-knitr}{}}%
---------. 2021b. \emph{Knitr: A General-Purpose Package for Dynamic Report Generation in r}. \url{https://yihui.org/knitr/}.

\leavevmode\vadjust pre{\hypertarget{ref-rmarkdown2018}{}}%
Xie, Yihui, J. J. Allaire, and Garrett Grolemund. 2018. \emph{R Markdown: The Definitive Guide}. Boca Raton, Florida: Chapman; Hall/CRC. \url{https://bookdown.org/yihui/rmarkdown}.

\leavevmode\vadjust pre{\hypertarget{ref-rmarkdown2020}{}}%
Xie, Yihui, Christophe Dervieux, and Emily Riederer. 2020. \emph{R Markdown Cookbook}. Boca Raton, Florida: Chapman; Hall/CRC. \url{https://bookdown.org/yihui/rmarkdown-cookbook}.

\end{CSLReferences}

\end{document}
